---
title: "Tutorial"
output:
  html_document:
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE, tidy = 'styler')
library(data.table)
library(ggplot2)
# Using dev directory if present
tutorial_dir = system.file("tutorial", package = 'cancereffectsizeR') 
```

This tutorial applies some of the package's key features to analyze publicly-available somatic variant data from tumor sequencing. We recommend that you start a fresh R/RStudio session. You can follow along by copy-and-pasting code into an R script.


## Installation
Start by installing the latest release:
```{r}
# Some dependencies are large, so we increase the download time limit to be safe
options(timeout = 600) 
install.packages('remotes')
remotes::install_github("Townsend-Lab-Yale/cancereffectsizeR@*release", 
                        dependencies = TRUE)
```

Regrettably, there is a bug in some older versions of the GenomeInfoDb package that may produce cryptic error messages in cancereffectsizeR, along the lines of `!anyNA(m32) is not TRUE`. If you encounter this issue, upgrade your Bioconductor version:

```{R}
# Only necessary if the current Bioconductor version is <3.14.
if (BiocManager::version() < as.package_version("3.14")) {
  BiocManager::install(version = "3.14")
}
```

In addition to installing cancereffectsizeR, you need to install a reference data set, or _refset_. These refsets include genomic and gene annotations, mutational signature definitions, and more. Currently, refsets supporting the hg38 and hg19 builds of the human genome are available as separate data packages. (If you want to build your own refset to use a custom genome build or gene set for almost any species, [you can](custom_refset_instructions.html).) For this tutorial, we'll use the hg38 refset package.
```{r}
options(timeout = 600)
remotes::install_github("Townsend-Lab-Yale/ces.refset.hg38@*release")
```

Restart R/RStudio after installation. Before continuing, you may want to create a directory for this tutorial to store all the data and output in one place.
```{r}
# setwd() somewhere, if desired
dir.create('CES_tutorial')
setwd('CES_tutorial')
```


## Quickstart

### Theoretical overview
Very briefly, we extract mutational signatures from each samples's SNV mutation profile using  [MutationalPatterns](https://bioconductor.org/packages/release/bioc/html/MutationalPatterns.html). The relative weights of biologically associated signatures are used to infer trinucleotide-context-specific relative rates of SNV mutations for each sample. Cohort-wide neutral gene mutation rates are calculated with [dNdScv](https://github.com/im3sanger/dndscv), with tissue-specific covariates provided by `ces.refset.hg38`. Combining this information, the rate of neutral mutation at a particular variant site is calculated by partitioning the gene mutation rate across all sites in the gene in accordance with the signature-informed relative rates. Comparing rates of observed and expected mutation under a model of somatic selection allows an inference of selection intensity, which we call cancer effect size.

### Example: Lung adenocarcinoma data from TCGA
If you want to skip all the details and run through a no-frills cancereffectsizeR analysis now, see below. Otherwise, take a peek to get a sense of what's to come, or keep reading for a more thorough walkthrough.

<details><summary>(Show/hide quickstart)</summary>
```{r}
library(cancereffectsizeR)
library(data.table)
library(ggplot2)

# Download lung adenocarcinoma (LUAD) somatic variant data.
# Note that the rest of the tutorial uses BRCA, not LUAD.
# This is an old TCGA file: Outside the tutorial, use get_TCGA_project_MAF() for the
# latest, most accurate data.
tcga_maf_file = 'TCGA.LUAD.mutect.0458c57f-316c-4a7c-9294-ccd11c97c2f9.DR-10.0.somatic.maf.gz'
if (! file.exists(tcga_maf_file)) {
  download.file('https://api.gdc.cancer.gov/data/0458c57f-316c-4a7c-9294-ccd11c97c2f9', 
                destfile = tcga_maf_file, mode = 'wb')
}

# Prepare data
maf = preload_maf(maf = tcga_maf_file, refset = "ces.refset.hg38")

# Create cancereffectsizeR analysis and load data
cesa = CESAnalysis(refset = "ces.refset.hg38")
cesa = load_maf(cesa = cesa, maf = maf)

# Infer trinculeotide-context-specific relative rates of SNV mutation from
# a mutational signature analysis (leaving out signatures not found in LUAD)
signature_exclusions = suggest_cosmic_signature_exclusions(cancer_type = 'LUAD', treatment_naive = TRUE)
cesa = trinuc_mutation_rates(cesa = cesa, signature_set = ces.refset.hg38$signatures$COSMIC_v3.2,
                             signature_exclusions = signature_exclusions)

# Estimate neutral gene mutation rates using dNdScv, with tissue-specific mutation rate covariates.
cesa = gene_mutation_rates(cesa, covariates = ces.refset.hg38$covariates$lung)

# Infer scaled selection coefficients under the default model of clonal selection.
# By default, inference is restricted to recurrent mutations.
cesa = ces_variant(cesa)

# Take top 15 variants, then sort lowest to highest (to plot left to right)
selection = cesa$selection[[1]]
top = selection[order(-selection_intensity)][1:15]
top = top[order(selection_intensity)]

# Plot top effects
top[, display_name := gsub('_', "\n", variant_name)]
top[, display_levels := factor(display_name, levels = display_name, ordered = T)]
plot_title = 'Top cancer effects in TCGA LUAD'
breaks = unique(as.numeric(round(quantile(top$included_with_variant))))

ggplot(top, aes(x = display_levels, y = selection_intensity)) + 
  geom_errorbar(aes(ymin = ci_low_95, ymax = ci_high_95), width = .2, color = 'darkgrey') +
  geom_point(aes(color = included_with_variant), size = 3) + 
  scale_x_discrete() + scale_y_log10() + 
  scale_color_viridis_c(name = 'variant prevalence', guide = 'colorbar', trans = 'log10', 
                        option = 'plasma', breaks = breaks) +
  xlab(element_blank()) +
  ylab(expression('cancer effect'~scriptstyle(~~(log[10])))) +
  ggtitle(plot_title) +
  guides(color = guide_colourbar(ticks = FALSE)) + 
  theme_minimal() + 
  theme(text = element_text(family = "Verdana"),
        axis.title.x = element_text(size = 14), 
        axis.text.x = element_text(size = 8),
        legend.position = 'bottom',
        legend.title = element_text(size = 10),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())

# There's lots more that can be done with cancereffectsizeR! See the full tutorial for some
# important details and a broader view of functionality.
```

```{r echo = F, eval = T}
p = readRDS(system.file("tutorial/top_LUAD_effects.rds", package = 'cancereffectsizeR'))
p
```

</details>


## Preparing data
Load cancereffectsizeR, as well as the data.table package. cancereffectsizeR makes extensive use of data tables, so it's handy to have the package loaded.
```{r}
library(cancereffectsizeR)
library(data.table)
```

For this tutorial, we'll use somatic variant data produced from exome sequencing by the TCGA BRCA (breast carcinoma) project. We will supplement the WXS data with targeted sequencing data from the Metastatic Breast Cancer data set [hosted at cBioPortal](https://www.cbioportal.org/study/summary?id=breast_ink4_msk_2021).

In cancereffectsizeR, data can be combined from multiple whole-exome, whole-genome, and targeted sequencing sources, although we always need at least one WXS or WGS source to anchor the analysis. Targeted sequencing data can't be used for mutation rate inference: there are too few mutations, typically, and since they're mostly in cancer hotspots, they don't provide a baseline of what mutation rates are like in the absence of selection. Instead, mutation rates in TGS samples will be assumed to be similar to those calculated in WXS/WGS samples.


### TCGA data (whole-exome)

We'll begin by downloading variant data from TCGA. The data comes in a large text file listing somatic mutations for a collection of tumors, in accordance with [MAF format](https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format). **This data is from an old release.** For future analyses, get the most recent, accurate TCGA somatic variant data with `get_TCGA_project_MAF()`.

```{r}
tcga_maf_file = 'TCGA.BRCA.mutect.995c0111-d90b-4140-bee7-3845436c3b42.DR-10.0.somatic.maf.gz'
if (! file.exists(tcga_maf_file)) {
  download.file('https://api.gdc.cancer.gov/data/995c0111-d90b-4140-bee7-3845436c3b42', destfile = tcga_maf_file, mode = 'wb')
}
```

Let's also load and examine a table of patient information. Although the TCGA samples in the data set are all taken from primary tumors, you'll see that some patients presented with metastatic disease (pM = M1). Hormone receptor status (progesterone or estrogen receptor positive) and HER2 amplification status are also recorded in a combined column.
```{r eval = T}
tcga_clinical = fread(system.file("tutorial/TCGA_BRCA_clinical.txt", package = "cancereffectsizeR"))
tcga_clinical[1:5]
```

The `preload_maf()` function takes in MAF data, extracts the columns needed by cancereffectsizeR, adds a couple of genomic annotations, and checks for common problems. Since TCGA data sets are well-curated, this particular run will not reveal any problems. (For your own analyses, see our [MAF data tips](MAF_filtering_tips.html).) This MAF file already uses the hg38 genome build, but if it didn't, we could use the `chain_file` argument to convert records via liftOver.

```{r}
tcga_maf = preload_maf(maf = tcga_maf_file, refset = "ces.refset.hg38")
```

If you compare the TCGA MAF and clinical file, you'll see we have a small problem: the clinical file uses patient IDs, while the MAF uses tissue sample IDs. (This column was originally called Tumor_Sample_Barcode, but `preload_maf()` changed it to Unique_Patient_Identifier. cancereffectsizeR uses this column name to emphasize that distinct MAF samples must derive from distinct patients.)

Luckily, TCGA IDs are easy to convert. After verifying that each TCGA patient has just one sample in the MAF data, we'll change tissue IDs to patient IDs.

```{r}
# TCGA sample IDs start with the patient ID, luckily. Here, we verify a 1-1 sample-patient
# relationship. If this check failed, we would need to figure out a sample exclusion
# strategy before proceeding with analysis.
tcga_maf[, patient_id := substr(Unique_Patient_Identifier, 1, 12)]
uniqueN(tcga_maf[, .(patient_id, Unique_Patient_Identifier)]) == uniqueN(tcga_maf$Unique_Patient_Identifier)

# We can now use patient_id as the unique identifier.
tcga_maf[, Unique_Patient_Identifier := patient_id]
tcga_maf[, patient_id := NULL] # remove redundant column
setnames(tcga_clinical, "patient_id", "Unique_Patient_Identifier")
```

### Metastatic data (TGS)
Our TGS data source (hosted on cBioPortal and [recently published](https://cancerdiscovery.aacrjournals.org/content/candisc/early/2021/09/17/2159-8290.CD-20-1726.full.pdf) in Cancer Discovery), consists of metastatic tumors that were sequenced using various MSK-Impact panels. In the full data set, some patients have multiple samples sequenced; the data provided with the package has been subsetted to one sample per patient, and the genome build has been converted to hg38.
```{r}
tgs_maf_file = system.file('tutorial/metastatic_breast_2021_hg38.maf', package = 'cancereffectsizeR')
tgs_maf = preload_maf(maf = tgs_maf_file, refset = 'ces.refset.hg38')
```

When combining data from multiple sources, `check_sample_overlap()` can detect unexpected sample duplication. If you feed in `tcga_maf` and `tgs_maf` and filter the output to `variants_shared > 2` (with panel data, a couple of shared variants doesn't imply sample duplication), you'll see that everything looks fine.

## Create CESAnalysis and load data
The CESAnalysis is the primary data structure of cancereffectsizeR. The cancereffectsizeR workflow consists of calling a series of functions that take a CESAnalysis as input and return an altered CESAnalysis. Re-using the same variable name will keep things tidy.

### Load whole-exome sequencing data
Let's create a CESAnalysis and call `load_maf()` to load the TCGA data.
```{r}
cesa = CESAnalysis(refset = "ces.refset.hg38")
cesa = load_maf(cesa = cesa, maf = tcga_maf)
```

You will see a message that some variants fall outside of the refset's exome definitions. Since we don't know exactly what exome capture techniques were used for the BRCA project (probably different methods at different study sites), this isn't unexpected. (On the off chance that you do know the exact exome capture intervals of whatever data you're using, you should supply those intervals with `load_maf()`'s `covered_regions` argument. For WGS data, simply set `coverage = "genome"`.)

To navigate a CESAnalysis, use the dollar sign ($). We can use these accessors to look at the MAF data as well more detailed variant annotations, and if we want we can perform various filtering operations, such as identifying the most prevalent variants in the loaded data.
```{r}
cesa$maf
cesa$variants
cesa$samples

# Let's see the top variants
(top_variants = cesa$variants[order(-maf_prevalence)][1:10, .(variant_name, chr, start, end, maf_prevalence)])
```

Let's load the clinical data into the analysis, too.
```{r}
cesa = load_sample_data(cesa, tcga_clinical)
```

Now that we've loaded the sample data, we can, for example, see how variant counts break down across different sample groups. Let's see a breakdown PIK3CA mutation status by BRCA receptor phenotype.
```{r}
# Get samples with 1 or more PIK3CA mutation
samples_with_pik3ca = cesa$maf[top_gene == "PIK3CA", unique(Unique_Patient_Identifier)]
pik3ca_sample_counts = cesa$samples[samples_with_pik3ca, .(with_PIK3CA = .N),
                                    by = 'receptor_status', on = 'Unique_Patient_Identifier']
total_sample_counts = cesa$samples[, .N, by = 'receptor_status']
combined_counts = merge.data.table(pik3ca_sample_counts, total_sample_counts, by = 'receptor_status')
combined_counts[, freq := with_PIK3CA / N]
combined_counts
```
PIK3CA mutation is least prevalent in triple negative breast cancer.


### Load targeted gene sequencing data
To load our TGS data, we need to be able to define its coverage. Why? To estimate the selection of a mutation, we need to know which samples have the mutation, which samples do not, and which are unknown due to lack of sequencing coverage. Our TGS samples were sequenced with multiple panels covering different genes, and unfortunately, the exact coverage (defined by genomic coordinates) is not publicly available for all of these. Therefore, for simplicity, we will filter the TGS data to mutations at a handful of top cancer genes that are covered in all of the panels. When we load the data, all records outside of these genes will be excluded, leaving us with greater power to infer selection in these genes without impacting our estimates outside these genes, which will use just the TCGA data. (As mentioned earlier, we shouldn't filter WXS/WGS data this way, since it would interfere with mutation rate calculation, but TGS samples are not involved in mutation rate calculation.)

```{r}
# Define coverage using the coding regions of these genes, as defined by the refset
top_tgs_genes = c("TP53", "PIK3CA", "ESR1","CDH1","GATA3","KMT2C",
                      "MAP3K1","AKT1","ARID1A","FOXA1","TBX3","PTEN")
tgs_coverage = ces.refset.hg38$gr_genes[ces.refset.hg38$gr_genes$gene %in% top_tgs_genes]
```

We don't have a clinical file for the TGS data, but we do know that they're all metastatic samples. Instead of creating another table and calling `load_sample_data()`, we can use the `sample_data_cols` feature of load_maf():
```{r}
tgs_maf$pM = 'M1'
cesa = load_maf(cesa, maf = tgs_maf, sample_data_cols = 'pM', 
                coverage = 'targeted', covered_regions = tgs_coverage, covered_regions_name = 'top_genes')
```

Now that all the data is loaded, we'll show one data-wrangling example: Let's look at variant prevalence by metastatic disease status. By default, the `variant_counts()` function considers all mutations, but we'll subset to recurrent mutations.
```{r}
counts_by_M = variant_counts(cesa = cesa, variant_ids = cesa$variants[maf_prevalence > 1, variant_id],
                             by = 'pM')
counts_by_M[M1_prevalence > 0][order(-M1_prevalence)]
```

## Mutational processes and relative mutation rates
For each (exome) sample in our data set, `trinuc_mutation_rates()` will perform mutational signature extraction in order to to attribute each sample's set of SNVs to a linear combination of mutational processes. By default, the signature extraction is done using the `MutationalPatterns` package, and `deconstructSigs` is also supported. The signature attribution allows us to infer sample-specific relative rates of SNV mutation for all trinucleotide contexts. TGS samples will be assumed to have mutational processes matching the group-average mutational processes of the exome data. (This isn't optimal, since the metastatic samples may be affected by mutational processes, such as chemotherapy, that are lacking in the primary tumor exome data, but it might be the best we can do without finding another data source.)

We will use signature definitions from the refset (it's also possible to create your own). To improve the accuracy of signature extraction, we will exclude signatures that can safely be presumed absent from the samples; a helper function, `suggest_cosmic_signature_exclusions()`, can provide some guidance.

```{r}
# We'll use all suggested exclusions (TCGA primary tumors are treatment-naive)
signature_exclusions = suggest_cosmic_signature_exclusions(cancer_type = 'BRCA', treatment_naive = TRUE)

cesa = trinuc_mutation_rates(cesa, signature_set = ces.refset.hg38$signatures$COSMIC_v3.2, 
                             signature_exclusions = signature_exclusions)
```


The `trinuc_mutation_rates()` run has added some useful information to the CESAnalysis:

* **snv_counts**: A matrix of SNV counts by trinucleotide context. (You can also generate such a matrix from MAF data with `trinuc_snv_counts()`.)
* **raw_attributions**: Signature attributions as generated by the extractor. For MutationalPatterns, this matrix contains the number of mutations attributed to signature for each sample. Naturally, samples with more mutations will tend to have higher raw attributions.
* **biological_weights**: The proportion of mutations attributed to each biologically-associated signature within each sample. Mutations attributed to signatures associated with sequencing/processing artifacts are left out. (Technical detail: Due to the instability of signature attributions on samples with few mutations, Samples with few MAF variants have their weights adjusted towards group-average weights, as indicated in the `group_avg_blended` column. If you want to make claims about subgroup differences in mutational processes, consider leaving these samples out or using the raw attributions. You should also leave out the TGS samples, which in this analysis are treated as having 0 mutations.)
* **trinuc_rates**: Inferred relative rates of mutation, produced by matrix-multiplying biological_weights and signature definitions. (These rates will not equal empirical rates of observed mutations.)

While we're here, MutationalPatterns has some nice plot functions that we can use. For example, we can plot the mean SNV profile for each BRCA receptor subtype. (We would want to exclude TGS samples in this plot because cancer hotspot mutations are not reflective of overall mutational processes, but our TGS samples lack receptor_status information, so they will already be left out without having to specifically exclude them.)

```{r eval = T, include = F}
snv_counts = readRDS(system.file("tutorial/BRCA_snv_counts.rds", package = 'cancereffectsizeR'))
cesa = list(samples = readRDS(system.file("tutorial/BRCA_cesa_samples.rds", package = 'cancereffectsizeR')))
```

```{r}
snv_counts = cesa$mutational_signatures$snv_counts
```
```{r eval = T}
# Counts look like this:
snv_counts[1:5, c("TCGA-AC-A3W5", "TCGA-AC-A3YI", "TCGA-AC-A3YJ", "TCGA-AC-A4ZE", "TCGA-AC-A5EH", "TCGA-AC-A5EI")]

# We'll sum up the counts by group, leaving out a sample that has no SNVs.
summed_snv_by_group = data.table()
receptor_groups = unique(na.omit(cesa$samples$receptor_status))
samples_with_snvs = cesa$samples[colnames(snv_counts), on = "Unique_Patient_Identifier"]
for (grp in receptor_groups) {
  curr_samples = samples_with_snvs[grp, Unique_Patient_Identifier, on = 'receptor_status']
  curr_snv_sum = rowSums(snv_counts[, curr_samples])
  summed_snv_by_group[, (grp) := curr_snv_sum]
}
summed_snv_by_group = as.matrix(summed_snv_by_group)
rownames(summed_snv_by_group) = rownames(snv_counts)
MutationalPatterns::plot_96_profile(summed_snv_by_group)
```

## Gene mutation rates
Next, we'll use `gene_mutation_rates()` to estimate regional rates of mutation in the absence of selection. We'll use the method provided in the dNdScv package, which uses dN/dS ratios and mutation rate covariates. Our refset has pre-computed covariates for a variety of tissue types. As with mutational processes, this analysis will not use the TGS samples, but the calculated rates will be assumed to hold for them.
```{r}
cesa = gene_mutation_rates(cesa, covariates = ces.refset.hg38$covariates$breast)
```

Here are the neutral gene mutation rates extracted from dNdScv's regression:
```{r eval = T, include = F}
cesa = list(gene_rates = readRDS(system.file("tutorial/BRCA_cesa_gene_rates.rds", 
                                             package = 'cancereffectsizeR')),
            dNdScv_results = readRDS(system.file("tutorial/BRCA_dndscv_out.rds", 
                                                 package = 'cancereffectsizeR')))
```

```{r eval = T}
head(cesa$gene_rates)
```

We can also look at dNdScv's identification of selection at the gene level. We'll filter results to q < .05. Since ces.refset.hg38 covers multiple isoforms for some genes, the dNdScv output is actually per protein, not per gene, so we'll  take the extra step of showing just the lowest q-value for each significant gene.

```{r eval = T}
dndscv_results = cesa$dNdScv_results[[1]]
sig_genes = dndscv_results[qallsubs_cv < .05]

# Take lowest q-value record for each gene, and sort by significance.
sig_genes = sig_genes[, .SD[which.min(qallsubs_cv)], by = 'gene'][order(qallsubs_cv)]
```

```{r eval = T, echo = F}
formattable::formattable(sig_genes)
```

See dNdScv's documentation if you're interested in interpretation of the dNdScv output.

Side note: If we had multiple tissue types in the analysis, we could call gene_mutation_rates() multiple times using the `samples` argument to specify sample groups, with appropriate covariates for each tissue type. Similar functionality is available in `trinuc_mutation_rates()`.

## Selection inference
The combination of cohort-level gene mutation rates and sample-level relative rates of trinucleotide-context-specific substitution allow us to estimate the rate at which any somatic substitution occurs in any patient's tumor sample. To be clear, by "rates," we don't mean the frequency of mutations in our somatic variant calls (which we already know anyway). These rates represent how often the mutations can be expected to occur in individual cells. (More specifically, we assume that mutation events in each specific sample and site follow a Poisson distribution, and these rates are the Poisson rates.) It's not typically necessary to look at these rates, but just to show that we can:
```{r}
# Let's take the top 3 variants by MAF prevalence. These happen to all be in PIK3CA.
variants_to_check = cesa$variants[order(-maf_prevalence), variant_id][1:3]

# A few random samples
samples_to_check = c('TCGA-A2-A3Y0', 'TCGA-XX-A89A', 'P-0000224')

baseline_mutation_rates(cesa = cesa, variant_ids = variants_to_check, samples = samples_to_check)
```

```{r eval = T, echo = F}
readRDS(system.file("tutorial/BRCA_site_rates_example.rds", package = "cancereffectsizeR"))
```

Interestingly, PIK3CA H1047R, the most prevalent variant in our MAF data (246/1875 samples), is estimated to have a substantially lower rate of occurrence than E545K and E542K in these particular samples.

### Basic model
We can use our estimated rates and the MAF variant data to infer scaled selection coefficients (aka cancer effects, or selection intensity) under a model of selection. These cancer effects are directly proportional to the proliferative advantage provided by each variant. First, let's calculate selection under the package's default model of selection with ces_variant():
```{r}
# Including an optional run_name
cesa = ces_variant(cesa = cesa, run_name = 'recurrents')
```

Let's visualize the most selected variants. The following should be fairly adaptable for a variety of data sets.
```{r}
library(ggplot2)

# Extract selection results from CESAnalysis and take top variants for visualization
top = cesa$selection$recurrents
top = top[order(-selection_intensity)][1:20] # take top 20 by SI
top = top[order(selection_intensity)] # will plot lowest to highest (left to right)

# Make variant names pretty for use in plot labels
top[, display_name := gsub('_', "\n", variant_name)]
top[, display_levels := factor(display_name, levels = display_name, ordered = T)]

plot_title = 'Top cancer effects in breast carcinoma (CES tutorial data)'
breaks = unique(as.numeric(round(quantile(top$included_with_variant))))
n.dodge = 2 # can reduce to 1 if labels still fit (e.g., if plotting fewer variants)
ggplot(top, aes(x = display_levels, y = selection_intensity)) + 
  geom_errorbar(aes(ymin = ci_low_95, ymax = ci_high_95), width = .2, color = 'darkgrey') +
  geom_point(aes(color = included_with_variant), size = 3) + 
  scale_x_discrete(guide = guide_axis(n.dodge = n.dodge)) + scale_y_log10() + 
  scale_color_viridis_c(name = 'variant prevalence', guide = 'colorbar', trans = 'log10', 
                        option = 'plasma', breaks = breaks) +
  xlab(element_blank()) +
  ylab(expression('cancer effect'~scriptstyle(~~(log[10])))) +
  ggtitle(plot_title) +
  guides(color = guide_colourbar(ticks = FALSE)) + 
  theme_minimal() + 
  theme(text = element_text(family = "Verdana"),
        axis.title.x = element_text(size = 14), 
        axis.text.x = element_text(size = 8),
        legend.position = 'bottom',
        legend.title = element_text(size = 10),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())
```

```{r eval = T, echo = F}
readRDS(paste0(tutorial_dir, '/top_BRCA_effects.rds'))
```

_PIK3CA_ and _ESR1_ variants take many of the top spots. The deconvolution of mutation rate from selection has revealed that some relatively low-prevalence variants have high effects. Interestingly, according to dNdScv, mutation in _ESR1_ at the gene level was not significantly greater than neutral in the TCGA cohort of primary tumors. The incorporation of metastatic TGS data and cancereffectsizeR's assessment of selection at the variant level confirm _ESR1_'s importance.

### Sequential model
In the sequential model, selection is allowed to vary across two or more theoretical timepoints represented in the data. These timepoints should be at least arguably chronological, though not necessarily strictly so. For example, if we assume that tumors progress sequentially from grade I to grade III, we could adopt a three-state sequentially model. For our data set, since the pM annotation tells us which patients had progressed to metastatic disease at the time of sample collection, we'll do a two-state model with states M0 and M1. (One could certainly argue about whether the few TCGA M1 patients should be lumped in with the M1 TGS data, since the TCGA samples are from primary tissue and the TGS samples are metastatic tissue, but we'll proceed under the assumption that metastasis occurred due to continued evolution of primary tumors.)

Under this model, we will find two cancer effects per variant. The idea is that the M0 effect reflects selection for the variant in the M0 samples. In the M1 samples, the M0 effect applied until a point of transition that they have all passed, and since then the M1 effect has held. (This transition may occur at different times in different samples.)

Since the goal here is to find variants that are differentially selected early and late in the disease, we will restrict our inference to variants that occur more than twice in M0/M1 samples. (We wouldn't be able to make a convincing case for differential selection for a variant that appears just once or twice in the cohort.) We can use the `counts_by_M` that we generated earlier to get the eligible variant IDs, and then pull their full annotations from the variants table:
```{r}
# counts_by_M produced earlier with a call to variant_counts()
variants_for_sequential = counts_by_M[M0_prevalence + M1_prevalence > 2, variant_id]
variants_for_sequential = cesa$variants[variants_for_sequential, on = 'variant_id']
```

Now, we'll run the sequential model, specifying our variants with the `variants` argument.
```{r}
cesa = ces_variant(cesa, variants = variants_for_sequential, model = 'sequential', run_name = 'sequential', 
                   ordering_col = 'pM', ordering = c('M0', 'M1'))
```

Since not all samples have pM annotation, we'll also re-evaluate single-step cancer effects at these sites, using the `samples` argument to restrict inference to the same samples used in the sequential run. This will let us perform a likelihood ratio test to identify variants whose prevalence is significantly better explained by the two-step model.
```{r}
# Assess the same variants in the same samples using the single model
cesa = ces_variant(cesa, variants = variants_for_sequential, model = 'basic', run_name = 'for_sequential_compare',
                   samples = cesa$samples[!is.na(pM)])
```


We merge the two selection output tables, do the likelihood ratio test, and print a simplified table of significantly differentially selected variants.
```{r}
combined_results = merge.data.table(cesa$selection$sequential, 
                                    cesa$selection$for_sequential_compare, 
                                    all.x = TRUE, all.y = FALSE, 
                                    by = c('variant_id', 'variant_name', 'variant_type'), 
                                    suffixes = c('.sequential', '.single'))

# Likelihood ratio test
combined_results[, chisquared := -2 * (loglikelihood.single - loglikelihood.sequential)]
combined_results[, p := pchisq(chisquared, df = 1, lower.tail = F)]

# Prep summary output for printing. Not shown here, but all of these variants are covered
# in both of our data sources.
for_print = combined_results[, .(variant_name, si_single = selection_intensity, si_M0, si_M1, p,
                                 M0_count = included_with_variant_M0, M1_count = included_with_variant_M1)]
sequential_signif_output = for_print[p < .05][order(p)]
sequential_signif_output
```

```{r eval = T, echo = F}
sequential_output_file = paste0(tutorial_dir, '/sequential_signif_output.rds')
formattable::formattable(readRDS(sequential_output_file), digits = 4)
```

The increased selection in M1 for _ESR1_ variants is immediately apparent, and it's especially convincing that so many distinct variants in the gene show the same relationship. Observe that the PIK3CA variants show reduced M1 selection even though the variants actually appear at similar or higher frequencies in the M1 samples. Under the model assumptions, the M1 samples previously experienced M0 selection pressures, so we actually expect these samples to have prevalent PIK3CA mutations regardless of M1 selection.

Relatively few variants are prevalent enough to test under sequential models, and confidence intervals are often wide. Several of the significant variants are given M0 effects of .001, the lower limit of our optimization, simply because the variants do not appear in the M0 group. Confidence intervals reflect that selection intensity at these sites could be much higher.

### More options
By default, all recurrent variants (variants that appear at least twice in MAF data) are included in a `ces_variant()` run. As demonstrated, the `variants` argument can be used to specify which variants to test. Variants with prevalences of 1 or even 0 (perhaps helpful to establish an effect ceiling) can be included: `variants = cesa$variants` will include all annotated mutations in the analysis. Note that most single-hit variants will have over-estimated effects because most are probably neutral passengers, and the occurrence of any non-selected variant is inherently improbable.

For each variant site, only samples with coverage at the site will inform selection inference. As also shown above, the `samples` argument further limits the inference process to a subset of eligible samples. This could be helpful to assess and compare selection in various groups (e.g., smoking vs. non-smoking).

Besides the basic and sequential models, `ces_variant()` supports user-supplied models of selection; see documentation for details. We hope that some enterprising users will come up with new models, and we are happy to discuss ideas and help work out any kinks.

The `CompoundVariantSet` feature, described in the next section, may sometimes provide improved resolution of somatic selection by batching related variants together. 

## Epistatic cancer effects
In the basic model, a variant is assumed to have a single cancer effect across all samples. In reality, we expect a variant's selection to be influenced by a complex combination of factors including a sample's mutational background (from substitutions to structural variants), the epigenetic state of the tumor, and various environmental factors. The `ces_epistasis()` function allows us to assess selection for any pair of variants under a model of pairwise epistasis, in which the selection intensity for each variant is allowed to very depending on the state of the other site.

### Variant-level epistasis
Let's try two prevalent BRCA mutations, PIK3CA E545K (n=159) and AKT1 E17K (n=82). Let's also test E545K with a nearby PIK3CA variant, E542K (n=93).

```{r}
# Start by pulling full variant IDs (with protein identifier) from variants table
group1 = cesa$variants[c("PIK3CA_E545K", "AKT1_E17K"), variant_id, on = 'variant_name']
group2 = cesa$variants[c("PIK3CA_E545K", "PIK3CA_E542K"), variant_id, on = 'variant_name']

cesa = ces_epistasis(cesa = cesa, variants = list(group1, group2), 
                     conf = .95, run_name = 'variant_epistasis_example')
cesa$epistasis$variant_epistasis_example
```

```{r eval = T, echo = F}
readRDS(paste0(tutorial_dir, '/variant_ep_example.rds'))
```

Confidence intervals tend to be wide in epistatic analyses. Even with the most prevalent somatic variants, both co-occurrence and mutual exclusivity can often be explained by chance. The output for PIK3CA E545K and E542K confirms the negative epistatic relationship that we would have expected given their mutual exclusivity in our data (and our knowledge of their biological effects). For E545K and AKT1 E17K, we see reduced selection for both after the acquisition of the other driver, but the confidence intervals leave open the possibility that selection for E545K is unaffected by AKT1 E17K status. (Additionally, the NA's on some of the lower bounds indicate that they fall below the lower limit of the optimization algorithm.)

### Epistasis and CompoundVariantSets
We can probe this PIK3CA/AKT1 relationship further if we assume that all prevalent PIK3CA mutations share the same epistatic relationship with AKT1 E17K. The `define_compound_variants()` feature lets us combine arbitrary variants into "compound variants" that are treated as if they were single variants by cancereffectsizeR's selection inference functions. (In brief, the mutation rate of the compound variant is equal to the sum of rates of constituent variants, and any sample with one or more of the constituent variants "has the compound variant.") Below, we define a CompoundVariantSet with two compound variants: All _PIK3CA_ variants with MAF prevalence > 1, and AKT E17K. We then pass the CompoundVariantSet to ces_epistasis(), which will test all pairs of compound variants (here, just the one pair).

```{r}
# Collect all the variants that we want in the CompoundVariantSet into a table
top_PIK3CA = cesa$variants[gene == 'PIK3CA' & maf_prevalence > 1]
top_akt1 = cesa$variants[variant_name == 'AKT1_E17K']
for_compound = rbind(top_PIK3CA, top_akt1)

# see define_compound_variants() documentation for details on arguments
comp = define_compound_variants(cesa = cesa, variant_table = for_compound, by = "gene", merge_distance = Inf)
cesa = ces_epistasis(cesa = cesa, variants = comp, run_name = "AKT1_E17K_vs_PIK3CA")
cesa$epistasis$AKT1_E17K_vs_PIK3CA
```

```{r eval = T, echo = F}
readRDS(paste0(tutorial_dir, '/comp_variant_ep.rds'))
```

Note that the effect sizes are smaller for PIK3CA here, since we mixed in lots of lower-effect PIK3CA variants. While four samples have both AKT1 E17K and one of the PIK3CA variants, this result strengthens the case for AKT1 mutation reducing selection for PIK3CA mutation.

### Gene-level epistasis
The convenience function `ces_gene_epistasis()` provides a simpler way to apply the model of `ces_epistasis()` at the gene level. The `variants` argument offers multiple methods of specifying which variants from each gene to include in the inference: "recurrent" uses all recurrent variants, "nonsilent" uses all coding variants except non-splice-site synonymous variants, or a custom table of variants can be input, as shown here.
```{r}

genes = c("AKT1", "PIK3CA", "TP53")

# We would use the "nonsilent" method, except a handful of variants are not covered in the TGS data,
# forcing us to exclude those so that the TGS samples can be included inthe inference.
variants = cesa$variants[variant_type == 'aac' & gene %in% genes & sapply(covered_in, length) == 2]
variants = variants[aa_ref != aa_alt | essential_splice == T]
cesa = ces_gene_epistasis(cesa = cesa, genes = genes, variants = variants, run_name = "gene_epistasis_example")
cesa$epistasis$gene_epistasis_example
```

```{r eval = T, echo = F}
readRDS(paste0(tutorial_dir, '/gene_ep_example.rds'))
```

Mutational selection in the two oncogenes appears to be maintained after mutation in TP53. In constrast, loss of TP53 is less selected after mutations in AKT1/PIK3CA; one explanation is that cell populations in which these powerful drivers have fixed must have already subverted tumor suppression mechanisms, making TP53 mutations superfluous. The two oncogenes also appear to show some mutual exclusivity, in that selection for each is reduced after mutation in the other.


## Save your work
Whether you've made it all the way through this tutorial--or simply want to take a break--you can save a commemorative (and space-efficient) copy of your CESAnalysis using `save_cesa()`. You can reload the analysis anytime with `load_cesa()`.

```{r}
save_cesa(cesa = cesa, 'cancerffectsizeR_tutorial_analysis.rds')

# In some future R session...
library(cancereffectsizeR)
cesa = load_cesa('cancerffectsizeR_tutorial_analysis.rds')
```






