# data from https://cancer.sanger.ac.uk/cosmic/signatures

signatures_names_matrix <- matrix(
  data =
    c("SBS1", "Deamination with age (1)",
      "SBS2", "APOBEC (2)",
      "SBS3" ,"Defective homologous recombination (3)",
      "SBS4", "Tobacco smoking (4)" ,
      "SBS5", "Unknown, clock-like (5)",
      "SBS6", "Defective DNA mismatch repair (6)",
      "SBS7a", "UV light (7a)",
      "SBS7b", "UV light (7b)",
      "SBS7c", "UV light (7c)",
      "SBS7d", "UV light (7d)",
      "SBS8", "Unknown (8)",
      "SBS9", "Polymerase eta (9)",
      "SBS10a", "Polymerase epsilon (10a)",
      "SBS10b", "Polymerase epsilon (10b)",
      "SBS11", "Alkylating agents (11)",
      "SBS12" , "Unknown (12)",
      "SBS13" , "APOBEC (13)",
      "SBS14", "Concurrent polymerase epsilon mutation and defective mismatch repair (14)",
      "SBS15", "Defective mismatch repair (15)",
      "SBS16", "Unknown (16)",
      "SBS17a", "Unknown (17a)",
      "SBS17b", "Unknown (17b)",
      "SBS18", "Damage by reactive oxygen species (18)",
      "SBS19", "Unknown (19)",
      "SBS20", "Concurrent POLD1 mutations and defective mismatch repair (20)",
      "SBS21", "DNA mismatch repair deficiency (21)",
      "SBS22", "Aristolochic acid exposure (22)",
      "SBS23" , "Unknown (23)",
      "SBS24", "Aflatoxin exposure (24)",
      "SBS25", "Unknown, perhaps chemotherapy (25)",
      "SBS26", "Defective DNA mismatch repair (26)",
      "SBS28", "Unknown (28)",
      "SBS29", "Tobacco chewing (29)",
      "SBS30", "Deficiency in base excision repair due to mutations in NTHL1 (30)",
      "SBS31", "Prior chemotherapy treatment with platinum drugs (31)",
      "SBS32","Treatment with azathioprine to induce immunosuppression (32)",
      "SBS33", "Unknown (33)",
      "SBS34", "Unknown (34)",
      "SBS35","Prior chemotherapy treatment with platinum drugs (35)",
      "SBS36", "Defective base excision repair and DNA damage from various sources (36)",
      "SBS37","Unknown (37)",
      "SBS38", "Unknown, potential indirect damage from UV-light (38)",
      "SBS39", "Unknown (39)",
      "SBS40", "Unknown (40)",
      "SBS41", "Unknown (41)",
      "SBS42", "Occupational exposure to haloalkanes (42)",
      "SBS44", "Defective DNA mismatch repair (44)",
      "SBS84", "Activity of activation-induced cytidine deaminase (AID) (84)",
      "SBS85", "Indirect effects of activation-induced cytidine deaminase (AID) (85)",
      "SBS27" , "* Possible sequencing artifact (27)",
      "SBS43", "* Possible sequencing artifact (28)",
      "SBS45", "* Possible artifact due to 8-oxo-guanine introduced during sequencing (45)",
      "SBS46", "* Possible sequencing artifact (46)",
      "SBS47", "* Possible sequencing artifact (47)",
      "SBS48", "* Possible sequencing artifact (48)" ,
      "SBS49", "* Possible sequencing artifact (49)",
      "SBS50", "* Possible sequencing artifact (50)",
      "SBS51" , "* Possible sequencing artifact (51)",
      "SBS52", "* Possible sequencing artifact (52)",
      "SBS53", "* Possible sequencing artifact (53)",
      "SBS54", "* Possible sequencing artifact and/or contamination with germline variants (54)",
      "SBS55", "* Possible sequencing artifact (55)",
      "SBS56", "* Possible sequencing artifact (56)",
      "SBS57" , "* Possible sequencing artifact (57)",
      "SBS58", "* Potential sequencing artifact (58)",
      "SBS59", "* Potential sequencing artifact (59)",
      "SBS60", "* Potential sequencing artifact (60)"
    ),ncol=2,byrow = T
)

rownames(signatures_names_matrix) <- signatures_names_matrix[,1]

save(signatures_names_matrix, file = "data/signatures_names_matrix.RData")
