<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Calculate Cancer Effect Size • cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Calculate Cancer Effect Size">
<meta name="description" content="cancereffectsizeR, an R package from the Townsend Lab at Yale School of Public Health for quantifying the effect size of somatic mutations in cancer
">
<meta property="og:description" content="cancereffectsizeR, an R package from the Townsend Lab at Yale School of Public Health for quantifying the effect size of somatic mutations in cancer
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/cancereffectsizeR.html">Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a></li>
    <li><a class="dropdown-item" href="articles/create_custom_covariates.html">Generate tissue covariates data</a></li>
    <li><a class="dropdown-item" href="articles/custom_refset_instructions.html">Running with custom reference data</a></li>
    <li><a class="dropdown-item" href="articles/MAF_filtering_tips.html">MAF filtering and sample validation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">

<h1 style="margin-bottom:0px" id="placeholder-leave-1">
<span class="grad1">cancer</span><span class="grad2">effect</span><span class="grad3">sizeR</span>
<a class="anchor" aria-label="anchor" href="#placeholder-leave-1"></a>
</h1>
<div class="section level4">
<h4 id="quantify-somatic-evolution-in-cancer">Quantify somatic evolution in cancer<a class="anchor" aria-label="anchor" href="#quantify-somatic-evolution-in-cancer"></a>
</h4>
<hr>
<p>Welcome to cancereffectsizeR! This R package provides a variety of tools for analyzing somatic variant data and characterizing the evolutionary trajectories of cancers. Key package features and related theory are presented in a <a href="https://aacrjournals.org/cancerres/article/83/4/500/716429/Estimation-of-Neutral-Mutation-Rates-and" class="external-link">recent article</a> in <em>Cancer Research</em>. As we continue development of cancereffectsizeR, we welcome your <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/issues" class="external-link">feedback, questions, and bug reports</a>.</p>
</div>
<div class="section level2">
<h2 id="installation-tutorial-and-customizations">Installation, tutorial, and customizations<a class="anchor" aria-label="anchor" href="#installation-tutorial-and-customizations"></a>
</h2>
<p>For simple installation instructions and a demonstration of a cancer effect analysis, see the <a href="articles/cancereffectsizeR.html">tutorial</a>. The quickstart section offers a condensed introduction to get you running a basic analysis in minutes.</p>
<p>The package has extensive support for testing specific research questions with customized analyses:</p>
<ul>
<li>Quantify the cancer effects of variants under the default model of selection, test epistatic models of selection, or define and test your own models. Arbitrarily batch variants by position, gene, or functional annotation, and quantify selection by batch. Assess differential selection among patient subgroups.</li>
<li>Combining mutational signature analysis with cancer effect estimation, compare signatures’ relative contributions to oncogenesis and mutagenesis, using either COSMIC signatures or any custom signature set.</li>
<li>Annotate somatic variants with built-in reference data, or create a <a href="articles/custom_refset_instructions.html">custom reference data set</a> for almost any species/genome.</li>
<li>Use provided tissue-specific covariates to inform calculation of gene mutation rates via <a href="https://github.com/im3sanger/dndscv" class="external-link">dNdScv</a>, or build your own <a href="articles/create_custom_covariates.html">custom covariates</a>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="selected-publications">Selected publications<a class="anchor" aria-label="anchor" href="#selected-publications"></a>
</h2>
<ul>
<li><p><strong><a href="https://aacrjournals.org/cancerres/article/83/4/500/716429/Estimation-of-Neutral-Mutation-Rates-and" class="external-link">Estimation of Neutral Mutation Rates and Quantification of Somatic Variant Selection Using cancereffectsizeR</a></strong>, <em>Cancer Research</em> (2023).<br>This resource report discusses the package’s key features and methods and presents analyses validating that cancer effects are a useful quantification of the cancer relevance of somatic variants.</p></li>
<li><p><strong><a href="https://academic.oup.com/mbe/article/39/5/msac084/6570859" class="external-link">Attribution of Cancer Origins to Endogenous, Exogenous, and Preventable Mutational Processes</a></strong>, <em>Molecular Biology and Evolution</em> (2022).<br>Cancer effects are incorporated into a novel method to calculate the relative contributions of various mutational processes to oncogenesis. Apply the method yourself with <code><a href="reference/mutational_signature_effects.html">mutational_signature_effects()</a></code>.</p></li>
<li><p><strong><a href="https://doi.org/10.1093/jnci/djy168" class="external-link">Effect Sizes of Somatic Mutations in Cancer</a></strong>, <em>Journal of the National Cancer Institute </em>(2018).<br>A pan-cancer analysis of cancer effects employing <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/releases/tag/0.1.0" class="external-link">version 0.1.0</a> of this package. This original version was developed by Vincent Cannataro, Stephen Gaffney, and Jeffrey Townsend.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>When reporting work that uses cancereffectsizeR, please cite our <a href="https://aacrjournals.org/cancerres/article/83/4/500/716429/Estimation-of-Neutral-Mutation-Rates-and" class="external-link">resource report</a> published in <em>Cancer Research</em>:</p>
<blockquote>
<p>Mandell JD, Cannataro VL, Townsend JP. Estimation of neutral mutation rates and quantification of somatic variant selection using cancereffectsizeR. Cancer Research. 2023 Feb 15; 83(4):500-505. <a href="https://www.doi.org/10.1158/0008-5472.CAN-22-1508" class="external-link">doi:10.1158/0008-5472.CAN-22-1508</a>.</p>
</blockquote>
</div>
</div>
  </main><aside class="col-md-3"><div class="provided-by">
<h2 data-toc-skip>Provided by</h2>
<ul class="list-unstyled">
<li><p>the <a href="https://medicine.yale.edu/lab/townsend/" class="external-link">Townsend Lab</a> at the<br><a href="https://ysph.yale.edu" class="external-link">Yale School of Public Health</a>.</p></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><p><a href="#citation">How to cite cancereffectsizeR</a></p></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Vincent L. Cannataro <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6364-7747" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jeff Mandell <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-3839-2543" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
