<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cancereffectsizeR, an R package from the Townsend Lab at Yale School of Public Health for quantifying the effect size of somatic mutations in cancer
">
<title>Calculate Cancer Effect Size • cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Calculate Cancer Effect Size">
<meta property="og:description" content="cancereffectsizeR, an R package from the Townsend Lab at Yale School of Public Health for quantifying the effect size of somatic mutations in cancer
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.3.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="articles/custom_refset_instructions.html">Running with custom reference data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">

<h1 style="margin-bottom:0px" id="placeholder-leave-1">
<span class="grad1">cancer</span><span class="grad2">effect</span><span class="grad3">sizeR</span>
<a class="anchor" aria-label="anchor" href="#placeholder-leave-1"></a>
</h1>
<div class="section level4">
<h4 id="quantify-somatic-evolution-in-cancer">Quantify somatic evolution in cancer<a class="anchor" aria-label="anchor" href="#quantify-somatic-evolution-in-cancer"></a>
</h4>
<hr>
<p>Welcome to cancereffectsizeR! This R package contains a variety of tools for analyzing somatic variant data and ultimately characterizing the evolutionary trajectories of cancers.</p>
<p>Features include:</p>
<ul>
<li>Annotate somatic variants with built-in reference data, or create a <a href="articles/custom_refset_instructions.html">custom reference data set</a> for almost any species/genome.</li>
<li>Attribute mutations to mutational processes by extracting COSMIC signatures, or any custom set of signature definitions.</li>
<li>Use provided tissue-specific covariates to inform calculation of gene mutation rates via <a href="https://github.com/im3sanger/dndscv" class="external-link">dNdScv</a>, or build your own <a href="articles/create_custom_covariates.html">custom covariates</a>.</li>
<li>Test single-variant, stage/grade-specific, and epistatic models of selection, or define and test your own models.</li>
<li>Arbitrarily batch variants by position, gene, or functional annotation, and quantify selection by batch.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>cancereffectsizeR requires R 3.5 or later and can be installed from its GitHub repository via the remotes package. Since we’re not able to test all installation environments, please help by letting us know if you have any installation problems.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Townsend-Lab-Yale/cancereffectsizeR@*release"</span>, 
                         dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>For a minimal installation, leave out the dependencies argument to install just required dependencies, rather than both required and suggested.</p>
<p>Regrettably, there is a bug in some older versions of the GenomeInfoDb package that may produce cryptic error messages in cancereffectsizeR, along the lines of <code>!anyNA(m32) is not TRUE</code>. If you encounter this issue, upgrade your Bioconductor version:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Only necessary if the current Bioconductor version is &lt;3.14.</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (BiocManager<span class="sc">::</span><span class="fu">version</span>() <span class="sc">&lt;</span> <span class="fu">as.package_version</span>(<span class="st">'3.14'</span>)) {</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="at">version =</span> <span class="st">"3.14"</span>)<span class="st">`</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="at">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="methods-overview">Methods overview<a class="anchor" aria-label="anchor" href="#methods-overview"></a>
</h2>
<p>For a thorough introduction to the package’s usage and underlying methods, see the <a href="articles/cancereffectsizeR.html">tutorial</a>. The quickstart section has a condesned intro to both usage and theory.</p>
<p>Currently, cancereffectsizeR directly supports analysis of both noncoding and amino-acid-changing SNVs. Current methods provide some insight into interactions with other types of selection; for example, samples could be grouped by copy number status (or any other feature of interest) and assessed for differential SNV selection. We plan to extend cancereffectsizeR’s functionality as we continue development, and we welcome your feedback, ideas, and bug reports.</p>
</div>
<div class="section level2">
<h2 id="publications-and-version-note">Publications and version note<a class="anchor" aria-label="anchor" href="#publications-and-version-note"></a>
</h2>
<p>Our 2018 JNCI paper <a href="https://doi.org/10.1093/jnci/djy168" class="external-link">Effect sizes of somatic mutations in cancer</a> describes <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/releases/tag/0.1.0" class="external-link">Version 0.1.0</a> of this package, which was developed by Cannataro, V. L., Gaffney, S. G., and Townsend, J. P.</p>
<p>This site describes the latest version, which has methodological and performance improvements, including support for whole-genome and targeted gene sequencing data. A user guide for v0.1.0 is available <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/user_guide/cancereffectsizeR_user_guide.md" class="external-link">here</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="provided-by">
<h2 data-toc-skip>Provided by</h2>
<ul class="list-unstyled">
<li><p>the <a href="https://medicine.yale.edu/lab/townsend/" class="external-link">Townsend Lab</a> at the<br><a href="https://ysph.yale.edu" class="external-link">Yale School of Public Health</a>.</p></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Vincent L. Cannataro <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6364-7747" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jeff Mandell <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-3839-2543" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
