<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Tutorial • cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.10.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a></li>
    <li><a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a></li>
    <li><a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a></li>
    <li><a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/vignettes/cancereffectsizeR.Rmd" class="external-link"><code>vignettes/cancereffectsizeR.Rmd</code></a></small>
      <div class="d-none name"><code>cancereffectsizeR.Rmd</code></div>
    </div>

    
    
<p>This tutorial applies some of the package’s key features to analyze
publicly-available somatic variant data from tumor sequencing. We
recommend that you start a fresh R/RStudio session. You can follow along
by copy-and-pasting code into an R script.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Start by installing the latest release:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Some dependencies are large, so we increase the download time limit to be safe</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"remotes"</span>, <span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Townsend-Lab-Yale/cancereffectsizeR"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span>, repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Regrettably, some Bioconductor releases contain versions of required
packages with bugs that cause cryptic messages in cancereffectsizeR
workflows, such as <code>!anyNA(m32) is not TRUE</code> or
<code>Error in data.frame(…, check.names=F): arguments imply differing number of rows</code>.
If you encounter these problems, the simplest solution is to upgrade
your Bioconductor version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only necessary if GenomeInfoDb or MutationalPatterns errors are causing problems.</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.20"</span><span class="op">)</span> <span class="co"># or newer, when available</span></span></code></pre></div>
<p>In addition to installing cancereffectsizeR, you need to install a
reference data set, or <em>refset</em>. These refsets include genomic
and gene annotations, mutational signature definitions, and more.
Currently, refsets supporting the hg38 and hg19 builds of the human
genome are available as separate data packages. (If you want to build
your own refset to use a custom genome build or gene set for almost any
species, <a href="custom_refset_instructions.html">you can</a>.) For
this tutorial, we’ll use the hg38 refset package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Townsend-Lab-Yale/ces.refset.hg38@*release"</span><span class="op">)</span></span></code></pre></div>
<p>Restart R/RStudio after installation. Before continuing, you may want
to create a directory for this tutorial to store all the data and output
in one place.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># setwd() somewhere, if desired</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"CES_tutorial"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"CES_tutorial"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quickstart">Quickstart<a class="anchor" aria-label="anchor" href="#quickstart"></a>
</h2>
<div class="section level3">
<h3 id="theoretical-overview">Theoretical overview<a class="anchor" aria-label="anchor" href="#theoretical-overview"></a>
</h3>
<p>Very briefly, we extract mutational signatures from each sample’s SNV
mutation profile using <a href="https://bioconductor.org/packages/release/bioc/html/MutationalPatterns.html" class="external-link">MutationalPatterns</a>.
The relative weights of biologically associated signatures are used to
infer trinucleotide-context-specific relative rates of SNV mutations for
each sample. Cohort-wide neutral gene mutation rates are calculated with
<a href="https://github.com/im3sanger/dndscv" class="external-link">dNdScv</a>, with
tissue-specific covariates provided by <code>ces.refset.hg38</code>.
Combining this information, the rate of neutral mutation at a particular
variant site is calculated by partitioning the gene mutation rate across
all sites in the gene in accordance with the signature-informed relative
rates. Comparing rates of observed and expected mutation under a model
of somatic selection allows inference of scaled selection coefficients,
which we also call cancer effects.</p>
</div>
<div class="section level3">
<h3 id="example-lung-adenocarcinoma-data-from-tcga">Example: Lung adenocarcinoma data from TCGA<a class="anchor" aria-label="anchor" href="#example-lung-adenocarcinoma-data-from-tcga"></a>
</h3>
<p>Here is a straightforward cancereffectsizeR workflow: We load TCGA
LUAD data, calculate mutation rates, and quantify selection. We also
attribute selection to mutational signatures and find that mutations
attributed to tobacco-associated signatures have disproportionately
large selective effects.</p>
<details><summary>
(Show/hide quickstart)
</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://townsend-lab-yale.github.io/cancereffectsizeR">cancereffectsizeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download TCGA lung adenocarcinoma (LUAD) somatic variant data.</span></span>
<span><span class="va">tcga_maf_file</span> <span class="op">&lt;-</span> <span class="st">"TCGA-LUAD.maf.gz"</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">tcga_maf_file</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/get_TCGA_project_MAF.html">get_TCGA_project_MAF</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"LUAD"</span>, filename <span class="op">=</span> <span class="st">"TCGA-LUAD.maf.gz"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Prepare data</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">tcga_maf_file</span>, refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create cancereffectsizeR analysis and load data</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, maf <span class="op">=</span> <span class="va">maf</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Infer trinculeotide-context-specific relative rates of SNV mutation from</span></span>
<span><span class="co"># a mutational signature analysis (leaving out signatures not found in LUAD)</span></span>
<span><span class="va">signature_exclusions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/suggest_cosmic_signature_exclusions.html">suggest_cosmic_signature_exclusions</a></span><span class="op">(</span>cancer_type <span class="op">=</span> <span class="st">"LUAD"</span>, treatment_naive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates</a></span><span class="op">(</span></span>
<span>  cesa <span class="op">=</span> <span class="va">cesa</span>, signature_set <span class="op">=</span> <span class="va">ces.refset.hg38</span><span class="op">$</span><span class="va">signatures</span><span class="op">$</span><span class="va">COSMIC_v3.4</span>,</span>
<span>  signature_exclusions <span class="op">=</span> <span class="va">signature_exclusions</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate neutral gene mutation rates using dNdScv, with tissue-specific mutation rate covariates.</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_mutation_rates.html">gene_mutation_rates</a></span><span class="op">(</span><span class="va">cesa</span>, covariates <span class="op">=</span> <span class="va">ces.refset.hg38</span><span class="op">$</span><span class="va">covariates</span><span class="op">$</span><span class="va">lung</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Infer scaled selection coefficients under the default model of clonal selection.</span></span>
<span><span class="co"># By default, inference is restricted to recurrent mutations.</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ces_variant.html">ces_variant</a></span><span class="op">(</span><span class="va">cesa</span>, run_name <span class="op">=</span> <span class="st">"example"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize top-effect variants.</span></span>
<span><span class="fu"><a href="../reference/plot_effects.html">plot_effects</a></span><span class="op">(</span>effects <span class="op">=</span> <span class="va">cesa</span><span class="op">$</span><span class="va">selection</span><span class="op">$</span><span class="va">example</span>, color_by <span class="op">=</span> <span class="st">"#DB382D"</span>, topn <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Attribute effects to mutational signatures</span></span>
<span><span class="va">mut_effects</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mutational_signature_effects.html">mutational_signature_effects</a></span><span class="op">(</span><span class="va">cesa</span>, <span class="va">cesa</span><span class="op">$</span><span class="va">selection</span><span class="op">$</span><span class="va">example</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot a comparison of how signatures contribute to mutation vs. selection</span></span>
<span><span class="fu"><a href="../reference/plot_signature_effects.html">plot_signature_effects</a></span><span class="op">(</span><span class="va">mut_effects</span>, viridis_option <span class="op">=</span> <span class="st">"F"</span>, num_sig_groups <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See the full tutorial for more details and a broader view of functionality!</span></span></code></pre></div>
<p><img src="cancereffectsizeR_files/figure-html/unnamed-chunk-6-1.png" width="700"><img src="cancereffectsizeR_files/figure-html/unnamed-chunk-6-2.png" width="700"></p>
</details>
</div>
</div>
<div class="section level2">
<h2 id="preparing-data">Preparing data<a class="anchor" aria-label="anchor" href="#preparing-data"></a>
</h2>
<p>Load cancereffectsizeR, as well as the data.table package.
cancereffectsizeR makes extensive use of data tables, so it’s handy to
have the package loaded.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://townsend-lab-yale.github.io/cancereffectsizeR">cancereffectsizeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span></code></pre></div>
<p>For this tutorial, we’ll use somatic variant data produced from exome
sequencing by the TCGA BRCA (breast carcinoma) project. We will
supplement the WXS data with targeted sequencing data from the
Metastatic Breast Cancer data set <a href="https://www.cbioportal.org/study/summary?id=breast_ink4_msk_2021" class="external-link">hosted
at cBioPortal</a>.</p>
<p>In cancereffectsizeR, data can be combined from multiple whole-exome,
whole-genome, and targeted sequencing sources, although we always need
at least one WXS or WGS source to anchor the analysis. Targeted
sequencing data can’t be used for mutation rate inference: there are too
few mutations, typically, and since they’re mostly in cancer hotspots,
they don’t provide a baseline of what mutation rates are like in the
absence of selection. Instead, mutation rates in TGS samples will be
assumed to be similar to those calculated in WXS/WGS samples.</p>
<div class="section level3">
<h3 id="tcga-data-whole-exome">TCGA data (whole-exome)<a class="anchor" aria-label="anchor" href="#tcga-data-whole-exome"></a>
</h3>
<p>We’ll begin by downloading variant data from TCGA. The function
downloads patient MAF data from the latest TCGA data release and
assembles a project-level MAF.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcga_maf_file</span> <span class="op">&lt;-</span> <span class="st">"TCGA-BRCA.maf.gz"</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">tcga_maf_file</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/get_TCGA_project_MAF.html">get_TCGA_project_MAF</a></span><span class="op">(</span>project <span class="op">=</span> <span class="st">"BRCA"</span>, filename <span class="op">=</span> <span class="va">tcga_maf_file</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Let’s also load and examine a table of patient information. Although
the TCGA samples in the data set are all taken from primary tumors,
you’ll see that some patients presented with metastatic disease (pM =
M1). Hormone receptor status (progesterone or estrogen receptor
positive) and HER2 amplification status are also recorded in a combined
column.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcga_clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/data.table/man/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"tutorial/TCGA_BRCA_clinical.txt"</span>, package <span class="op">=</span> <span class="st">"cancereffectsizeR"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Change patient identifier column name in clinical table to match the MAF.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/data.table/man/setattr.html" class="external-link">setnames</a></span><span class="op">(</span><span class="va">tcga_clinical</span>, <span class="st">"patient_id"</span>, <span class="st">"Unique_Patient_Identifier"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Peek at data</span></span>
<span><span class="va">tcga_clinical</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##    Unique_Patient_Identifier     pM receptor_status</span></span>
<span><span class="co">##                       &lt;char&gt; &lt;char&gt;          &lt;char&gt;</span></span>
<span><span class="co">## 1:              TCGA-3C-AAAU   &lt;NA&gt;       HR+/HER2-</span></span>
<span><span class="co">## 2:              TCGA-3C-AALI     M0       HR+/HER2+</span></span>
<span><span class="co">## 3:              TCGA-3C-AALJ     M0            &lt;NA&gt;</span></span>
<span><span class="co">## 4:              TCGA-3C-AALK     M0       HR+/HER2+</span></span>
<span><span class="co">## 5:              TCGA-4H-AAAK     M0            &lt;NA&gt;</span></span></code></pre>
<p>The <code><a href="../reference/preload_maf.html">preload_maf()</a></code> function takes in MAF data, extracts
the columns needed by cancereffectsizeR, adds a couple of genomic
annotations, and checks for common problems. (For your own analyses, see
our <a href="MAF_filtering_tips.html">MAF data tips</a>.) This MAF file
already uses the hg38 genome build, but if it didn’t, we could use the
<code>chain_file</code> argument to convert records via liftOver.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcga_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">tcga_maf_file</span>, refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span></span></code></pre></div>
<p>When MAFs created by are fed into , TCGA sample replicates are
effectively merged, since the Unique_Patient_Identifier column
supersedes Tumor_Sample_Barcode (the original sample identifiers). You
will see a note that preload_maf() caught and handled the ensuing
duplicate mutation records.</p>
</div>
<div class="section level3">
<h3 id="metastatic-data-tgs">Metastatic data (TGS)<a class="anchor" aria-label="anchor" href="#metastatic-data-tgs"></a>
</h3>
<p>Our TGS data source (hosted on cBioPortal and <a href="https://cancerdiscovery.aacrjournals.org/content/candisc/early/2021/09/17/2159-8290.CD-20-1726.full.pdf" class="external-link">recently
published</a> in Cancer Discovery), consists of metastatic tumors that
were sequenced using various MSK-Impact panels. In the full data set,
some patients have multiple samples sequenced; the data provided with
the package has been subsetted to one sample per patient, and the genome
build has been converted to hg38.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tgs_maf_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"tutorial/metastatic_breast_2021_hg38.maf"</span>, package <span class="op">=</span> <span class="st">"cancereffectsizeR"</span><span class="op">)</span></span>
<span><span class="va">tgs_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">tgs_maf_file</span>, refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span></span></code></pre></div>
<p>When combining data from multiple sources,
<code><a href="../reference/check_sample_overlap.html">check_sample_overlap()</a></code> can detect unexpected sample
duplication. If you feed in <code>tcga_maf</code> and
<code>tgs_maf</code> and filter the output to
<code>variants_shared &gt; 2</code> (with panel data, a couple of shared
variants doesn’t imply sample duplication), you’ll see that everything
looks fine.</p>
</div>
</div>
<div class="section level2">
<h2 id="create-cesanalysis-and-load-data">Create CESAnalysis and load data<a class="anchor" aria-label="anchor" href="#create-cesanalysis-and-load-data"></a>
</h2>
<p>The CESAnalysis is the primary data structure of cancereffectsizeR.
The cancereffectsizeR workflow consists of calling a series of functions
that take a CESAnalysis as input and return an altered CESAnalysis.</p>
<div class="section level3">
<h3 id="load-whole-exome-sequencing-data">Load whole-exome sequencing data<a class="anchor" aria-label="anchor" href="#load-whole-exome-sequencing-data"></a>
</h3>
<p>Let’s create a CESAnalysis and call <code><a href="../reference/load_maf.html">load_maf()</a></code> to load
the TCGA data. Since we’re going to load more than one MAF, we’ll
provide an optional <code>maf_name</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, maf <span class="op">=</span> <span class="va">tcga_maf</span>, maf_name <span class="op">=</span> <span class="st">"BRCA"</span><span class="op">)</span></span></code></pre></div>
<p>You will see a message that some variants fall outside of the
refset’s exome definitions. Since we don’t know exactly what exome
capture techniques were used for the BRCA project (probably different
methods at different study sites), this isn’t unexpected. When you do
know the exact exome capture intervals of whatever data you’re using,
you should supply those intervals with <code><a href="../reference/load_maf.html">load_maf()</a></code>’s
<code>covered_regions</code> argument. For WGS data, set
<code>coverage = "genome"</code>.</p>
<p>To navigate a CESAnalysis, use the dollar sign ($). We can use these
accessors to look at the MAF data as well as more detailed variant
annotations, and if we want we can perform various filtering operations,
such as identifying the most prevalent variants in the loaded data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cesa</span><span class="op">$</span><span class="va">maf</span></span>
<span><span class="va">cesa</span><span class="op">$</span><span class="va">variants</span></span>
<span><span class="va">cesa</span><span class="op">$</span><span class="va">samples</span></span>
<span></span>
<span><span class="co"># Let's see the top variants</span></span>
<span><span class="op">(</span><span class="va">top_variants</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">maf_prevalence</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu">.</span><span class="op">(</span><span class="va">variant_name</span>, <span class="va">chr</span>, <span class="va">start</span>, <span class="va">end</span>, <span class="va">maf_prevalence</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Let’s load the clinical data into the analysis, too.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_sample_data.html">load_sample_data</a></span><span class="op">(</span><span class="va">cesa</span>, <span class="va">tcga_clinical</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-targeted-gene-sequencing-data">Load targeted gene sequencing data<a class="anchor" aria-label="anchor" href="#load-targeted-gene-sequencing-data"></a>
</h3>
<p>To load our TGS data, we need to be able to define its coverage. Why?
To estimate the selection of a mutation, we need to know which samples
have the mutation, which samples do not, and which are unknown due to
lack of sequencing coverage. Our TGS samples were sequenced with
multiple panels covering different genes, and unfortunately, the exact
coverage (defined by genomic coordinates) is not publicly available for
all of these. Therefore, for simplicity, we will filter the TGS data to
mutations at a handful of top cancer genes that are covered in all of
the panels. When we load the data, all records outside of these genes
will be excluded, leaving us with greater power to infer selection in
these genes without impacting our estimates outside these genes, which
will use just the TCGA data. (As mentioned earlier, we shouldn’t filter
WXS/WGS data this way, since it would interfere with mutation rate
calculation, but TGS samples are not involved in mutation rate
calculation.)</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define coverage using the coding regions of these genes, as defined by the refset</span></span>
<span><span class="va">top_tgs_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"TP53"</span>, <span class="st">"PIK3CA"</span>, <span class="st">"ESR1"</span>, <span class="st">"CDH1"</span>, <span class="st">"GATA3"</span>, <span class="st">"KMT2C"</span>,</span>
<span>  <span class="st">"MAP3K1"</span>, <span class="st">"AKT1"</span>, <span class="st">"ARID1A"</span>, <span class="st">"FOXA1"</span>, <span class="st">"TBX3"</span>, <span class="st">"PTEN"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">tgs_coverage</span> <span class="op">&lt;-</span> <span class="va">ces.refset.hg38</span><span class="op">$</span><span class="va">gr_genes</span><span class="op">[</span><span class="va">ces.refset.hg38</span><span class="op">$</span><span class="va">gr_genes</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">top_tgs_genes</span><span class="op">]</span></span></code></pre></div>
<p>We don’t have a clinical file for the TGS data, but we do know that
they’re all metastatic samples. Instead of creating another table and
calling <code><a href="../reference/load_sample_data.html">load_sample_data()</a></code>, we can use the
<code>sample_data_cols</code> argument in load_maf(). We will also add a
little bit of padding (10 bp) to our coverage definitions to allow
variants just outside coding regions to be considered covered.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tgs_maf</span><span class="op">$</span><span class="va">pM</span> <span class="op">&lt;-</span> <span class="st">"M1"</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span><span class="va">cesa</span>,</span>
<span>  maf <span class="op">=</span> <span class="va">tgs_maf</span>, sample_data_cols <span class="op">=</span> <span class="st">"pM"</span>, maf_name <span class="op">=</span> <span class="st">"MBC"</span>,</span>
<span>  coverage <span class="op">=</span> <span class="st">"targeted"</span>, covered_regions <span class="op">=</span> <span class="va">tgs_coverage</span>,</span>
<span>  covered_regions_name <span class="op">=</span> <span class="st">"top_genes"</span>, covered_regions_padding <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="mutational-processes-and-relative-mutation-rates">Mutational processes and relative mutation rates<a class="anchor" aria-label="anchor" href="#mutational-processes-and-relative-mutation-rates"></a>
</h2>
<p>For each (exome) sample in our data set,
<code><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code> will perform mutational signature
extraction in order to to attribute each sample’s set of SNVs to a
linear combination of mutational processes. By default, the signature
extraction is done using the <code>MutationalPatterns</code> package,
and <code>deconstructSigs</code> is also supported. The signature
attribution allows us to infer sample-specific relative rates of SNV
mutation for all trinucleotide contexts. TGS samples will be assumed to
have mutational processes matching the group-average mutational
processes of the exome data. (This isn’t optimal, since the metastatic
samples may be affected by mutational processes, such as chemotherapy,
that are lacking in the primary tumor exome data, but it might be the
best we can do without finding another data source.)</p>
<p>We will use signature definitions from the refset (it’s also possible
to create your own). To improve the accuracy of signature extraction, we
will exclude signatures that can safely be presumed absent from the
samples; a helper function,
<code><a href="../reference/suggest_cosmic_signature_exclusions.html">suggest_cosmic_signature_exclusions()</a></code>, can provide some
guidance.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We'll use all suggested exclusions (TCGA primary tumors are treatment-naive)</span></span>
<span><span class="va">signature_exclusions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/suggest_cosmic_signature_exclusions.html">suggest_cosmic_signature_exclusions</a></span><span class="op">(</span>cancer_type <span class="op">=</span> <span class="st">"BRCA"</span>, treatment_naive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates</a></span><span class="op">(</span><span class="va">cesa</span>,</span>
<span>  signature_set <span class="op">=</span> <span class="va">ces.refset.hg38</span><span class="op">$</span><span class="va">signatures</span><span class="op">$</span><span class="va">COSMIC_v3.4</span>,</span>
<span>  signature_exclusions <span class="op">=</span> <span class="va">signature_exclusions</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code> run has added some useful
information to the CESAnalysis:</p>
<ul>
<li>
<strong>snv_counts</strong>: A matrix of SNV counts by trinucleotide
context. (You can also generate such a matrix from MAF data with
<code><a href="../reference/trinuc_snv_counts.html">trinuc_snv_counts()</a></code>.)</li>
<li>
<strong>raw_attributions</strong>: Signature attributions as
generated by the extractor. For MutationalPatterns, this matrix contains
the number of mutations attributed to signature for each sample.
Naturally, samples with more mutations will tend to have higher raw
attributions.</li>
<li>
<strong>biological_weights</strong>: The proportion of mutations
attributed to each biologically-associated signature within each sample.
Mutations attributed to signatures associated with sequencing/processing
artifacts are left out. (Technical detail: Due to the instability of
signature attributions on samples with few mutations, Samples with few
MAF variants have their weights adjusted towards group-average weights,
as indicated in the <code>group_avg_blended</code> column. If you want
to make claims about subgroup differences in mutational processes,
consider leaving these samples out or using the raw attributions. You
should also leave out the TGS samples, which in this analysis are
treated as having 0 mutations.)</li>
<li>
<strong>trinuc_rates</strong>: Inferred relative rates of mutation,
produced by matrix-multiplying biological_weights and signature
definitions. (These rates will not equal empirical rates of observed
mutations.)</li>
</ul>
</div>
<div class="section level2">
<h2 id="gene-mutation-rates">Gene mutation rates<a class="anchor" aria-label="anchor" href="#gene-mutation-rates"></a>
</h2>
<p>Next, we’ll use <code><a href="../reference/gene_mutation_rates.html">gene_mutation_rates()</a></code> to estimate
regional rates of mutation in the absence of selection. We’ll use the
method provided in the dNdScv package, which uses dN/dS ratios and
mutation rate covariates. Our refset has pre-computed covariates for a
variety of tissue types. As with mutational processes, this analysis
will not use the TGS samples, but the calculated rates will be assumed
to hold for them.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_mutation_rates.html">gene_mutation_rates</a></span><span class="op">(</span><span class="va">cesa</span>, covariates <span class="op">=</span> <span class="va">ces.refset.hg38</span><span class="op">$</span><span class="va">covariates</span><span class="op">$</span><span class="va">breast</span><span class="op">)</span></span></code></pre></div>
<p>Here are the neutral gene mutation rates extracted from dNdScv’s
regression:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cesa</span><span class="op">$</span><span class="va">gene_rates</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       gene         rate rate_grp_1_95_low rate_grp_1_95_high</span></span>
<span><span class="co">##     &lt;char&gt;        &lt;num&gt;             &lt;num&gt;              &lt;num&gt;</span></span>
<span><span class="co">## 1:    A1BG 6.127126e-07      5.613872e-07       6.677908e-07</span></span>
<span><span class="co">## 2:    A1CF 7.084595e-07      6.658288e-07       7.532645e-07</span></span>
<span><span class="co">## 3:     A2M 7.896587e-07      7.532342e-07       8.269888e-07</span></span>
<span><span class="co">## 4:   A2ML1 6.255639e-07      5.869469e-07       6.656543e-07</span></span>
<span><span class="co">## 5: A3GALT2 5.508752e-07      5.100439e-07       5.946168e-07</span></span>
<span><span class="co">## 6:  A4GALT 7.261875e-07      6.757761e-07       7.797770e-07</span></span></code></pre>
<p>We can also look at dNdScv’s identification of selection at the gene
level. We’ll filter results to q &lt; .05.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View top dNdScv genes, sorted by significance</span></span>
<span><span class="va">dndscv_results</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">dNdScv_results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sig_genes</span> <span class="op">&lt;-</span> <span class="va">dndscv_results</span><span class="op">[</span><span class="va">qallsubs_cv</span> <span class="op">&lt;</span> <span class="fl">.05</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">qallsubs_cv</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<table class="table table table-condensed">
<thead><tr>
<th style="text-align:right;">
gene_name
</th>
<th style="text-align:right;">
n_syn
</th>
<th style="text-align:right;">
n_mis
</th>
<th style="text-align:right;">
n_non
</th>
<th style="text-align:right;">
n_spl
</th>
<th style="text-align:right;">
wmis_cv
</th>
<th style="text-align:right;">
wnon_cv
</th>
<th style="text-align:right;">
wspl_cv
</th>
<th style="text-align:right;">
pmis_cv
</th>
<th style="text-align:right;">
ptrunc_cv
</th>
<th style="text-align:right;">
pallsubs_cv
</th>
<th style="text-align:right;">
qmis_cv
</th>
<th style="text-align:right;">
qtrunc_cv
</th>
<th style="text-align:right;">
qallsubs_cv
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
TP53
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
198
</td>
<td style="text-align:right;">
46
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
188.063121
</td>
<td style="text-align:right;">
580.81334
</td>
<td style="text-align:right;">
580.81334
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
</tr>
<tr>
<td style="text-align:right;">
PIK3CA
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
336
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
49.946758
</td>
<td style="text-align:right;">
0.00000
</td>
<td style="text-align:right;">
0.00000
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
1.741979e-01
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
8.485461e-01
</td>
<td style="text-align:right;">
0.000000e+00
</td>
</tr>
<tr>
<td style="text-align:right;">
PTEN
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
15
</td>
<td style="text-align:right;">
11
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
9.088199
</td>
<td style="text-align:right;">
91.77341
</td>
<td style="text-align:right;">
91.77341
</td>
<td style="text-align:right;">
5.238135e-06
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
1.299282e-02
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
</tr>
<tr>
<td style="text-align:right;">
CDH1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
16
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
4.475685
</td>
<td style="text-align:right;">
141.11002
</td>
<td style="text-align:right;">
141.11002
</td>
<td style="text-align:right;">
4.219075e-04
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
5.232557e-01
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
</tr>
<tr>
<td style="text-align:right;">
KMT2C
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
25
</td>
<td style="text-align:right;">
27
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1.532738
</td>
<td style="text-align:right;">
14.61201
</td>
<td style="text-align:right;">
14.61201
</td>
<td style="text-align:right;">
2.523294e-01
</td>
<td style="text-align:right;">
3.663736e-15
</td>
<td style="text-align:right;">
2.220446e-16
</td>
<td style="text-align:right;">
7.803695e-01
</td>
<td style="text-align:right;">
1.188736e-11
</td>
<td style="text-align:right;">
7.710721e-13
</td>
</tr>
<tr>
<td style="text-align:right;">
MAP2K4
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
10.810199
</td>
<td style="text-align:right;">
76.15288
</td>
<td style="text-align:right;">
76.15288
</td>
<td style="text-align:right;">
2.433399e-06
</td>
<td style="text-align:right;">
4.107825e-15
</td>
<td style="text-align:right;">
8.881784e-16
</td>
<td style="text-align:right;">
7.041852e-03
</td>
<td style="text-align:right;">
1.188736e-11
</td>
<td style="text-align:right;">
2.570240e-12
</td>
</tr>
<tr>
<td style="text-align:right;">
MAP3K1
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
17
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2.988223
</td>
<td style="text-align:right;">
24.31043
</td>
<td style="text-align:right;">
24.31043
</td>
<td style="text-align:right;">
3.202730e-03
</td>
<td style="text-align:right;">
2.331468e-15
</td>
<td style="text-align:right;">
2.253753e-14
</td>
<td style="text-align:right;">
7.803695e-01
</td>
<td style="text-align:right;">
1.012032e-11
</td>
<td style="text-align:right;">
5.590273e-11
</td>
</tr>
<tr>
<td style="text-align:right;">
AKT1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
26
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
17.900089
</td>
<td style="text-align:right;">
0.00000
</td>
<td style="text-align:right;">
0.00000
</td>
<td style="text-align:right;">
2.093881e-13
</td>
<td style="text-align:right;">
5.882058e-01
</td>
<td style="text-align:right;">
9.905410e-13
</td>
<td style="text-align:right;">
1.211868e-09
</td>
<td style="text-align:right;">
8.485461e-01
</td>
<td style="text-align:right;">
2.149845e-09
</td>
</tr>
<tr>
<td style="text-align:right;">
ARID1A
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
12
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2.100089
</td>
<td style="text-align:right;">
28.62643
</td>
<td style="text-align:right;">
28.62643
</td>
<td style="text-align:right;">
1.212432e-01
</td>
<td style="text-align:right;">
5.273559e-13
</td>
<td style="text-align:right;">
2.593925e-12
</td>
<td style="text-align:right;">
7.803695e-01
</td>
<td style="text-align:right;">
1.308069e-09
</td>
<td style="text-align:right;">
5.004258e-09
</td>
</tr>
<tr>
<td style="text-align:right;">
NCOR1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
2.637482
</td>
<td style="text-align:right;">
18.67881
</td>
<td style="text-align:right;">
18.67881
</td>
<td style="text-align:right;">
2.289725e-02
</td>
<td style="text-align:right;">
8.482326e-12
</td>
<td style="text-align:right;">
5.398637e-11
</td>
<td style="text-align:right;">
7.803695e-01
</td>
<td style="text-align:right;">
1.840983e-08
</td>
<td style="text-align:right;">
9.373654e-08
</td>
</tr>
</tbody>
</table>
<p>See dNdScv’s documentation if you’re interested in interpretation of
the dNdScv output.</p>
<p>Side note: If we had multiple tissue types in the analysis, we could
call gene_mutation_rates() multiple times using the <code>samples</code>
argument to specify sample groups, with appropriate covariates for each
tissue type. Similar functionality is available in
<code><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="selection-inference">Selection inference<a class="anchor" aria-label="anchor" href="#selection-inference"></a>
</h2>
<p>The combination of cohort-level gene mutation rates and sample-level
relative rates of trinucleotide-context-specific substitution allow us
to estimate the rate at which any somatic substitution occurs in any
patient’s tumor sample. To be clear, by “rates,” we don’t mean the
frequency of mutations in our somatic variant calls (which we already
know anyway). These rates represent how often the mutations can be
expected to occur in individual cells. (More specifically, we assume
that mutation events in each specific sample and site follow a Poisson
distribution, and these rates are the Poisson rates.) It’s not typically
necessary to look at these rates, but just to show that we can:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's take the top 3 variants by MAF prevalence. These happen to all be in PIK3CA.</span></span>
<span><span class="va">variants_to_check</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">maf_prevalence</span><span class="op">)</span>, <span class="va">variant_id</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># A few random samples</span></span>
<span><span class="va">samples_to_check</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TCGA-A2-A3Y0"</span>, <span class="st">"TCGA-XX-A89A"</span>, <span class="st">"P-0000224"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/baseline_mutation_rates.html">baseline_mutation_rates</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, variant_ids <span class="op">=</span> <span class="va">variants_to_check</span>, samples <span class="op">=</span> <span class="va">samples_to_check</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-condensed">
<thead><tr>
<th style="text-align:right;">
Unique_Patient_Identifier
</th>
<th style="text-align:right;">
PIK3CA_E542K_ENSP00000263967.3
</th>
<th style="text-align:right;">
PIK3CA_E545K_ENSP00000263967.3
</th>
<th style="text-align:right;">
PIK3CA_H1047R_ENSP00000263967.3
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
P-0000224
</td>
<td style="text-align:right;">
9.15e-06
</td>
<td style="text-align:right;">
9.15e-06
</td>
<td style="text-align:right;">
1.11e-06
</td>
</tr>
<tr>
<td style="text-align:right;">
TCGA-A2-A3Y0
</td>
<td style="text-align:right;">
1.84e-06
</td>
<td style="text-align:right;">
1.84e-06
</td>
<td style="text-align:right;">
8.73e-07
</td>
</tr>
<tr>
<td style="text-align:right;">
TCGA-XX-A89A
</td>
<td style="text-align:right;">
1.32e-05
</td>
<td style="text-align:right;">
1.32e-05
</td>
<td style="text-align:right;">
2.23e-08
</td>
</tr>
</tbody>
</table>
<p>Interestingly, PIK3CA H1047R, the most prevalent variant in our MAF
data (246/1875 samples), is estimated to have a substantially lower rate
of occurrence than E545K and E542K in these particular samples.</p>
<div class="section level3">
<h3 id="default-model">Default model<a class="anchor" aria-label="anchor" href="#default-model"></a>
</h3>
<p>We can use our estimated rates and the MAF variant data to infer
scaled selection coefficients (aka cancer effects, or selection
intensity) under a model of selection. These cancer effects are directly
proportional to the proliferative advantage provided by each variant.
First, let’s calculate selection under the package’s default model of
selection with ces_variant():</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Including an optional run_name</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ces_variant.html">ces_variant</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, run_name <span class="op">=</span> <span class="st">"recurrents"</span><span class="op">)</span></span></code></pre></div>
<p>Let’s visualize the most selected variants with
<code><a href="../reference/plot_effects.html">plot_effects()</a></code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_effects.html">plot_effects</a></span><span class="op">(</span>effects <span class="op">=</span> <span class="va">cesa</span><span class="op">$</span><span class="va">selection</span><span class="op">$</span><span class="va">recurrents</span><span class="op">)</span></span></code></pre></div>
<p><img src="cancereffectsizeR_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p><em>PIK3CA</em> and <em>ESR1</em> variants take many of the top
spots. The deconvolution of mutation rate from selection has revealed
that some relatively low-prevalence variants have high effects.
Interestingly, according to dNdScv, mutation in <em>ESR1</em> at the
gene level was not significantly greater than neutral in the TCGA cohort
of primary tumors. The incorporation of metastatic TGS data and
cancereffectsizeR’s assessment of selection at variant-level resolution
confirm <em>ESR1</em>’s importance.</p>
<p>Here’s a look at the effects of all recurrent variants across some of
the highest-effect genes.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_effects.html">plot_effects</a></span><span class="op">(</span><span class="va">cesa</span><span class="op">$</span><span class="va">selection</span><span class="op">$</span><span class="va">recurrents</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"gene"</span>, topn <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  label_individual_variants <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="cancereffectsizeR_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="more-options">More options<a class="anchor" aria-label="anchor" href="#more-options"></a>
</h3>
<p>By default, all recurrent variants (variants that appear at least
twice in MAF data) are included in a <code><a href="../reference/ces_variant.html">ces_variant()</a></code> run. As
demonstrated, the <code>variants</code> argument can be used to specify
which variants to test. Variants with prevalences of 1 or even 0
(perhaps helpful to establish an effect ceiling) can be included:
<code>variants = cesa$variants</code> will include all annotated
mutations in the analysis. Note that most single-hit variants will have
over-estimated effects because most are probably neutral passengers, and
the occurrence of any non-selected variant is inherently improbable.</p>
<p>For each variant site, only samples with coverage at the site will
inform selection inference. As also shown above, the
<code>samples</code> argument further limits the inference process to a
subset of eligible samples. This could be helpful to assess and compare
selection in various groups (e.g., smoking vs. non-smoking).</p>
<p>Besides the default model, <code><a href="../reference/ces_variant.html">ces_variant()</a></code> supports
user-supplied models of selection; see documentation for details. We
hope that some enterprising users will come up with new models, and we
are happy to discuss ideas and help work out any kinks.</p>
<p>The <code>CompoundVariantSet</code> feature, described in the next
section, may sometimes provide improved resolution of somatic selection
by batching related variants together.</p>
</div>
</div>
<div class="section level2">
<h2 id="epistatic-cancer-effects">Epistatic cancer effects<a class="anchor" aria-label="anchor" href="#epistatic-cancer-effects"></a>
</h2>
<p>In the default model, a variant is assumed to have a single cancer
effect across all samples. In reality, we expect a variant’s selection
to be influenced by a complex combination of factors including a
sample’s mutational background (from substitutions to structural
variants), the epigenetic state of the tumor, and various environmental
factors. The <code><a href="../reference/ces_epistasis.html">ces_epistasis()</a></code> function allows us to assess
selection for any pair of variants under a model of pairwise epistasis,
in which the selection intensity for each variant is allowed to very
depending on the state of the other site.</p>
<div class="section level3">
<h3 id="variant-level-epistasis">Variant-level epistasis<a class="anchor" aria-label="anchor" href="#variant-level-epistasis"></a>
</h3>
<p>Let’s try two prevalent BRCA mutations, PIK3CA E545K (n=159) and AKT1
E17K (n=82). Let’s also test E545K with a nearby PIK3CA variant, E542K
(n=93).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Start by pulling full variant IDs (with protein identifier) from variants table</span></span>
<span><span class="va">group1</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PIK3CA E545K"</span>, <span class="st">"AKT1 E17K"</span><span class="op">)</span>, <span class="va">variant_id</span>, on <span class="op">=</span> <span class="st">"variant_name"</span><span class="op">]</span></span>
<span><span class="va">group2</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PIK3CA E545K"</span>, <span class="st">"PIK3CA E542K"</span><span class="op">)</span>, <span class="va">variant_id</span>, on <span class="op">=</span> <span class="st">"variant_name"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ces_epistasis.html">ces_epistasis</a></span><span class="op">(</span></span>
<span>  cesa <span class="op">=</span> <span class="va">cesa</span>, variants <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">group1</span>, <span class="va">group2</span><span class="op">)</span>,</span>
<span>  conf <span class="op">=</span> <span class="fl">.95</span>, run_name <span class="op">=</span> <span class="st">"variant_epistasis_example"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cesa</span><span class="op">$</span><span class="va">epistasis</span><span class="op">$</span><span class="va">variant_epistasis_example</span></span></code></pre></div>
<table class="table table table-condensed">
<thead><tr>
<th style="text-align:right;">
variant_A
</th>
<th style="text-align:right;">
variant_B
</th>
<th style="text-align:right;">
ces_A0
</th>
<th style="text-align:right;">
ces_B0
</th>
<th style="text-align:right;">
ces_A_on_B
</th>
<th style="text-align:right;">
ces_B_on_A
</th>
<th style="text-align:right;">
p_A_change
</th>
<th style="text-align:right;">
p_B_change
</th>
<th style="text-align:right;">
p_epistasis
</th>
<th style="text-align:right;">
expected_nAB_epistasis
</th>
<th style="text-align:right;">
expected_nAB_null
</th>
<th style="text-align:right;">
AB_epistatic_ratio
</th>
<th style="text-align:right;">
nA0
</th>
<th style="text-align:right;">
nB0
</th>
<th style="text-align:right;">
nAB
</th>
<th style="text-align:right;">
n00
</th>
<th style="text-align:right;">
n_total
</th>
<th style="text-align:right;">
ces_A_null
</th>
<th style="text-align:right;">
ces_B_null
</th>
<th style="text-align:right;">
ci_low_95_ces_A0
</th>
<th style="text-align:right;">
ci_high_95_ces_A0
</th>
<th style="text-align:right;">
ci_low_95_ces_B0
</th>
<th style="text-align:right;">
ci_high_95_ces_B0
</th>
<th style="text-align:right;">
ci_low_95_ces_A_on_B
</th>
<th style="text-align:right;">
ci_high_95_ces_A_on_B
</th>
<th style="text-align:right;">
ci_low_95_ces_B_on_A
</th>
<th style="text-align:right;">
ci_high_95_ces_B_on_A
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
PIK3CA_E545K_ENSP00000263967.3
</td>
<td style="text-align:right;">
AKT1_E17K_ENSP00000497822.1
</td>
<td style="text-align:right;">
10855
</td>
<td style="text-align:right;">
26453
</td>
<td style="text-align:right;">
0.001
</td>
<td style="text-align:right;">
6541.491
</td>
<td style="text-align:right;">
0.17701
</td>
<td style="text-align:right;">
0.20577
</td>
<td style="text-align:right;">
6.76e-03
</td>
<td style="text-align:right;">
1.02e+00
</td>
<td style="text-align:right;">
7.54
</td>
<td style="text-align:right;">
1.35e-01
</td>
<td style="text-align:right;">
159
</td>
<td style="text-align:right;">
80
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1614
</td>
<td style="text-align:right;">
1854
</td>
<td style="text-align:right;">
10572
</td>
<td style="text-align:right;">
25487
</td>
<td style="text-align:right;">
9257
</td>
<td style="text-align:right;">
12626
</td>
<td style="text-align:right;">
21071
</td>
<td style="text-align:right;">
32683
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
8735
</td>
<td style="text-align:right;">
373
</td>
<td style="text-align:right;">
28908
</td>
</tr>
<tr>
<td style="text-align:right;">
PIK3CA_E545K_ENSP00000263967.3
</td>
<td style="text-align:right;">
PIK3CA_E542K_ENSP00000263967.3
</td>
<td style="text-align:right;">
10913
</td>
<td style="text-align:right;">
6343
</td>
<td style="text-align:right;">
0.001
</td>
<td style="text-align:right;">
0.001
</td>
<td style="text-align:right;">
0.00732
</td>
<td style="text-align:right;">
0.00671
</td>
<td style="text-align:right;">
4.94e-05
</td>
<td style="text-align:right;">
1.31e-06
</td>
<td style="text-align:right;">
9.34
</td>
<td style="text-align:right;">
1.40e-07
</td>
<td style="text-align:right;">
160
</td>
<td style="text-align:right;">
93
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1601
</td>
<td style="text-align:right;">
1854
</td>
<td style="text-align:right;">
10572
</td>
<td style="text-align:right;">
6005
</td>
<td style="text-align:right;">
9307
</td>
<td style="text-align:right;">
12694
</td>
<td style="text-align:right;">
5139
</td>
<td style="text-align:right;">
7722
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
4171
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
2360
</td>
</tr>
</tbody>
</table>
<p>Confidence intervals tend to be wide in epistatic analyses. Even with
the most prevalent somatic variants, both co-occurrence and mutual
exclusivity can often be explained by chance. The output for PIK3CA
E545K and E542K confirms the negative epistatic relationship that we
would have expected given their mutual exclusivity in our data (and our
knowledge of their biological effects). For E545K and AKT1 E17K, we see
reduced selection for both after the acquisition of the other driver,
but the confidence intervals leave open the possibility that selection
for AKT1 E17K is unaffected by PIK3CA status. (Additionally, the NA’s on
some of the lower bounds indicate that they fall below the lower limit
of the optimization algorithm.)</p>
</div>
<div class="section level3">
<h3 id="epistasis-and-compoundvariantsets">Epistasis and CompoundVariantSets<a class="anchor" aria-label="anchor" href="#epistasis-and-compoundvariantsets"></a>
</h3>
<p>We can probe this PIK3CA/AKT1 relationship further if we assume that
all prevalent PIK3CA mutations share the same epistatic relationship
with AKT1 E17K. The <code><a href="../reference/define_compound_variants.html">define_compound_variants()</a></code> feature lets
us combine arbitrary variants into “compound variants” that are treated
as if they were single variants by cancereffectsizeR’s selection
inference functions. (In brief, the mutation rate of the compound
variant is equal to the sum of rates of constituent variants, and any
sample with one or more of the constituent variants “has the compound
variant.”) Below, we define a CompoundVariantSet with two compound
variants: All <em>PIK3CA</em> variants with MAF prevalence &gt; 1, and
AKT E17K. We then pass the CompoundVariantSet to ces_epistasis(), which
will test all pairs of compound variants (here, just the one pair).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Collect all the variants that we want in the CompoundVariantSet into a table</span></span>
<span><span class="va">top_PIK3CA</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"PIK3CA"</span> <span class="op">&amp;</span> <span class="va">maf_prevalence</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">top_akt1</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="va">variant_name</span> <span class="op">==</span> <span class="st">"AKT1 E17K"</span><span class="op">]</span></span>
<span><span class="va">for_compound</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">top_PIK3CA</span>, <span class="va">top_akt1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># see define_compound_variants() documentation for details on arguments</span></span>
<span><span class="va">comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/define_compound_variants.html">define_compound_variants</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, variant_table <span class="op">=</span> <span class="va">for_compound</span>, by <span class="op">=</span> <span class="st">"gene"</span>, merge_distance <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ces_epistasis.html">ces_epistasis</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, variants <span class="op">=</span> <span class="va">comp</span>, run_name <span class="op">=</span> <span class="st">"AKT1_E17K_vs_PIK3CA"</span><span class="op">)</span></span>
<span><span class="va">cesa</span><span class="op">$</span><span class="va">epistasis</span><span class="op">$</span><span class="va">AKT1_E17K_vs_PIK3CA</span></span></code></pre></div>
<table class="table table table-condensed">
<thead><tr>
<th style="text-align:right;">
variant_A
</th>
<th style="text-align:right;">
variant_B
</th>
<th style="text-align:right;">
ces_A0
</th>
<th style="text-align:right;">
ces_B0
</th>
<th style="text-align:right;">
ces_A_on_B
</th>
<th style="text-align:right;">
ces_B_on_A
</th>
<th style="text-align:right;">
p_A_change
</th>
<th style="text-align:right;">
p_B_change
</th>
<th style="text-align:right;">
p_epistasis
</th>
<th style="text-align:right;">
expected_nAB_epistasis
</th>
<th style="text-align:right;">
expected_nAB_null
</th>
<th style="text-align:right;">
AB_epistatic_ratio
</th>
<th style="text-align:right;">
nA0
</th>
<th style="text-align:right;">
nB0
</th>
<th style="text-align:right;">
nAB
</th>
<th style="text-align:right;">
n00
</th>
<th style="text-align:right;">
n_total
</th>
<th style="text-align:right;">
ces_A_null
</th>
<th style="text-align:right;">
ces_B_null
</th>
<th style="text-align:right;">
ci_low_95_ces_A0
</th>
<th style="text-align:right;">
ci_high_95_ces_A0
</th>
<th style="text-align:right;">
ci_low_95_ces_B0
</th>
<th style="text-align:right;">
ci_high_95_ces_B0
</th>
<th style="text-align:right;">
ci_low_95_ces_A_on_B
</th>
<th style="text-align:right;">
ci_high_95_ces_A_on_B
</th>
<th style="text-align:right;">
ci_low_95_ces_B_on_A
</th>
<th style="text-align:right;">
ci_high_95_ces_B_on_A
</th>
</tr></thead>
<tbody><tr>
<td style="text-align:right;">
PIK3CA
</td>
<td style="text-align:right;">
AKT1
</td>
<td style="text-align:right;">
3756
</td>
<td style="text-align:right;">
30457
</td>
<td style="text-align:right;">
6.95
</td>
<td style="text-align:right;">
6214
</td>
<td style="text-align:right;">
0.000182
</td>
<td style="text-align:right;">
0.000392
</td>
<td style="text-align:right;">
4.87e-11
</td>
<td style="text-align:right;">
3.98
</td>
<td style="text-align:right;">
29.1
</td>
<td style="text-align:right;">
0.137
</td>
<td style="text-align:right;">
642
</td>
<td style="text-align:right;">
77
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1131
</td>
<td style="text-align:right;">
1854
</td>
<td style="text-align:right;">
3636
</td>
<td style="text-align:right;">
25504
</td>
<td style="text-align:right;">
3471
</td>
<td style="text-align:right;">
4056
</td>
<td style="text-align:right;">
24153
</td>
<td style="text-align:right;">
37801
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
1014
</td>
<td style="text-align:right;">
1873
</td>
<td style="text-align:right;">
14410
</td>
</tr></tbody>
</table>
<p>Note that the effect sizes are smaller for PIK3CA here, since we
mixed in lots of lower-effect PIK3CA variants. While four samples carry
both AKT1 E17K and a PIK3CA variant, this result strengthens the case
for AKT1 mutation reducing selection for PIK3CA mutation.</p>
</div>
<div class="section level3">
<h3 id="gene-level-epistasis">Gene-level epistasis<a class="anchor" aria-label="anchor" href="#gene-level-epistasis"></a>
</h3>
<p>The convenience function <code><a href="../reference/ces_gene_epistasis.html">ces_gene_epistasis()</a></code> provides a
simpler way to apply the model of <code><a href="../reference/ces_epistasis.html">ces_epistasis()</a></code> at the
gene level. The <code>variants</code> argument provides three options
for which variants from each gene to include in the inference:
“recurrent” uses all recurrent variants; “nonsilent” uses nonsynonymous
coding variants and any variants in essential splice sites; or,
alternatively, supply a custom table of variants. In order to take
advantage of the targeted sequencing data (which doesn’t quite cover all
mutated sites in these genes), we will supply a custom table of variants
that are covered across all samples.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AKT1"</span>, <span class="st">"PIK3CA"</span>, <span class="st">"TP53"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get consensus covered regions</span></span>
<span><span class="va">combined_coverage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">cesa</span><span class="op">$</span><span class="va">coverage_ranges</span><span class="op">$</span><span class="va">exome</span><span class="op">$</span><span class="va">`exome+`</span>, <span class="va">cesa</span><span class="op">$</span><span class="va">coverage_ranges</span><span class="op">$</span><span class="va">targeted</span><span class="op">$</span><span class="va">top_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get variants in the genes of interest that have sequencing coverage in all samples</span></span>
<span><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_variants.html">select_variants</a></span><span class="op">(</span><span class="va">cesa</span>, genes <span class="op">=</span> <span class="va">genes</span>, gr <span class="op">=</span> <span class="va">combined_coverage</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ces_gene_epistasis.html">ces_gene_epistasis</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, genes <span class="op">=</span> <span class="va">genes</span>, variants <span class="op">=</span> <span class="va">variants</span>, run_name <span class="op">=</span> <span class="st">"gene_epistasis_example"</span><span class="op">)</span></span>
<span><span class="va">cesa</span><span class="op">$</span><span class="va">epistasis</span><span class="op">$</span><span class="va">gene_epistasis_example</span></span></code></pre></div>
<table class="table table table-condensed">
<thead><tr>
<th style="text-align:right;">
variant_A
</th>
<th style="text-align:right;">
variant_B
</th>
<th style="text-align:right;">
ces_A0
</th>
<th style="text-align:right;">
ces_B0
</th>
<th style="text-align:right;">
ces_A_on_B
</th>
<th style="text-align:right;">
ces_B_on_A
</th>
<th style="text-align:right;">
p_A_change
</th>
<th style="text-align:right;">
p_B_change
</th>
<th style="text-align:right;">
p_epistasis
</th>
<th style="text-align:right;">
expected_nAB_epistasis
</th>
<th style="text-align:right;">
expected_nAB_null
</th>
<th style="text-align:right;">
AB_epistatic_ratio
</th>
<th style="text-align:right;">
nA0
</th>
<th style="text-align:right;">
nB0
</th>
<th style="text-align:right;">
nAB
</th>
<th style="text-align:right;">
n00
</th>
<th style="text-align:right;">
n_total
</th>
<th style="text-align:right;">
ces_A_null
</th>
<th style="text-align:right;">
ces_B_null
</th>
<th style="text-align:right;">
ci_low_95_ces_A0
</th>
<th style="text-align:right;">
ci_high_95_ces_A0
</th>
<th style="text-align:right;">
ci_low_95_ces_B0
</th>
<th style="text-align:right;">
ci_high_95_ces_B0
</th>
<th style="text-align:right;">
ci_low_95_ces_A_on_B
</th>
<th style="text-align:right;">
ci_high_95_ces_A_on_B
</th>
<th style="text-align:right;">
ci_low_95_ces_B_on_A
</th>
<th style="text-align:right;">
ci_high_95_ces_B_on_A
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:right;">
AKT1
</td>
<td style="text-align:right;">
PIK3CA
</td>
<td style="text-align:right;">
3370
</td>
<td style="text-align:right;">
1739
</td>
<td style="text-align:right;">
0.001
</td>
<td style="text-align:right;">
547.634
</td>
<td style="text-align:right;">
0.00215
</td>
<td style="text-align:right;">
0.0014
</td>
<td style="text-align:right;">
2.00e-10
</td>
<td style="text-align:right;">
7.04
</td>
<td style="text-align:right;">
33.5
</td>
<td style="text-align:right;">
0.210
</td>
<td style="text-align:right;">
83
</td>
<td style="text-align:right;">
667
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
1097
</td>
<td style="text-align:right;">
1854
</td>
<td style="text-align:right;">
2653
</td>
<td style="text-align:right;">
1697
</td>
<td style="text-align:right;">
2719
</td>
<td style="text-align:right;">
4116
</td>
<td style="text-align:right;">
1609
</td>
<td style="text-align:right;">
1876
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
886
</td>
<td style="text-align:right;">
234
</td>
<td style="text-align:right;">
1071
</td>
</tr>
<tr>
<td style="text-align:right;">
AKT1
</td>
<td style="text-align:right;">
TP53
</td>
<td style="text-align:right;">
2521
</td>
<td style="text-align:right;">
1616
</td>
<td style="text-align:right;">
3576.737
</td>
<td style="text-align:right;">
0.001
</td>
<td style="text-align:right;">
0.94556
</td>
<td style="text-align:right;">
0.8365
</td>
<td style="text-align:right;">
2.32e-01
</td>
<td style="text-align:right;">
15.36
</td>
<td style="text-align:right;">
21.6
</td>
<td style="text-align:right;">
0.712
</td>
<td style="text-align:right;">
75
</td>
<td style="text-align:right;">
426
</td>
<td style="text-align:right;">
15
</td>
<td style="text-align:right;">
1338
</td>
<td style="text-align:right;">
1854
</td>
<td style="text-align:right;">
2653
</td>
<td style="text-align:right;">
1575
</td>
<td style="text-align:right;">
1993
</td>
<td style="text-align:right;">
3136
</td>
<td style="text-align:right;">
1470
</td>
<td style="text-align:right;">
1773
</td>
<td style="text-align:right;">
2053
</td>
<td style="text-align:right;">
5723
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
1192
</td>
</tr>
<tr>
<td style="text-align:right;">
PIK3CA
</td>
<td style="text-align:right;">
TP53
</td>
<td style="text-align:right;">
1591
</td>
<td style="text-align:right;">
1983
</td>
<td style="text-align:right;">
2343.782
</td>
<td style="text-align:right;">
0.001
</td>
<td style="text-align:right;">
0.23271
</td>
<td style="text-align:right;">
0.0154
</td>
<td style="text-align:right;">
8.42e-06
</td>
<td style="text-align:right;">
126.90
</td>
<td style="text-align:right;">
161.9
</td>
<td style="text-align:right;">
0.784
</td>
<td style="text-align:right;">
551
</td>
<td style="text-align:right;">
318
</td>
<td style="text-align:right;">
123
</td>
<td style="text-align:right;">
862
</td>
<td style="text-align:right;">
1854
</td>
<td style="text-align:right;">
1697
</td>
<td style="text-align:right;">
1575
</td>
<td style="text-align:right;">
1460
</td>
<td style="text-align:right;">
1729
</td>
<td style="text-align:right;">
1802
</td>
<td style="text-align:right;">
2176
</td>
<td style="text-align:right;">
1936
</td>
<td style="text-align:right;">
2813
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
319
</td>
</tr>
</tbody>
</table>
<p>Mutational selection in the two oncogenes appears to be maintained
after mutation in TP53. In contrast, loss of TP53 is less selected after
mutations in AKT1/PIK3CA. One explanation is that cell populations in
which these powerful drivers have fixed must have already subverted
tumor suppression mechanisms, making TP53 mutations superfluous. The two
oncogenes show significant mutual exclusivity, in that selection for
each is reduced after mutation in the other.</p>
<p>Here’s a plot showing the change in selection for mutated sites in
each gene after mutation in the other gene. The epistatic model is not
significantly better than a no-epistasis model for AKT1/TP53. The other
gene pairs do show significant epistatic effects. Asterisks indicate
significant changes in selection for individual genes. Estimates/CIs
depicted on the dashed gray line are at the floor of the inference; such
estimates are consistent with low or negative selection.</p>
<p><img src="cancereffectsizeR_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<p>Even when epistatic effects are significant, it’s not uncommon for
gene-specific changes in selection to be insignificant. The reason is
that patterns of co-occurrence can often be explained by a strong enough
change in selection for either gene individually. (This also explains
why confidence intervals on the parameters are typically wide.)</p>
</div>
</div>
<div class="section level2">
<h2 id="save-your-work">Save your work<a class="anchor" aria-label="anchor" href="#save-your-work"></a>
</h2>
<p>Whether you’ve made it all the way through this tutorial–or simply
want to take a break–you can save a commemorative (and space-efficient)
copy of your CESAnalysis using <code><a href="../reference/save_cesa.html">save_cesa()</a></code>. You can reload
the analysis anytime with <code><a href="../reference/load_cesa.html">load_cesa()</a></code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/save_cesa.html">save_cesa</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, <span class="st">"cancerffectsizeR_tutorial_analysis.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># In some future R session...</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://townsend-lab-yale.github.io/cancereffectsizeR">cancereffectsizeR</a></span><span class="op">)</span></span>
<span><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_cesa.html">load_cesa</a></span><span class="op">(</span><span class="st">"cancerffectsizeR_tutorial_analysis.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
