<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started • cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Getting started">
<meta property="og:description" content="cancereffectsizeR">
<meta property="og:image" content="https://townsend-lab-yale.github.io/cancereffectsizeR/index.html/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancereffectsizeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cancereffectsizeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    </li>
    <li>
      <a href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    </li>
    <li>
      <a href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="cancereffectsizeR_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/vignettes/cancereffectsizeR.Rmd"><code>vignettes/cancereffectsizeR.Rmd</code></a></small>
      <div class="hidden name"><code>cancereffectsizeR.Rmd</code></div>

    </div>

    
    
<p>cancereffectsizeR is used to analyze somatic single-nucleotide variant data in <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format">MAF format</a>. If you want to try out the software with publicly-available data, you can download MAF files covering collections of tumors from many cancer types from the TCGA data portal.</p>
<div id="install-reference-data" class="section level2">
<h2 class="hasAnchor">
<a href="#install-reference-data" class="anchor"></a>Install reference data</h2>
<p>In addition to installing cancereffectsizeR, you need to install a reference data set, or <em>refset</em>. Currently, a refset supporting the hg19 genome build is available as a separate data package. Install like this:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"Townsend-Lab-Yale/ces.refset.hg19@*release"</span><span class="op">)</span></code></pre></div>
<p>Refsets for more genome builds and species may be added in the future; please let us know if you’re interested. There is also full support for <a href="custom_refset_instructions.html">building your own custom refset</a>.</p>
</div>
<div id="maf-pre-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#maf-pre-processing" class="anchor"></a>MAF pre-processing</h2>
<div id="preload_maf" class="section level3">
<h3 class="hasAnchor">
<a href="#preload_maf" class="anchor"></a>preload_maf()</h3>
<p>After acquiring <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/#gdc-maf-format-v100">MAF data</a>, start by running <code><a href="../reference/preload_maf.html">preload_maf()</a></code>, a function that can catch potential problems in your MAF(s). Here is an example of preloading three MAF files:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://townsend-lab-yale.github.io/cancereffectsizeR">cancereffectsizeR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ces.refset.hg19</span><span class="op">)</span>
<span class="va">maf1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="st">"my_maf_1.txt"</span>, refset <span class="op">=</span> <span class="va">ces.refset.hg19</span><span class="op">)</span>

<span class="co"># If you want, load the whole MAF file, rather than just essential columns (may make object large)</span>
<span class="va">maf2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="st">"my_maf_2.txt"</span>, refset <span class="op">=</span> <span class="va">ces.refset.hg19</span>, more_cols <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>

<span class="co"># Imagine this MAF has non-standard column names</span>
<span class="va">maf3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>
  maf <span class="op">=</span> <span class="st">"my_maf_3.txt"</span>, refset <span class="op">=</span> <span class="va">ces.refset.hg19</span>,
  chr_col <span class="op">=</span> <span class="st">"chr"</span>, start_col <span class="op">=</span> <span class="st">"start"</span>, sample_col <span class="op">=</span> <span class="st">"sample_id"</span>
<span class="op">)</span></code></pre></div>
<p>Review any problematic records and fix them if possible. (If you are able to fix the problems, you should save a copy of the corrected data and re-run preload_maf() before continuing.)</p>
<div id="liftover-support" class="section level4">
<h4 class="hasAnchor">
<a href="#liftover-support" class="anchor"></a>LiftOver support</h4>
<p>If the MAF data you are loading is from a different genome build (e.g., you want to use ces.refset.hg19 with hg38 data), you can use the <code>chain_file</code> option to supply a UCSC-style chain file, and then your MAF coordinates will be automatically converted with liftOver.</p>
</div>
</div>
<div id="quality-filtering" class="section level3">
<h3 class="hasAnchor">
<a href="#quality-filtering" class="anchor"></a>Quality filtering</h3>
<p>You want to be confident that nearly all of the somatic variants in your MAF file(s) are real. Pervasive false positives will lead to unreliable effect size estimates. Here are some guidelines:</p>
<ul>
<li>There should be few to no mutations at sites where population databases show common germline variation.</li>
<li>There should be few to no mutations in repetitive or poorly mapped regions of the genome, as mutations in these sites are difficult to call confidently and should be rare anyways since these regions are usually not functionally important at the small-variant level.</li>
<li>When combining data from multiple sources, look out for re-use of samples.</li>
</ul>
<p>When running with ces.refset.hg19 reference data, preloaded MAFs get annotations relevant to these guidelines. (See the <code><a href="../reference/preload_maf.html">preload_maf()</a></code> reference for details.) In this example, we’ll exclude mutations at sites of germline variation, and we’ll also exclude mutations in repetitive regions except when the sites are of known cancer signficance (COSMIC mutation census tiers 1-3). Next, we’ll combine the three MAF data tables and run a helper function <code><a href="../reference/check_sample_overlap.html">check_sample_overlap()</a></code>, to identify possible duplicate samples across data sources. Lastly, we save one analysis-ready MAF per original MAF.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create named list of MAF tables</span>
<span class="va">maf1</span> <span class="op">&lt;-</span> <span class="st">"my_maf_1.txt"</span>
<span class="va">maf2</span> <span class="op">&lt;-</span> <span class="st">"my_maf_2.txt"</span>
<span class="va">maf3</span> <span class="op">&lt;-</span> <span class="st">"my_maf_3.txt"</span>
<span class="va">all_mafs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>maf1 <span class="op">=</span> <span class="va">maf1</span>, maf2 <span class="op">=</span> <span class="va">maf2</span>, maf3 <span class="op">=</span> <span class="va">maf3</span><span class="op">)</span>
<span class="va">combined_maf</span> <span class="op">&lt;-</span> <span class="fu">rbindlist</span><span class="op">(</span><span class="va">all_mafs</span>, idcol <span class="op">=</span> <span class="st">"source"</span>, fill <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>

<span class="co"># remove problematic records (assuming they can't be fixed)</span>
<span class="va">combined_maf</span> <span class="op">&lt;-</span> <span class="va">combined_maf</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">problem</span><span class="op">)</span><span class="op">]</span>

<span class="co"># Exclude records that overlap germline variant sites, and also</span>
<span class="co"># exclude records in repetitive regions except at COSMIC Tier 1-3 sites.</span>
<span class="va">combined_maf</span> <span class="op">&lt;-</span> <span class="va">combined_maf</span><span class="op">[</span><span class="va">germline_variant_site</span> <span class="op">==</span> <span class="cn">F</span> <span class="op">&amp;</span> <span class="op">(</span><span class="va">repetitive_region</span> <span class="op">==</span> <span class="cn">F</span> <span class="op">|</span> <span class="va">cosmic_site_tier</span> <span class="op">%in%</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span>

<span class="co"># check for sample duplicates and sample contamination</span>
<span class="va">possible_dups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_sample_overlap.html">check_sample_overlap</a></span><span class="op">(</span><span class="va">combined_maf</span><span class="op">)</span>

<span class="co"># After manual review, let's say that we've identified 3 samples that should be removed.</span>
<span class="va">samples_to_remove</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample_9"</span>, <span class="st">"Sample_100"</span>, <span class="st">"Sample_150"</span><span class="op">)</span>
<span class="va">combined_maf</span> <span class="op">&lt;-</span> <span class="va">combined_maf</span><span class="op">[</span><span class="op">!</span><span class="va">Unique_Patient_Identifier</span> <span class="op">%in%</span> <span class="va">samples_to_remove</span><span class="op">]</span>

<span class="co"># Divide back into three tables.</span>
<span class="co"># You'll probably want to save these as your analysis-ready MAF text files.</span>
<span class="va">maf_1_final</span> <span class="op">&lt;-</span> <span class="va">combined_maf</span><span class="op">[</span><span class="va">source</span> <span class="op">==</span> <span class="st">"maf1"</span><span class="op">]</span>
<span class="va">maf_2_final</span> <span class="op">&lt;-</span> <span class="va">combined_maf</span><span class="op">[</span><span class="va">source</span> <span class="op">==</span> <span class="st">"maf2"</span><span class="op">]</span>
<span class="va">maf_3_final</span> <span class="op">&lt;-</span> <span class="va">combined_maf</span><span class="op">[</span><span class="va">source</span> <span class="op">==</span> <span class="st">"maf3"</span><span class="op">]</span>

<span class="co"># fwrite(maf_1_final, "maf_1_final.txt", sep = "\t") (etc.)</span></code></pre></div>
<p>Once you have a high-confidence set of variants, <strong>do not</strong> filter out records based on silent mutation status or other functional criteria. cancereffectsizeR needs the full set of high-confidence mutations from all samples. (A note on indels: It’s better to leave indels in the data if they’re present (and to fix them if they’re misformatted), but it’s okay if they’re missing from the data source.)</p>
</div>
</div>
<div id="create-cesanalysis" class="section level2">
<h2 class="hasAnchor">
<a href="#create-cesanalysis" class="anchor"></a>Create CESAnalysis</h2>
<p>The CESAnalysis object is the primary data structure of cancereffectsizeR. A CESAnalysis always has an associated reference data set, or <em>refset</em>. Optionally, you can also define a set of sample groupings, if you want to be able to run future steps with different parameter choices for each group of samples. For example, if you will want to calculate gene mutation rates separately for sample subsets A and B, and run signature analysis separately for subsets 1, 2, and 3, then you may need to define six groups to account for all combinations: A1, A2, A3, B1, B2, B3. If not already present, you can add a group column to the table(s) generated by <code><a href="../reference/preload_maf.html">preload_maf()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create CESAnalysis with default reference data set (ces.refset.hg19)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Same, but with pre-defined sample groups (specific to the data/experiment)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>sample_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"smoking"</span>, <span class="st">"nonsmoking"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Same, but with sample groups that arguably have an evolutionary ordering. The order</span>
<span class="co"># used to define sample_groups doesn't matter; you'll specify the order later.</span>
<span class="co"># (Keep in mind that only 1 sample per individual can be included in analysis.)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>sample_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"early"</span>, <span class="st">"stage3"</span>, <span class="st">"stage4"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="load-maf-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-maf-data" class="anchor"></a>Load MAF data</h2>
<p>Next, use <code><a href="../reference/load_maf.html">load_maf()</a></code> to load your MAF data into your CESAnalysis. If your CESAnalyses specifies sample groups (see Create CESAnalysis, above), specify <code>group_col</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use the more_cols option to keep extra data column(s) during preload_maf</span>
<span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="st">"my_maf_with_groups.txt"</span>, more_cols <span class="op">=</span> <span class="st">"tumor_grade"</span><span class="op">)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>sample_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"I"</span>, <span class="st">"II"</span>, <span class="st">"III"</span><span class="op">)</span><span class="op">)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, maf <span class="op">=</span> <span class="va">maf</span>, group_col <span class="op">=</span> <span class="st">"tumor_grade"</span><span class="op">)</span>

<span class="co"># Or, carefully create the group column yourself using other data.</span>
<span class="co"># This is a simplified example. Often, you'll need to deal with NA values, missing samples, etc.</span>
<span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="st">"my_maf.txt"</span><span class="op">)</span>
<span class="va">smoking_by_sample</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="st">"my_clinical_data.csv"</span><span class="op">)</span>
<span class="va">smokers</span> <span class="op">&lt;-</span> <span class="va">smoking_by_sample</span><span class="op">[</span><span class="va">is_smoker</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="va">Unique_Patient_Identifier</span><span class="op">]</span>

<span class="va">maf</span><span class="op">[</span>, <span class="va">group</span> <span class="op">:=</span> <span class="st">"nonsmoker"</span><span class="op">]</span> <span class="co"># assuming everyone not marked as a smoker is a nonsmoker</span>
<span class="va">maf</span><span class="op">[</span><span class="va">smokers</span>, <span class="va">group</span> <span class="op">:=</span> <span class="st">"smoker"</span>, on <span class="op">=</span> <span class="st">"Unique_Patient_Identifier"</span><span class="op">]</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>sample_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"smoker"</span>, <span class="st">"nonsmoker"</span><span class="op">)</span><span class="op">)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, maf <span class="op">=</span> <span class="va">maf</span>, group_col <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></code></pre></div>
<div id="sequencing-data-coverage" class="section level3">
<h3 class="hasAnchor">
<a href="#sequencing-data-coverage" class="anchor"></a>Sequencing data coverage</h3>
<p>By default, data is assumed to be derived from whole-exome sequencing. Whole-genome data and targeted sequencing data are also supported when the <code>coverage</code> option is specified. For targeted sequencing data, you must provide the set of target regions as a BED file or GRanges object. You can (and should) also supply this information for exome data, if you have it available. See ?load_maf() for details.</p>
</div>
<div id="more-examples" class="section level3">
<h3 class="hasAnchor">
<a href="#more-examples" class="anchor"></a>More examples</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Example 1</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>refset <span class="op">=</span> <span class="st">"ces.refset.hg19"</span><span class="op">)</span>

<span class="co"># load in some whole-exome MAF data (directly from a clean MAF text file)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span><span class="va">cesa</span>, maf <span class="op">=</span> <span class="st">"wes_lung_adenocarcinoma.maf"</span><span class="op">)</span>

<span class="co"># add some whole-genome data (and lift from hg38 to hg19 with a user-provided chain file)</span>
<span class="va">wgs_luad</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span><span class="st">"wgs_luad.maf"</span>, chain_file <span class="op">=</span> <span class="st">"hg38ToHg19.over.chain"</span><span class="op">)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span><span class="va">cesa</span>, maf <span class="op">=</span> <span class="va">wgs_luad</span>, coverage <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span>

<span class="co"># Add some targeted gene sequencing data, supplying a name for the coverage intervals (targeted regions)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span><span class="va">cesa</span>,
  maf <span class="op">=</span> <span class="st">"cancer-gene-targeted_luad.maf"</span>, coverage <span class="op">=</span> <span class="st">"targeted"</span>,
  covered_regions <span class="op">=</span> <span class="st">"my_target_regions.bed"</span>, covered_regions_name <span class="op">=</span> <span class="st">"my_panel"</span>
<span class="op">)</span>


<span class="co">## Example 2</span>
<span class="co"># Create a CESAnalysis with samples divided into tumor stage groups</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>refset <span class="op">=</span> <span class="st">"ces.refset.hg19"</span>, sample_groups <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>

<span class="co"># Load exome data with known coverage intervals (should be available whenever the exome capture kit is known).</span>
<span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="st">"my_multistage.maf"</span>, coverage_intervals_to_check <span class="op">=</span> <span class="st">"my_exome_targets.bed"</span><span class="op">)</span>

<span class="co"># Imagine that we inspected the preloaded MAF and found that a good number of variants</span>
<span class="co"># are just outside covered intervals, which is reasonable since some high-quality</span>
<span class="co"># sequencing reads can be expected to extend past the targeted regions. Therefore, we</span>
<span class="co"># supply covered_regions_padding to accept most of these variants. (Those far outside</span>
<span class="co"># the covered regions may come from multi-mapping reads or other sequencing error that</span>
<span class="co"># should have been filtered out.)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span><span class="va">cesa</span>,
  maf <span class="op">=</span> <span class="va">maf</span>, group_col <span class="op">=</span> <span class="st">"stage"</span>,
  coverage <span class="op">=</span> <span class="st">"exome"</span>, covered_regions <span class="op">=</span> <span class="st">"my_exome_targets.bed"</span>,
  covered_regions_name <span class="op">=</span> <span class="st">"my_exome"</span>, covered_regions_padding <span class="op">=</span> <span class="fl">100</span>
<span class="op">)</span></code></pre></div>
</div>
</div>
<div id="calculate-mutation-rates-and-estimate-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-mutation-rates-and-estimate-selection" class="anchor"></a>Calculate mutation rates and estimate selection</h2>
<p>After loading in data, run the following functions to prepare for calculation of cancer effect size. See each function’s documentation for details.</p>
<ul>
<li>
<code><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code> uses deconstructSigs to extract mutational signature weightings from each tumor’s SNV mutation profile, and from there calculates trinucleotide-context-specific SNV mutation rates. See also <code><a href="../articles/cosmic_cancer_type_note.html">vignette("cosmic_cancer_type_note")</a></code> for advice on excluding certain signatures that may be presumed absent from your tumors.</li>
<li>
<code><a href="../reference/gene_mutation_rates.html">gene_mutation_rates()</a></code> uses dNdScv to calculate gene-level mutation rates. It’s highly recommended to supply tissue-specific covariate data if available. Covariates for many human tissue types are included in <code>ces.refset.hg19</code>. If no appropriate covariates are available, you can <a href="create_custom_covariates.html">make your own</a>, or use dNdScv’s non-specific hg19 covariates.</li>
</ul>
<p>If you have defined <code>sample_groups</code> in your CESAnalysis, you can call each of these functions multiple times, specifying one or more sample groups per run.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create CESAnalysis object and define chronological tumor progression states</span>
<span class="va">analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>refset <span class="op">=</span> <span class="st">"ces.refset.hg19"</span>, sample_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Primary"</span>, <span class="st">"Metastatic"</span><span class="op">)</span><span class="op">)</span>
<span class="va">analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span><span class="va">analysis</span>, maf <span class="op">=</span> <span class="st">"luad_data.maf"</span>, group_col <span class="op">=</span> <span class="st">"pri_met"</span><span class="op">)</span>

<span class="co"># Before running trinuc_mutation_rates, which includes a mutational signature analysis,</span>
<span class="co"># you can run this function to identify possible signatures to exclude. It's recommended</span>
<span class="co"># to manually review the suggestions and decide which signatures to exclude.</span>
<span class="co">#</span>
<span class="co"># Here, we decide to accept all suggestions, so we assign the output of the function</span>
<span class="co"># (a vector of signatures to exclude) to to_remove.</span>
<span class="va">signatures_to_remove</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/suggest_cosmic_signatures_to_remove.html">suggest_cosmic_signatures_to_remove</a></span><span class="op">(</span>cancer_type <span class="op">=</span> <span class="st">"LUAD"</span>, treatment_naive <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>


<span class="va">analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates</a></span><span class="op">(</span><span class="va">analysis</span>,
  signature_set <span class="op">=</span> <span class="st">"COSMIC_v3.1"</span>,
  signatures_to_remove <span class="op">=</span> <span class="va">signatures_to_remove</span>
<span class="op">)</span>

<span class="co"># calculate neutral gene mutation rates within groups, since group differences are likely</span>
<span class="va">analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_mutation_rates.html">gene_mutation_rates</a></span><span class="op">(</span><span class="va">analysis</span>, covariates <span class="op">=</span> <span class="st">"lung"</span>, sample_group <span class="op">=</span> <span class="st">"Primary"</span><span class="op">)</span>
<span class="va">analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gene_mutation_rates.html">gene_mutation_rates</a></span><span class="op">(</span><span class="va">analysis</span>, covariates <span class="op">=</span> <span class="st">"lung"</span>, sample_group <span class="op">=</span> <span class="st">"Metastatic"</span><span class="op">)</span></code></pre></div>
<p>Once all rates have been calculated, you are ready to run <code><a href="../reference/ces_variant.html">ces_variant()</a></code> to find effect sizes under a model of selection. There are two built-in models. <em>sswm</em> and <em>sswm_sequential</em>. (SSWM stands for the “strong selection, weak mutation” assumption of these models. See our <a href="https://doi.org/10.1093/jnci/djy168">paper</a> for details.) The default model, <em>sswm</em>, assumes that each variant has a consistent selection intensity across all samples being tested. <em>sswm_sequential</em> assumes that selection intensity varies among sample groups that are (presumably) in sequential states of disease progression. Naturally, this model can only be used if you have defined sample groups, and if those groups can be thought of as sequential. The group ordering is defined in the call to <code>ces_variant</code>. In this example, primary and metastatic samples are sequential, so we can use <em>sswm_sequential</em>. However, it may still be worth running the default model to get better resolution of the overall effects of lower-prevalence variants.</p>
<p>By default, all variants with frequency &gt;1 will be tested for selection. You can specify precisely which variants to include (and, if you like, include all variants) using the <code><a href="../reference/select_variants.html">select_variants()</a></code> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Calculate effect sizes under SSWM model (by default, on all variants with frequency &gt; 1 in the MAF data)</span>
<span class="co"># supplying optional run names for clarity</span>
<span class="va">analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ces_variant.html">ces_variant</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">analysis</span>, run_name <span class="op">=</span> <span class="st">"general"</span><span class="op">)</span>


<span class="co"># Calculate stage-specific effects under sequential model</span>
<span class="va">analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ces_variant.html">ces_variant</a></span><span class="op">(</span>
  cesa <span class="op">=</span> <span class="va">analysis</span>, model <span class="op">=</span> <span class="st">"sswm_sequential"</span>,
  groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Primary"</span>, <span class="st">"Metastatic"</span><span class="op">)</span>, run_name <span class="op">=</span> <span class="st">"stage_specific"</span>
<span class="op">)</span></code></pre></div>
<p>To visualize highly selected variants from selection output, use <code><a href="../reference/lollipops.html">lollipops()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Examine SIs from three major drivers in the general selection output</span>
<span class="va">si</span> <span class="op">&lt;-</span> <span class="va">analysis</span><span class="op">$</span><span class="va">selection</span><span class="op">$</span><span class="va">general</span>
<span class="va">gene_tables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>KRAS <span class="op">=</span> <span class="va">si</span><span class="op">[</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"KRAS"</span><span class="op">]</span>, EGFR <span class="op">=</span> <span class="va">si</span><span class="op">[</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"EGFR"</span><span class="op">]</span>, TP53 <span class="op">=</span> <span class="va">si</span><span class="op">[</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"TP53"</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="../reference/lollipops.html">lollipops</a></span><span class="op">(</span><span class="va">gene_tables</span>, max_sites <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="kras_egfr_tp53.jpg" alt=""><p class="caption">(not real data)</p>
</div>
<p>We can also visualize selection intensities from the <em>sswm_sequential</em> model:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show early/late SIs of high-frequency variants</span>
<span class="co"># Setting a selection floor ensures variants with frequency of zero in either group don't appear</span>
<span class="va">si</span> <span class="op">&lt;-</span> <span class="va">analysis</span><span class="op">$</span><span class="va">selection</span><span class="op">$</span><span class="va">stage_specific</span><span class="op">[</span><span class="va">si_Primary</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">si_Metastatic</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span>
<span class="fu"><a href="../reference/lollipops.html">lollipops</a></span><span class="op">(</span><span class="va">si</span>, group_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Early"</span>, <span class="st">"Late"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="primary_met_lollipops.jpg" alt=""><p class="caption">(not real data)</p>
</div>
</div>
<div id="going-further" class="section level2">
<h2 class="hasAnchor">
<a href="#going-further" class="anchor"></a>Going further</h2>
<p>Use the <code><a href="../reference/select_variants.html">select_variants()</a></code> function to find variants in your data and view cancereffectsizeR’s annotations. You can use this function to collect variants for input into <code>ces_variant</code>.</p>
<p>In addition to testing for selection of single variants, you can test epistatic models of selection at the variant and gene levels using <code><a href="../reference/ces_epistasis.html">ces_epistasis()</a></code> and <code><a href="../reference/ces_gene_epistasis.html">ces_gene_epistasis()</a></code>. Output can identify cases of both positive epistasis and mutual exclusivity. There is also support for testing custom models of selection in <code><a href="../reference/ces_variant.html">ces_variant()</a></code>; see the function’s documentation for details.</p>
<p>It’s sometimes helpful to combine sets of variants, based on genomic position or functional hypotheses, into batches that can be tested in selection functions as if they were single variants. (The simplifying assumption is that all samples with any mutation in the batch experience similar effects.) We call these <em>compound variants</em>, and you can use them with <code>ces_variant</code> and <code>ces_epistasis</code>. See <code><a href="../reference/define_compound_variants.html">?define_compound_variants</a></code> for details.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '85c732458d5af0ba2cd9feb9eb7f9a97',
    indexName: 'cancereffectsizer',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
