<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generate tissue covariates data • cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Generate tissue covariates data">
<meta property="og:description" content="cancereffectsizeR">
<meta property="og:image" content="https://townsend-lab-yale.github.io/cancereffectsizeR/index.html/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancereffectsizeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cancereffectsizeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    </li>
    <li>
      <a href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    </li>
    <li>
      <a href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="create_custom_covariates_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Generate tissue covariates data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/vignettes/create_custom_covariates.Rmd"><code>vignettes/create_custom_covariates.Rmd</code></a></small>
      <div class="hidden name"><code>create_custom_covariates.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Covariates data specific to a tissue, cancer, or cell type improves the mutation rate estimates produced by <code><a href="../reference/gene_mutation_rates.html">gene_mutation_rates()</a></code>. This guide shows how to generate covariates from experimental data by walking through the creation of the lung tissue covariates offered in version 1.0 of the cancereffectsizeR data package <a href="https://github.com/Townsend-Lab-Yale/ces.refset.hg19">ces.refset.hg19</a>. The source data includes RNA-Seq, histone mark, and replication timing data. The data files are large and can be downloaded from their respective project websites. Note that new versions of some of these data files are available and will possibly be used in future updates to ces.refset.hg19.</p>
<p>Presumably, if you are going to the trouble of generating your own covariates data, your data sources, tissue type, genome build, or even species varies from this example. If you read through the code line by line, it will hopefully be apparent what changes need to be made for compatibility with your data. Please don’t hesitate to contact us with questions.</p>
<p>The key task is to associate the genes defined in your refset (whether ces.refset.hg19, or some custom refset) with the information in your data sources. Usually, this means either matching up gene identifiers or lining up genes with their genomic positions. In the latter case, be sure the genome build of the experimental data matches that used in the refset. (If it doesn’t, you may be able to convert it using a tool such as liftOver.)</p>
<p>Note that if you run through this code exactly with all the same source data, the output covariates data will still differ slightly from the ces.refset.hg19 lung data because of improvements to the workflow.</p>
</div>
<div id="example-lung-tissue-covariates-for-ces-refset-hg19" class="section level1">
<h1 class="hasAnchor">
<a href="#example-lung-tissue-covariates-for-ces-refset-hg19" class="anchor"></a>Example: Lung tissue covariates for ces.refset.hg19</h1>
<p>Start by loading required packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span>

  <span class="co"># Load refset data package, if not using a custom refset directory</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ces.refset.hg19</span><span class="op">)</span></code></pre></div>
<p>We need to gather gene information used by the refset of interest. If using a custom refset directory, use <code>readRDS</code> to load the required RefCDS and gr_genes data; here, we can access the data from the refset environment:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get gene names and gene IDs; remove version number suffixes for short IDs</span>
<span class="va">refset_genes</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/rbindlist.html">rbindlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">ces.refset.hg19</span><span class="op">$</span><span class="va">RefCDS</span>, <span class="st">'['</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_name"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">refset_genes</span><span class="op">[</span>, <span class="va">short_gene_id</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'\\..*'</span>, <span class="st">''</span>, <span class="va">gene_id</span><span class="op">)</span><span class="op">]</span>
<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setkey.html">setkey</a></span><span class="op">(</span><span class="va">refset_genes</span>, <span class="st">"short_gene_id"</span><span class="op">)</span>

<span class="co"># Get genomic intervals for each gene</span>
<span class="va">gene_gr</span> <span class="op">=</span> <span class="va">ces.refset.hg19</span><span class="op">$</span><span class="va">gr_genes</span></code></pre></div>
<p>Next, we’ll load and process the experimental data. Since some processing steps can be time-consuming, you may wish to use <code>saveRDS</code> to save processed data as you go, in case you need to repeat the workflow.</p>
<div id="process-gene-based-experimental-data" class="section level2">
<h2 class="hasAnchor">
<a href="#process-gene-based-experimental-data" class="anchor"></a>Process gene-based experimental data</h2>
<p>First, we’ll load lung tissue RNA-Seq gene-level read counts from GTEx (Genotype-Tissue expression project). The data file is large because it contains many tissue types. We’ll use an accompanying sample attributes file to subset to lung samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gtex</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct.gz"</span><span class="op">)</span>
<span class="va">gtex_key</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"GTEx_v7_Annotations_SampleAttributesDS.txt"</span><span class="op">)</span>

<span class="co"># Subset to lung samples</span>
<span class="va">gtex_lung_samples</span> <span class="op">=</span> <span class="va">gtex_key</span><span class="op">[</span><span class="va">SMTSD</span> <span class="op">==</span> <span class="st">"Lung"</span>, <span class="va">SAMPID</span><span class="op">]</span>
<span class="va">gtex_lung_samples</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setops.html">intersect</a></span><span class="op">(</span><span class="va">gtex_lung_samples</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">gtex</span><span class="op">)</span><span class="op">)</span>
<span class="va">gtex</span> <span class="op">=</span> <span class="va">gtex</span><span class="op">[</span>, <span class="va">.SD</span>, .SDcols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Name"</span>, <span class="va">gtex_lung_samples</span><span class="op">)</span><span class="op">]</span>

<span class="co"># The GTEx Ensembl gene IDs (in "Name") include version suffixes; strip these for short IDs</span>
<span class="va">gtex</span><span class="op">[</span>, <span class="va">short_gene_id</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'\\..*'</span>, <span class="st">''</span>, <span class="va">Name</span><span class="op">)</span><span class="op">]</span>
<span class="va">gtex</span><span class="op">[</span>, <span class="va">Name</span> <span class="op">:=</span> <span class="cn">NULL</span><span class="op">]</span>

<span class="co"># Consider saving work in progress with saveRDS(gtex, "my_gtex.rds")</span></code></pre></div>
<p>We repeat this process with cancer cell line RNA-Seq data from DepMap’s CCLE (Cancer Cell Line Encyclopedia). As above, the data file contains many tissue types, and we subset to lung samples.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ccle</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"CCLE_DepMap_18Q1_RNAseq_RPKM_20180214.gct"</span><span class="op">)</span>
<span class="va">ccle_lung_samples</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ccle</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ccle</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"lung"</span>, ignore.case <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">]</span>
<span class="va">ccle</span> <span class="op">=</span> <span class="va">ccle</span><span class="op">[</span>, <span class="va">.SD</span>, .SDcols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Name"</span>, <span class="va">ccle_lung_samples</span><span class="op">)</span><span class="op">]</span>

<span class="co"># The CCLE Ensembl gene IDs (in "Name") include version suffixes; strip these for short IDs</span>
<span class="va">ccle</span><span class="op">[</span>, <span class="va">short_gene_id</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'\\..*'</span>, <span class="st">''</span>, <span class="va">Name</span><span class="op">)</span><span class="op">]</span>
<span class="va">ccle</span><span class="op">[</span>, <span class="va">Name</span> <span class="op">:=</span> <span class="cn">NULL</span><span class="op">]</span></code></pre></div>
</div>
<div id="process-position-based-experimental-data" class="section level2">
<h2 class="hasAnchor">
<a href="#process-position-based-experimental-data" class="anchor"></a>Process position-based experimental data</h2>
<p>We will use histone mark data from Roadmap Epigenomics.</p>
<p>First, the data should be converted from Wig to BigWig format. It’s possible to do this outside of R with other utilities, but we’ll show how to do it in R. This will result in the creation of new data files with the “.bw” extension. They will be significantly larger than the .wig.gz files, so you may want to delete them later.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">marks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GSM1013123_UCSD.Lung.H3K27ac.STL001.wig.gz"</span>,
          <span class="st">"GSM1059437_UCSD.Lung.H3K36me3.STL001.wig.gz"</span>,
          <span class="st">"GSM1059443_UCSD.Lung.H3K4me1.STL001.wig.gz"</span>,
          <span class="st">"GSM1120355_UCSD.Lung.H3K9me3.STL001.wig.gz"</span>,
          <span class="st">"GSM906395_UCSD.Lung.H3K27ac.STL002.wig.gz"</span>,
          <span class="st">"GSM906411_UCSD.Lung.H3K9me3.STL002.wig.gz"</span>,
          <span class="st">"GSM910572_UCSD.Lung.H3K4me1.STL002.wig.gz"</span>,
          <span class="st">"GSM915336_UCSD.Lung.H3K4me3.STL002.wig.gz"</span>,
          <span class="st">"GSM956014_UCSD.Lung.H3K36me3.STL002.wig.gz"</span><span class="op">)</span>
<span class="va">hg19_seqinfo</span> <span class="op">=</span> <span class="fu">seqinfo</span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg19</span><span class="op">)</span> <span class="co"># or, whatever genome you're using</span>

<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">marks</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">bw</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/wigToBigWig.html">wigToBigWig</a></span><span class="op">(</span><span class="va">marks</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, seqinfo <span class="op">=</span> <span class="va">hg19_seqinfo</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="st">"Finished "</span>, <span class="va">marks</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"."</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># BigWig-formatted files have names based on originals</span>
<span class="va">bw</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'.wig.gz'</span>, <span class="st">'.bw'</span>, <span class="va">marks</span><span class="op">)</span></code></pre></div>
<p>Now, we can load each converted data file as a GRanges object, with a score annotation that indicates the level of enrichment of the given histone mark in each equally-sized genomic interval. We will assign scores to each gene based on the average score across all overlapping genomic intervals.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">marks_by_gene</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span><span class="op">(</span>gene_name <span class="op">=</span> <span class="va">refset_genes</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">bw</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">mark_gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html">import.bw</a></span><span class="op">(</span><span class="va">bw</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">mark_gr</span><span class="op">)</span> <span class="op">=</span> <span class="st">"NCBI"</span> <span class="co"># to match ces.refset.hg19's gene_gr</span>
  <span class="va">overlaps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">mark_gr</span>, subject <span class="op">=</span> <span class="va">gene_gr</span><span class="op">)</span>
  <span class="va">overlaps_dt</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span><span class="op">(</span>chromatin_hits <span class="op">=</span> <span class="fu">queryHits</span><span class="op">(</span><span class="va">overlaps</span><span class="op">)</span>,gene_hits <span class="op">=</span> <span class="fu">subjectHits</span><span class="op">(</span><span class="va">overlaps</span><span class="op">)</span><span class="op">)</span>
  <span class="va">overlaps_dt</span><span class="op">[</span>, <span class="va">gene_name</span> <span class="op">:=</span> <span class="va">gene_gr</span><span class="op">$</span><span class="va">names</span><span class="op">[</span><span class="va">gene_hits</span><span class="op">]</span><span class="op">]</span>
  <span class="va">overlaps_dt</span><span class="op">[</span>, <span class="va">score</span> <span class="op">:=</span> <span class="va">mark_gr</span><span class="op">$</span><span class="va">score</span><span class="op">[</span><span class="va">chromatin_hits</span><span class="op">]</span><span class="op">]</span>
  <span class="va">score_by_gene</span> <span class="op">=</span> <span class="va">overlaps_dt</span><span class="op">[</span>, <span class="fu">.</span><span class="op">(</span>mean_score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">]</span>
  <span class="va">current_mark</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'.bw'</span>, <span class="st">''</span>, <span class="va">bw</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="va">marks_by_gene</span><span class="op">[</span><span class="va">score_by_gene</span>, <span class="op">(</span><span class="va">current_mark</span><span class="op">)</span> <span class="op">:=</span> <span class="va">mean_score</span>, on <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">]</span>
  <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="st">"Finished "</span>, <span class="va">bw</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"."</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># Fill entries with no data (i.e., no chromatin marks) with zeroes</span>
<span class="va">marks_by_gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">marks_by_gene</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span>

<span class="co"># Considering saving: saveRDS(marks_by_gene, "marks_by_gene.rds")</span></code></pre></div>
<p>Finally, we will use replication timing data from ReplicationDomain, processed nearly identically:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rep_timing</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"RT_IMR90_Lung Fibroblast_Int49605910_hg19.bedgraph"</span>,
               <span class="st">"RT_IMR90_Lung Fibroblast_Int78679848_hg19.bedgraph"</span><span class="op">)</span>

<span class="va">rt_by_gene</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span><span class="op">(</span>gene_name <span class="op">=</span> <span class="va">refset_genes</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">rep_timing</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">rt_gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/BEDFile-class.html">import.bedGraph</a></span><span class="op">(</span><span class="va">rep_timing</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="fu">seqlevelsStyle</span><span class="op">(</span><span class="va">rt_gr</span><span class="op">)</span> <span class="op">=</span> <span class="st">"NCBI"</span> <span class="co"># to match ces.refset.hg19's gene_gr</span>
  <span class="va">overlaps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">rt_gr</span>, subject <span class="op">=</span> <span class="va">gene_gr</span><span class="op">)</span>
  <span class="va">overlaps_dt</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/data.table.html">data.table</a></span><span class="op">(</span>rt_hits <span class="op">=</span> <span class="fu">queryHits</span><span class="op">(</span><span class="va">overlaps</span><span class="op">)</span>,gene_hits <span class="op">=</span> <span class="fu">subjectHits</span><span class="op">(</span><span class="va">overlaps</span><span class="op">)</span><span class="op">)</span>
  <span class="va">overlaps_dt</span><span class="op">[</span>, <span class="va">gene_name</span> <span class="op">:=</span> <span class="va">gene_gr</span><span class="op">$</span><span class="va">names</span><span class="op">[</span><span class="va">gene_hits</span><span class="op">]</span><span class="op">]</span>
  <span class="va">overlaps_dt</span><span class="op">[</span>, <span class="va">score</span> <span class="op">:=</span> <span class="va">rt_gr</span><span class="op">$</span><span class="va">score</span><span class="op">[</span><span class="va">rt_hits</span><span class="op">]</span><span class="op">]</span>
  <span class="va">score_by_gene</span> <span class="op">=</span> <span class="va">overlaps_dt</span><span class="op">[</span>, <span class="fu">.</span><span class="op">(</span>mean_score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">]</span>
  <span class="va">current_rt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'_hg19.bedgraph'</span>, <span class="st">''</span>, <span class="va">rep_timing</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
  <span class="va">rt_by_gene</span><span class="op">[</span><span class="va">score_by_gene</span>, <span class="op">(</span><span class="va">current_rt</span><span class="op">)</span> <span class="op">:=</span> <span class="va">mean_score</span>, on <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">]</span>
  <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="st">"Finished "</span>, <span class="va">rep_timing</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"."</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># Fill in zeroes for genes with no RT data</span>
<span class="va">rt_by_gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">rt_by_gene</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></code></pre></div>
</div>
<div id="combine-processed-data-and-run-prcomp" class="section level2">
<h2 class="hasAnchor">
<a href="#combine-processed-data-and-run-prcomp" class="anchor"></a>Combine processed data and run <em>prcomp</em>
</h2>
<p>We combine information across all refset genes that were present in all gene-based data sources. (Position-based data, such as the chromatin and replication timing data, are presumed to cover all genes.)</p>
<p>[Aside: refset genes that are not present in all gene-based data sources will not have any covariates data generated, and cancereffectsizeR’s <code><a href="../reference/gene_mutation_rates.html">gene_mutation_rates()</a></code> will instead use covariates of the nearest available gene. In this example, over 95% of refset.ces.hg19 genes are covered in both CCLE and GTEx data. If you want to use a gene-based data source that leaves out many genes, you may want to adopt another strategy, such as manually inserting an appropriate value for missing genes, or using the values of nearby genes.]</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">final_gene_ids</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">refset_short_gene_ids</span>, <span class="va">ccle</span><span class="op">$</span><span class="va">short_gene_id</span>, <span class="va">gtex</span><span class="op">$</span><span class="va">short_gene_id</span><span class="op">)</span><span class="op">)</span>
<span class="va">covariates_input</span> <span class="op">=</span> <span class="va">refset_genes</span><span class="op">[</span><span class="va">final_gene_ids</span>, <span class="fu">.</span><span class="op">(</span><span class="va">short_gene_id</span>, <span class="va">gene_name</span><span class="op">)</span>, on <span class="op">=</span> <span class="st">"short_gene_id"</span><span class="op">]</span>

<span class="co"># Merge all data for the chosen gene IDs</span>
<span class="va">covariates_input</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html">merge.data.table</a></span><span class="op">(</span><span class="va">covariates_input</span>, <span class="va">gtex</span>, all.x <span class="op">=</span> <span class="cn">T</span>, all.y <span class="op">=</span> <span class="cn">F</span>, by <span class="op">=</span> <span class="st">"short_gene_id"</span><span class="op">)</span>
<span class="va">covariates_input</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html">merge.data.table</a></span><span class="op">(</span><span class="va">covariates_input</span>, <span class="va">ccle</span>, all.x <span class="op">=</span> <span class="cn">T</span>, all.y <span class="op">=</span> <span class="cn">F</span>, by <span class="op">=</span> <span class="st">"short_gene_id"</span><span class="op">)</span>
<span class="va">covariates_input</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html">merge.data.table</a></span><span class="op">(</span><span class="va">covariates_input</span>, <span class="va">marks_by_gene</span>, all.x <span class="op">=</span> <span class="cn">T</span>, all.y <span class="op">=</span> <span class="cn">F</span>, by <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span>
<span class="va">covariates_input</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html">merge.data.table</a></span><span class="op">(</span><span class="va">covariates_input</span>, <span class="va">rt_by_gene</span>, all.x <span class="op">=</span> <span class="cn">T</span>, all.y <span class="op">=</span> <span class="cn">F</span>, by <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span>

<span class="co"># drop any data columns with no variance (none, if using the example data)</span>
<span class="va">has_variance</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">covariates_input</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">uniqueN</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">columns_with_variance</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">has_variance</span><span class="op">)</span><span class="op">[</span><span class="va">has_variance</span><span class="op">]</span>
<span class="va">covariates_input</span> <span class="op">=</span> <span class="va">covariates_input</span><span class="op">[</span>, <span class="va">..columns_with_variance</span><span class="op">]</span></code></pre></div>
<p>We finish by running the PCA function <em>prcomp</em> to generate a prcomp-class object, which contains the covariates information used by cancereffectsizeR, and saving the object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># To format for prcomp, convert to a data-only matrix and transpose</span>
<span class="va">final_gene_names</span> <span class="op">=</span> <span class="va">covariates_input</span><span class="op">$</span><span class="va">gene_name</span>
<span class="va">covariates_input</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">covariates_input</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_name"</span>, <span class="st">"short_gene_id"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">covariates_input</span><span class="op">)</span> <span class="op">=</span> <span class="va">final_gene_names</span>

<span class="co"># Run prcomp (with specification of 20 principal components and data scaling)</span>
<span class="va">covariates_output</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="va">covariates_input</span>, rank. <span class="op">=</span> <span class="fl">20</span>, scale. <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>

<span class="co"># Save it!</span>
<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">covariates_output</span>, <span class="st">"lung.rds"</span><span class="op">)</span></code></pre></div>
<p>The resulting covariates RDS file can be placed in a <a href="custom_refset_instructions.html">custom refset directory</a>, or it can be loaded into an R session with <code>readRDS</code> and passed directly to <code><a href="../reference/gene_mutation_rates.html">gene_mutation_rates()</a></code>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '85c732458d5af0ba2cd9feb9eb7f9a97',
    indexName: 'cancereffectsizer',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
