<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cancereffectsizeR">
<title>Running with custom reference data • cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Running with custom reference data">
<meta property="og:description" content="cancereffectsizeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a>
    <a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Running with custom reference data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/vignettes/custom_refset_instructions.Rmd" class="external-link"><code>vignettes/custom_refset_instructions.Rmd</code></a></small>
      <div class="d-none name"><code>custom_refset_instructions.Rmd</code></div>
    </div>

    
    
<p>A refset is a collection of all the reference data required to run a
cancereffectsizeR analysis. Currently, refsets for the hg38 and hg19
human genome builds are available in the form as <a href="https://github.com/Townsend-Lab-Yale/ces.refset.hg38" class="external-link">ces.refset.hg38</a>
and <a href="https://github.com/Townsend-Lab-Yale/ces.refset.hg19" class="external-link">ces.refset.hg19</a>
data packages. It’s possible to generate your own reference data for
pretty much any species or genome build. Here is the general method:</p>
<ol style="list-style-type: decimal">
<li><p>Find a GTF file containing coding sequence (CDS) definitions,
filter to high-quality and desired CDS regions, and run
<code><a href="../reference/build_RefCDS.html">build_RefCDS()</a></code> to get a collection of gene/transcript/CDS
information.</p></li>
<li><p>Run <code><a href="../reference/create_refset.html">create_refset()</a></code>, which will require your RefCDS
output and some additional information, to save reference data to an
output directory.</p></li>
</ol>
<p>The output of <code>create_refset</code> is usable with
cancereffectsizeR, but there are two additional sources of data that you
will probably want to either add to your refset directory, or at least
have available to supply to cancereffectsizeR functions:</p>
<ol start="3" style="list-style-type: decimal">
<li><p>A CES Signature Set, containing SNV signature definitions and
metadata, is required for mutational signature extraction. As long as
you have a matrix of signature definitions (e.g., see those published by
COSMIC), you can create your own set. You can include one or more
signature set in your refset, or you can pass a signature set as an
argument to <code><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code>. (In either case, you
can also copy a signature set from an existing refset if the genome
build is compatible.) See the details in
<code><a href="../reference/trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code> for how to make a signature set,
then validate with <code><a href="../reference/validate_signature_set.html">validate_signature_set()</a></code>. Finally, to
save with your refset, create a subdirectory called “signatures” in your
refset directory, and then use <code>saveRDS</code> to save your
signature set with the name “[set_name]_signatures.rds”.</p></li>
<li><p>Tissue covariates data inform the calculation of gene mutation
rates when cancereffectsizeR calls dNdScv. Covariates data can be saved
in a refset subdirectory called “covariates”, or you can supply them as
an argument to <code><a href="../reference/gene_mutation_rates.html">gene_mutation_rates()</a></code>. (Example:
“covariates/lung.rds” can be specified with
<code>covariates = "lung"</code>.) <a href="create_custom_covariates.html">This guide</a> shows how to
generate covariates data by combining and processing tissue-specific
experimental data from several sources. As a last resort,
<code>gene_mutation_rates</code> can be run without covariates, but it’s
best to use them if available.</p></li>
</ol>
<p>For an example refset, see the ces.refset.hg38 data package’s refset
directory <a href="https://github.com/Townsend-Lab-Yale/ces.refset.hg38/tree/main/inst/refset" class="external-link">here</a>.
If your refset also uses the hg19 or hg38, you can copy signatures and
covariates from the refset packages into your own refset directory if
desired.</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
