<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="cancereffectsizeR">
<title>MAF filtering and sample validation â€¢ cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="MAF filtering and sample validation">
<meta property="og:description" content="cancereffectsizeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.7.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a>
    <a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>MAF filtering and sample validation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/vignettes/MAF_filtering_tips.Rmd" class="external-link"><code>vignettes/MAF_filtering_tips.Rmd</code></a></small>
      <div class="d-none name"><code>MAF_filtering_tips.Rmd</code></div>
    </div>

    
    
<p>Since our goal is to quantify somatic selection, we want MAF data to
represent a complete set of high-confidence somatic variants for the
sample set. We accordingly expect the following:</p>
<ul>
<li>There should be few to no mutations at sites where population
databases show common germline variation.</li>
<li>There should be few to no mutations in repetitive or poorly mapped
regions of the genome.</li>
<li>Samples should have little mutational overlap, especially at sites
without known cancer association.</li>
</ul>
<div class="section level4">
<h4 id="filtering-to-reduce-false-positive-variants">Filtering to reduce false positive variants<a class="anchor" aria-label="anchor" href="#filtering-to-reduce-false-positive-variants"></a>
</h4>
<p>Well-curated data, such as MAF files produced using the Genomic Data
Commons Aliquot Ensemble Somatic Variant Merging and Masking workflow,
should not need quality filtering. For data produced with other or
unknown somatic calling methods, reading an MAF file with
<code><a href="../reference/preload_maf.html">preload_maf()</a></code> provides three relevant annotation
columns:</p>
<ul>
<li>germline_variant_site: The variant overlaps a region that contains a
common germline variant according to gnomAD (common being &gt;1%
prevalence in some population).</li>
<li>repetitive_region: The variant is in a region of the genome marked
as repetitive by the RepeatMasker tool.</li>
<li>cosmic_site_tier: Indicates if the variant overlaps a site annotated
as cancer-related (tiers 1, 2, and 3) by COSMIC.</li>
</ul>
<p>A simple strategy to reduce false positive calls is to filter out all
germline site records, as well as records from repetitive regions except
for the few with COSMIC annotations. We can apply this filtering like
this:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://townsend-lab-yale.github.io/cancereffectsizeR">cancereffectsizeR</a></span><span class="op">)</span></span>
<span><span class="va">maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span><span class="st">"my_data.maf"</span>, refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span> <span class="co"># also works with ces.refset.hg19</span></span>
<span><span class="va">maf</span> <span class="op">=</span> <span class="va">maf</span><span class="op">[</span><span class="va">germline_variant_site</span> <span class="op">==</span> <span class="cn">F</span><span class="op">]</span><span class="op">[</span><span class="va">repetitive_region</span> <span class="op">==</span> <span class="cn">F</span> <span class="op">|</span> <span class="va">cosmic_site_tier</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="sample-re-use-contamination-and-multi-sample-sequencing">Sample re-use, contamination, and multi-sample sequencing<a class="anchor" aria-label="anchor" href="#sample-re-use-contamination-and-multi-sample-sequencing"></a>
</h4>
<p>When combining data sources, itâ€™s important to verify that a
patientâ€™s mutation data is not duplicated. Since it can be hard to be
sure, we recommend both careful manual curation and the use of
<code><a href="../reference/check_sample_overlap.html">check_sample_overlap()</a></code> to flag possible sample overlap.</p>
<p>Sometimes, patients from the same data source will show suspiciously
high mutational overlap. This could be due to shared calling error, or
worse, contamination between samples during sequencing. If the latter
appears likely, the data should not be used.</p>
<p>Relatedly, patients with multiple distinct sequenced samples
(multi-region sequencing, or multiple timepoints) should contribute just
one sample to an effect analysis, unless there is evidence that the
tissues evolved independently (unusual).</p>
</div>
<div class="section level4">
<h4 id="what-not-to-filter">What not to filter<a class="anchor" aria-label="anchor" href="#what-not-to-filter"></a>
</h4>
<p>We donâ€™t apply the above filters to targeted gene sequencing data
sets, since they presumably come from high-depth sequencing of cancer
hotspots.</p>
<p>To allow a complete picture of the mutational processes present in
tissues for mutation rate estimation, whole-exome/whole-genome variants
should not be filtered on any sort of functional criteria. One thing to
watch out for: Occasionally, researchers will leave out synonymous
variants when publishing their study data. As synonymous variants are
essential for calculating neutral gene mutation rates, they must be
included in any WXS/WGS data.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
