<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial • cancereffectsizeR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Tutorial">
<meta property="og:description" content="cancereffectsizeR">
<meta property="og:image" content="https://townsend-lab-yale.github.io/cancereffectsizeR/index.html/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancereffectsizeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cancereffectsizeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tutorial.html">Tutorial</a>
    </li>
    <li>
      <a href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    </li>
    <li>
      <a href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
    </li>
    <li>
      <a href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tutorial_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/vignettes/tutorial.Rmd"><code>vignettes/tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>tutorial.Rmd</code></div>

    </div>

    
    
<p>Welcome to cancereffectsizeR! This tutorial walks you through some of the package’s key features to analyze publicly-available tumor sequencing data. We recommend that you start a fresh R/RStudio session and set the working directory to a new folder dedicated to this tutorial. From there, you can follow allowing by copy-and-pasting code into an R script (or the console, but it’s nice to be able to save the finished product.)</p>
<div id="installation-and-data-acquisition" class="section level1">
<h1 class="hasAnchor">
<a href="#installation-and-data-acquisition" class="anchor"></a>Installation and data acquisition</h1>
<p>If you haven’t installed the package already, or if it’s been a while since installation, run this to install the latest release:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Some dependencies are large, so we increase the download time limit to be safe</span>
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"Townsend-Lab-Yale/cancereffectsizeR@*release"</span><span class="op">)</span></code></pre></div>
<p>In addition to installing cancereffectsizeR, you need to install a reference data set, or <em>refset</em>. These refsets include genomic and gene annotations, mutational signature definitions, and more. Currently, refsets supporting the hg38 and hg19 builds of the human genome are available as separate data packages. (If you want to build your own refset to use a custom genome build or gene set for almost any species, <a href="articles/custom_refset_instructions.html">you can</a>.) For this tutorial, we’ll use the hg38 refset package. Installation</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">600</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"Townsend-Lab-Yale/ces.refset.hg38@*release"</span><span class="op">)</span></code></pre></div>
</div>
<div id="preparing-data" class="section level1">
<h1 class="hasAnchor">
<a href="#preparing-data" class="anchor"></a>Preparing data</h1>
<p>Load cancereffectsizeR, as well as the data.table package. cancereffectsizeR makes extensive use of data tables, so it’s handy to have the package loaded to be able to take advantage of all data.table functionality.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://townsend-lab-yale.github.io/cancereffectsizeR">cancereffectsizeR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></code></pre></div>
<div id="tcga-data-whole-exome" class="section level2">
<h2 class="hasAnchor">
<a href="#tcga-data-whole-exome" class="anchor"></a>TCGA data (whole-exome)</h2>
<p>We’ll begin by downloading somatic variant data from the TCGA BRCA (breast carcinoma) project. The data comes in a large text file listing somatic mutations for a collection of tumors, in accordance with <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format">MAF format</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcga_maf_file</span> <span class="op">&lt;-</span> <span class="st">"TCGA.BRCA.mutect.995c0111-d90b-4140-bee7-3845436c3b42.DR-10.0.somatic.maf.gz"</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">tcga_maf_file</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://api.gdc.cancer.gov/data/995c0111-d90b-4140-bee7-3845436c3b42"</span>, destfile <span class="op">=</span> <span class="va">tcga_maf_file</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>Let’s also load and examine a table of patient information. Although the TCGA samples in the data set are all taken from primary tumors, you’ll see that some patients presented with metastatic disease (pM = M1). Hormone receptor status (progesterone or estrogen receptor positive) and HER2 amplicfication status are also recorded in a combined column.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcga_clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tutorial/TCGA_BRCA_clinical.txt"</span>, package <span class="op">=</span> <span class="st">"cancereffectsizeR"</span><span class="op">)</span><span class="op">)</span>
<span class="va">tcga_clinical</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<p>The <code><a href="../reference/preload_maf.html">preload_maf()</a></code> takes in MAF data, finds the columns needed by cancereffectsizeR, adds a couple of genomic annotations, and checks for common problems. Since TCGA data sets are well-curated, this partiular run will not reveal any problems. This MAF file already uses the hg38 genome build, but if it didn’t, we could use the <code>chain_file</code> argument to convert records via liftOver.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcga_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preload_maf.html">preload_maf</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">tcga_maf_file</span>, refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span></code></pre></div>
<p>If you compare the TCGA MAF and clinical file, you’ll see we have a small problem: the clinical file uses patient IDs, while the MAF uses tissue sample IDs. (This column was originally called Tumor_Sample_Barcode, but <code><a href="../reference/preload_maf.html">preload_maf()</a></code> changed it to Unique_Patient_Identifier. cancereffectsizeR uses this column name to emphasize the point that all MAF samples must derive from distinct patients.)</p>
<p>Luckily, TCGA IDs are easy to convert. After verifying that each TCGA patient has just one sample in the MAF data, we’ll change tissue IDs to patient IDs.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># TCGA sample IDs start with the patient ID, luckily. Verify a 1-1 sample-patient</span>
<span class="co"># relationship. If this check failed, we would need to figure out a sample exclusion</span>
<span class="co"># strategy before proceeding with analysis.</span>
<span class="va">tcga_maf</span><span class="op">[</span>, <span class="va">patient_id</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="va">Unique_Patient_Identifier</span>, <span class="fl">1</span>, <span class="fl">12</span><span class="op">)</span><span class="op">]</span>
<span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">uniqueN</a></span><span class="op">(</span><span class="va">tcga_samples</span><span class="op">[</span>, <span class="fu">.</span><span class="op">(</span><span class="va">patient_id</span>, <span class="va">Unique_Patient_Identifier</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">uniqueN</a></span><span class="op">(</span><span class="va">tcga_samples</span><span class="op">$</span><span class="va">Unique_Patient_Identifier</span><span class="op">)</span>

<span class="co"># We can now use patient_id as the unique identifier.</span>
<span class="va">tcga_maf</span><span class="op">[</span>, <span class="va">Unique_Patient_Identifier</span> <span class="op">:=</span> <span class="va">patient_id</span><span class="op">]</span>
<span class="va">tcga_maf</span><span class="op">[</span>, <span class="va">patient_id</span> <span class="op">:=</span> <span class="cn">NULL</span><span class="op">]</span> <span class="co"># remove redundant column</span></code></pre></div>
</div>
<div id="pre-analysis-variant-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-analysis-variant-filtering" class="anchor"></a>Pre-analysis variant filtering</h2>
<p>Since our goal is to quantify somatic selection, we want the MAF data to be a complete set of high-confidence somatic variants. It’s probably unwise to go too far down the rabbithole of second-guessing the variant calling methods used in the data sources. However, there is some simple filtering that we often use to reduce the number of false positive calls:</p>
<ul>
<li>There should be few to no mutations at sites where population databases show common germline variation.</li>
<li>There should be few to no mutations in repetitive or poorly mapped regions of the genome, as mutations in these sites are difficult to sequence confidently, and they should be rare anyways since these regions are usually not functionally important at the small-variant level.</li>
</ul>
<p>When using ces.refset.hg38 (hg19), the preloaded MAF has three relevant annotation columns: * germline_variant_site: The variant overlaps a region that contains a common germline variant accoring to gnomAD (common being &gt;1% prevalence in some population). * repetitive_region: The variant is in a region of the genome marked as repetitive by the RepeatMasker tool. * cosmic_site_tier: Indicates if the variant overlaps a site annotated as cancer-related (tiers 1, 2, and 3) by COSMIC.</p>
<p>We suggest filtering out all germline site records, as well as records from repetive regions except for the few with COSMIC annotations. We can apply this filtering like this:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcga_maf</span> <span class="op">&lt;-</span> <span class="va">tcga_maf</span><span class="op">[</span><span class="va">germline_variant_site</span> <span class="op">==</span> <span class="cn">F</span><span class="op">]</span>
<span class="va">tcga_maf</span> <span class="op">&lt;-</span> <span class="va">tcga_maf</span><span class="op">[</span><span class="va">repetitive_region</span> <span class="op">==</span> <span class="cn">F</span> <span class="op">|</span> <span class="va">cosmic_site_tier</span> <span class="op">%in%</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<div id="what-not-to-filter" class="section level4">
<h4 class="hasAnchor">
<a href="#what-not-to-filter" class="anchor"></a>What not to filter</h4>
<p>To get a complete picture of the mutational processes present in tissues, variants should not be filtered on any sort of functional criteria. One thing to watch out for: Occasionally, researchers will leave out synonymous variants when publishing their study data. As synonymous variants are essential for calculating gene mutation rates, the synonymous variants should be tracked down before using the data.</p>
</div>
<div id="notes-on-data-sources" class="section level4">
<h4 class="hasAnchor">
<a href="#notes-on-data-sources" class="anchor"></a>Notes on data sources</h4>
<ul>
<li>While TCGA data is whole-exome, cancereffectsizeR can also incorporate whole-genome and targeted sequencing data from multiple sources into a single analysis.</li>
<li>Targeted sequencing of known cancer genes is useful to increase sample size, with the caveat that at least one exome or genome data source must be included in the analysis, since targeted sequencing data can’t be used for inferences of mutational processes or gene mutation rates. (There are too few mutations, typically, and since they’re mostly in cancer hotspots, they don’t provide a baseline of what mutation rates are like in the absence of selection.)</li>
<li>When combining data sources, or using data from projects that performed multi-region sequencing, it’s important to verify that only one sample per patient is included in your MAF (and in your downstream cancereffectsizeR analysis), since samples from the same patient likely did not evolve independently. Since it’s often hard to be sure that there wasn’t re-use of samples across different data sources, we recommend inputting your MAFs to <code><a href="../reference/check_sample_overlap.html">check_sample_overlap()</a></code> to catch unexpected sample duplication.</li>
</ul>
</div>
</div>
</div>
<div id="create-cesanalysis-and-load-data" class="section level1">
<h1 class="hasAnchor">
<a href="#create-cesanalysis-and-load-data" class="anchor"></a>Create CESAnalysis and load data</h1>
<p>The CESAnalysis is the primary data structure of cancereffectsizeR. The cancereffectsizeR workflow consists of calling a series of functions that take a CESAnalysis as input and return an altered CESAnalysis. Re-using the same variable name will keep things tidy.</p>
<p>Let’s create a CESAnalysis and call <code><a href="../reference/load_maf.html">load_maf()</a></code> to load the TCGA data.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CESAnalysis.html">CESAnalysis</a></span><span class="op">(</span>refset <span class="op">=</span> <span class="st">"ces.refset.hg38"</span><span class="op">)</span>
<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_maf.html">load_maf</a></span><span class="op">(</span>cesa <span class="op">=</span> <span class="va">cesa</span>, maf <span class="op">=</span> <span class="va">tcga_maf</span><span class="op">)</span></code></pre></div>
<p>You will see a message that some variants fall outside of the refset’s exome definitions. Since we don’t know exactly what exome capture techniques were used for the BRCA project (probably a combination of multiple kits at different study sites), this isn’t unexpected. (On the off chance that you do know the exact exome capture intervals of whatever data you’re using, you should supply those intervals with load_maf()’s covered_regions argument.)</p>
<p>To navigate a CESAnalysis, use the dollar sign ($). We can use these accessors to look at the MAF data as well more detailed variant annotations, and if we want we can perform various filtering operations, such as identifying the most prevalent variants.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cesa</span><span class="op">$</span><span class="va">maf</span>
<span class="va">cesa</span><span class="op">$</span><span class="va">variants</span>
<span class="va">cesa</span><span class="op">$</span><span class="va">samples</span>

<span class="co"># Let's see the top variants</span>
<span class="op">(</span><span class="va">top_variants</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">maf_prevalence</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu">.</span><span class="op">(</span><span class="va">variant_name</span>, <span class="va">chr</span>, <span class="va">start</span>, <span class="va">end</span>, <span class="va">maf_prevalence</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>Let’s load the clinical data into the analysis, too. There’s a small data management issue to resolve: the clinical file uses patient IDs, while the MAF data uses tissue sample IDs. Luckily, we can match up IDs using the CESAnalysis samples table.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcga_samples</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">samples</span>



<span class="va">cesa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_sample_data.html">load_sample_data</a></span><span class="op">(</span><span class="va">cesa</span>, <span class="va">tcga_clinical</span><span class="op">)</span>
<span class="va">cesa</span><span class="op">$</span><span class="va">samples</span></code></pre></div>
<p>Now that we’ve loaded the sample data, we can, for example, see how variant counts break down across different sample groups. Let’s see a breakdown PIK3CA mutation status by BRCA receptor phenotype.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get samples with 1 or more PIK3CA mutation</span>
<span class="va">samples_with_pik3ca</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">maf</span><span class="op">[</span><span class="va">top_gene</span> <span class="op">==</span> <span class="st">"PIK3CA"</span>, <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html">unique</a></span><span class="op">(</span><span class="va">Unique_Patient_Identifier</span><span class="op">)</span><span class="op">]</span>
<span class="va">pik3ca_sample_counts</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">samples</span><span class="op">[</span><span class="va">samples_with_pik3ca</span>, <span class="fu">.</span><span class="op">(</span>with_PIK3CA <span class="op">=</span> <span class="va">.N</span><span class="op">)</span>,
  by <span class="op">=</span> <span class="st">"receptor_status"</span>, on <span class="op">=</span> <span class="st">"Unique_Patient_Identifier"</span>
<span class="op">]</span>
<span class="va">total_sample_counts</span> <span class="op">&lt;-</span> <span class="va">cesa</span><span class="op">$</span><span class="va">samples</span><span class="op">[</span>, <span class="va">.N</span>, by <span class="op">=</span> <span class="st">"receptor_status"</span><span class="op">]</span>
<span class="va">combined_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/merge.html">merge.data.table</a></span><span class="op">(</span><span class="va">pik3ca_sample_counts</span>, <span class="va">total_sample_counts</span>, by <span class="op">=</span> <span class="st">"receptor_status"</span><span class="op">)</span>
<span class="va">combined_counts</span><span class="op">[</span>, <span class="va">freq</span> <span class="op">:=</span> <span class="va">with_PIK3CA</span> <span class="op">/</span> <span class="va">N</span><span class="op">]</span>
<span class="va">combined_counts</span></code></pre></div>
<p>PIK3CA mutation is least prevalent in triple negative breast cancer.</p>
<p>We can also look at variant prevalence by metastatic disease status. By default, the <code>variant_counts</code> function considers all mutations, but we’ll subset to recurrent mutations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_by_M</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/variant_counts.html">variant_counts</a></span><span class="op">(</span>
  cesa <span class="op">=</span> <span class="va">cesa</span>, variant_ids <span class="op">=</span> <span class="va">cesa</span><span class="op">$</span><span class="va">variants</span><span class="op">[</span><span class="va">maf_prevalence</span> <span class="op">&gt;</span> <span class="fl">1</span>, <span class="va">variant_id</span><span class="op">]</span>,
  by <span class="op">=</span> <span class="st">"pM"</span>
<span class="op">)</span>

<span class="co"># Might any mutations be protective?</span>
<span class="va">counts_by_M</span><span class="op">[</span><span class="va">M1_prevalence</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/setorder.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">M1_prevalence</span><span class="op">)</span><span class="op">]</span></code></pre></div>
</div>
<div id="relative-trinucleotide-rate-calculation" class="section level1">
<h1 class="hasAnchor">
<a href="#relative-trinucleotide-rate-calculation" class="anchor"></a>Relative trinucleotide rate calculation</h1>
<div outline:0="">
<details><summary>
(Show/hide)
</summary><p><br></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># even code</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Hello world!"</span><span class="op">)</span></code></pre></div>
</details>
</div>
<p>Some text</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '85c732458d5af0ba2cd9feb9eb7f9a97',
    indexName: 'cancereffectsizer',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
