<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>cancereffectsizeR's RefCDS builder — build_RefCDS • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="cancereffectsizeR's RefCDS builder — build_RefCDS"><meta name="description" content="Based on the buildref function in Inigo Martincorena's package dNdScv, this function
takes in gene/transcript/CDS definitions and creates a dNdScv-style RefCDS object and
an associated GenomicRanges object also required to run dNdScv."><meta property="og:description" content="Based on the buildref function in Inigo Martincorena's package dNdScv, this function
takes in gene/transcript/CDS definitions and creates a dNdScv-style RefCDS object and
an associated GenomicRanges object also required to run dNdScv."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a></li>
    <li><a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a></li>
    <li><a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a></li>
    <li><a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>cancereffectsizeR's RefCDS builder</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/build_RefCDS.R" class="external-link"><code>R/build_RefCDS.R</code></a></small>
      <div class="d-none name"><code>build_RefCDS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Based on the buildref function in Inigo Martincorena's package dNdScv, this function
takes in gene/transcript/CDS definitions and creates a dNdScv-style RefCDS object and
an associated GenomicRanges object also required to run dNdScv.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">build_RefCDS</span><span class="op">(</span></span>
<span>  <span class="va">gtf</span>,</span>
<span>  <span class="va">genome</span>,</span>
<span>  use_all_transcripts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cds_ranges_lack_stop_codons <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  additional_essential_splice_pos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  numcode <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  chromosome_style <span class="op">=</span> <span class="st">"NCBI"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gtf">gtf<a class="anchor" aria-label="anchor" href="#arg-gtf"></a></dt>
<dd><p>Path of a Gencode-style GTF file, or an equivalently formatted data
table. See details for required columns (features). It's possible to build such a
table using data pulled from biomaRt, but it's easier to use a GTF.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>Genome assembly name (e.g., "hg19"); an associated BSgenome object must be
available to load. Alternatively, supply a BSgenome object directly.</p></dd>


<dt id="arg-use-all-transcripts">use_all_transcripts<a class="anchor" aria-label="anchor" href="#arg-use-all-transcripts"></a></dt>
<dd><p>T/F (default TRUE): Whether to use all complete transcripts or just the longest
one for each gene.</p></dd>


<dt id="arg-cds-ranges-lack-stop-codons">cds_ranges_lack_stop_codons<a class="anchor" aria-label="anchor" href="#arg-cds-ranges-lack-stop-codons"></a></dt>
<dd><p>The CDS records in Gencode GTFs don't include the
stop codons in their genomic intervals. If your input does include the stop
codons within CDS records, set to FALSE.</p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>how many cores to use for parallel computations</p></dd>


<dt id="arg-additional-essential-splice-pos">additional_essential_splice_pos<a class="anchor" aria-label="anchor" href="#arg-additional-essential-splice-pos"></a></dt>
<dd><p>Usually not needed. A list of
additional essential splice site positions to combine with those calculated
automatically by this function. Each element of the list should have a name
matching a protein_id in the input and consist of a numeric vector of
additional positions. This option exists so that mutations at chr17:7579312
on TP53 are treated as splice site mutations in cancereffectsizeR's default
hg19 reference data set. (Variants at this coding position, which are
always synonymous, have validated effects on splicing, even though the
position misses automatic "essential splice" annotation by 1 base.)</p></dd>


<dt id="arg-numcode">numcode<a class="anchor" aria-label="anchor" href="#arg-numcode"></a></dt>
<dd><p>(don't use) NCBI genetic code number; currently only code 1, the
standard genetic code, is supported</p></dd>


<dt id="arg-chromosome-style">chromosome_style<a class="anchor" aria-label="anchor" href="#arg-chromosome-style"></a></dt>
<dd><p>Chromosome naming style to use. Defaults to "NCBI". For the human
genome, that means 1, 2,..., as opposed to "UCSC" style (chr1, chr2, ...). Value gets
passed to genomeInfoDb's seqlevelsStyle().</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A two-item list: RefCDS (which is itself a big list, with each entry containing
  information on one coding sequence (CDS)), and a GRanges object that defines the
  genomic intervals covered by each CDS.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Required columns are seqnames, start, end, strand, gene_name, gene_id, protein_id, and
type.  Only rows that have type == "CDS" will be used. Strand should be
"+" or "-".</p>
<p>By default, only one the longest complete transcript is used from each gene in the
input. If you set use_all_transcripts = TRUE, then all complete transcripts will be
used, resulting in multiple RefCDS entries for some genes. If you do this, you may
want to first eliminate low-confidence or superfluous transcripts from the input data.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

