<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function calculates variant effect sizes under the chosen model of selection. Under the
default model, a variant is assumed to have a consistent scaled selection coefficient (cancer
effect) across all included samples."><title>Calculate cancer effects of variants — ces_variant • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate cancer effects of variants — ces_variant"><meta property="og:description" content="This function calculates variant effect sizes under the chosen model of selection. Under the
default model, a variant is assumed to have a consistent scaled selection coefficient (cancer
effect) across all included samples."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.10.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
    <a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Calculate cancer effects of variants</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/ces_variant.R" class="external-link"><code>R/ces_variant.R</code></a></small>
      <div class="d-none name"><code>ces_variant.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates variant effect sizes under the chosen model of selection. Under the
default model, a variant is assumed to have a consistent scaled selection coefficient (cancer
effect) across all included samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ces_variant</span><span class="op">(</span></span>
<span>  cesa <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variants <span class="op">=</span> <span class="fu"><a href="select_variants.html">select_variants</a></span><span class="op">(</span><span class="va">cesa</span>, min_freq <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  model <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  run_name <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  lik_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  optimizer_args <span class="op">=</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">model</span>, <span class="st">"default"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"L-BFGS-B"</span>, lower <span class="op">=</span></span>
<span>    <span class="fl">0.001</span>, upper <span class="op">=</span> <span class="fl">1e+09</span><span class="op">)</span> <span class="kw">else</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  return_fit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  hold_out_same_gene_samples <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  conf <span class="op">=</span> <span class="fl">0.95</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>cesa</dt>
<dd><p>CESAnalysis object</p></dd>


<dt>variants</dt>
<dd><p>Which variants to estimate effects for, specified with a variant table such as
from <code>[CESAnalysis]$variants</code> or <code><a href="select_variants.html">select_variants()</a></code>, or a <code>CompoundVariantSet</code>
from <code><a href="define_compound_variants.html">define_compound_variants()</a></code>. Defaults to all recurrent mutations; that is,
<code>[CESAnalysis]$variants[maf_prevalence &gt; 1]</code>. To include all variants, set to
<code>[CESAnalysis]$variants</code>.</p></dd>


<dt>samples</dt>
<dd><p>Which samples to include in inference. Defaults to all samples. Can be a vector of
Unique_Patient_Identifiers, or a data.table containing rows from the CESAnalysis sample table.</p></dd>


<dt>model</dt>
<dd><p>Set to "basic" (default) or "sequential" (not yet available) to use built-in
models of selection, or supply a custom function factory (see details).</p></dd>


<dt>run_name</dt>
<dd><p>Optionally, a name to identify the current run.</p></dd>


<dt>lik_args</dt>
<dd><p>Extra arguments, given as a list, to pass to custom likelihood functions.</p></dd>


<dt>optimizer_args</dt>
<dd><p>Named list of arguments to pass to the optimizer, bbmle::mle2. Use, for example,
to choose optimization algorithm or parameter boundaries on custom models.</p></dd>


<dt>return_fit</dt>
<dd><p>TRUE/FALSE (default FALSE): Embed model fit for each variant in a "fit"
attribute of the selection results table. Use <code>attr(selection_table, 'fit')</code> to access the
list of fitted models. Defaults to FALSE to save memory. Model fit objects can be of moderate
or large size. If you run thousands of variants at once, you may exhaust your system memory.</p></dd>


<dt>hold_out_same_gene_samples</dt>
<dd><p>When finding likelihood of each variant, hold out samples that
lack the variant but have any other mutations in the same gene. By default, TRUE when running
with single variants, FALSE with a CompoundVariantSet.</p></dd>


<dt>cores</dt>
<dd><p>Number of cores to use for processing variants in parallel (not useful for Windows
systems).</p></dd>


<dt>conf</dt>
<dd><p>Cancer effect confidence interval width (NULL skips calculation, speeds runtime). Ignored
when running custom models.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>CESAnalysis object with selection results appended to the selection output list</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Definitions of the sample count columns in the effects output:</p><ul><li><p>included_with_variant: Number of samples that have the variant and were included in the inference.</p></li>
<li><p>included_total: Number of samples that have coverage at the site and were included in the inference.</p></li>
<li><p>held_out: Samples that have coverage at the site, but were held out of the inference due to <code>hold_out_same_gene_samples = TRUE</code>.</p></li>
<li><p>uncovered: Samples that were not included in the inference because their sequencing did not cover the variant site.</p></li>
</ul><p>Note that if a table of samples to include in the inference is specified with <code>samples</code>, any
CESAnalysis samples not present in the table will not be included in any of the above accounts.</p>
<p>It's possible to pass in your own selection model. You'll need to create a "function factory"
that, for any variant, produces a likelihood function that can be evaluated on the data. The
first two arguments must be <code>rates_tumors_with</code> and <code>rates_tumors_without</code>, which give the baseline
site mutation rates in samples with and without the variant. The third argument must be
<code>sample_index</code>, a data.table that associates <code>Unique_Patient_Identifier</code> with group names and
indices. (Your function factory must accept this argument, but it doesn't have to use its value.)
Values for all three of these arguments will be calculated by ces_variant and passed to your
function factory automatically. Your function can take whatever additional arguments you like,
and you can pass in values using <code>lik_args</code>. The likelihood function parameters that
ces_variant will optimize should be named and have default values. See the source code of
<code><a href="sswm_lik.html">sswm_lik()</a></code> for an example.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

