<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Takes in a data.table of MAF data (produced, typically, with preload_maf()) and
identifies samples with relatively high proportions of shared SNV mutations. Some
flagged sample pairs may reflect shared driver mutations or chance overlap of variants
in SNV or sequencing error hotspots. Very high overlap may indicate sample duplication,
re-use of samples across data sources, or within-experiment sample contamination. To limit
the influence of shared calling error, it's recommended to run this function after
any quality filtering of MAF records, as a final step."><title>Catch duplicate samples — check_sample_overlap • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Catch duplicate samples — check_sample_overlap"><meta property="og:description" content="Takes in a data.table of MAF data (produced, typically, with preload_maf()) and
identifies samples with relatively high proportions of shared SNV mutations. Some
flagged sample pairs may reflect shared driver mutations or chance overlap of variants
in SNV or sequencing error hotspots. Very high overlap may indicate sample duplication,
re-use of samples across data sources, or within-experiment sample contamination. To limit
the influence of shared calling error, it's recommended to run this function after
any quality filtering of MAF records, as a final step."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.6.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a>
    <a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Catch duplicate samples</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/preload_maf.R" class="external-link"><code>R/preload_maf.R</code></a></small>
      <div class="d-none name"><code>check_sample_overlap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Takes in a data.table of MAF data (produced, typically, with <code><a href="preload_maf.html">preload_maf()</a></code>) and
identifies samples with relatively high proportions of shared SNV mutations. Some
flagged sample pairs may reflect shared driver mutations or chance overlap of variants
in SNV or sequencing error hotspots. Very high overlap may indicate sample duplication,
re-use of samples across data sources, or within-experiment sample contamination. To limit
the influence of shared calling error, it's recommended to run this function after
any quality filtering of MAF records, as a final step.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">check_sample_overlap</span><span class="op">(</span><span class="va">maf_list</span><span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>maf_list</dt>
<dd><p>A list of data.tables (or a single data.table) with MAF data and cancereffectsizeR-style column names,
as generated by <code><a href="preload_maf.html">preload_maf()</a></code>.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a data.table with overlap statistics</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Sample pairs are flagged when...</p><ul><li><p>Both samples have &lt;6 total SNVs and any shared SNVs.</p></li>
<li><p>Both samples have &lt;21 total SNVs and &gt;1 shared mutation.</p></li>
<li><p>One sample has just 1 or 2 total SNVs and has any overlaps with the other sample.</p></li>
<li><p>The samples have &gt;2 shared SNVs and at least one percent of SNVs are shared (in the sample with fewer SNVs).</p></li>
</ul><p>These thresholds err on the side of reporting too many possible duplicates. In general,
and especially when dealing with targeted sequencing data, the presence of 1 or 2
shared mutations between a pair of samples is not strong evidence of sample
duplication. It's up to the user to filter and interpret the output.</p>
<p>In addition to reporting SNV counts, this function divides the genome into 1000-bp
windows and reports the following:</p><ul><li><p>variant_windows_A: Number of windows in which sample A has a variant.</p></li>
<li><p>variant_windows_B: Same for B.</p></li>
<li><p>windows_shared: Number of windows that contain a variant shared between both samples.</p></li>
</ul><p>Sometimes, samples have little overlap except for a few hotspots that may derive from
shared calling error or highly mutable regions. These window counts can help
distinguish such samples from those with more pervasive SNV overlap.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

