<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a custom refset — create_refset • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Create a custom refset — create_refset"><meta name="description" content="Use this function to create and save a directory of custom reference data that can be
used with cancereffectsizeR instead of supplied refsets like ces.refset.hg19. All
arguments are required except default_exome/exome_interval_padding, which are recommended."><meta property="og:description" content="Use this function to create and save a directory of custom reference data that can be
used with cancereffectsizeR instead of supplied refsets like ces.refset.hg19. All
arguments are required except default_exome/exome_interval_padding, which are recommended."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a></li>
    <li><a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a></li>
    <li><a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a></li>
    <li><a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create a custom refset</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/create_refset.R" class="external-link"><code>R/create_refset.R</code></a></small>
      <div class="d-none name"><code>create_refset.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Use this function to create and save a directory of custom reference data that can be
used with cancereffectsizeR instead of supplied refsets like <code>ces.refset.hg19</code>. All
arguments are required except default_exome/exome_interval_padding, which are recommended.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_refset</span><span class="op">(</span></span>
<span>  <span class="va">output_dir</span>,</span>
<span>  <span class="va">refcds_dndscv</span>,</span>
<span>  refcds_anno <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">species_name</span>,</span>
<span>  <span class="va">genome_build_name</span>,</span>
<span>  <span class="va">BSgenome_name</span>,</span>
<span>  supported_chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span>,</span>
<span>  default_exome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exome_interval_padding <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  transcripts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>Name/path of an existing, writable output directory where all data
will be saved. The name of this directory will serve as the name of the custom refset.</p></dd>


<dt id="arg-refcds-dndscv">refcds_dndscv<a class="anchor" aria-label="anchor" href="#arg-refcds-dndscv"></a></dt>
<dd><p>Transcript information in the two-item list (consisting of RefCDS
and gr_genes) that is output by <code>build_RefCDS</code>. This transcript information will be used with dNdScv.</p></dd>


<dt id="arg-refcds-anno">refcds_anno<a class="anchor" aria-label="anchor" href="#arg-refcds-anno"></a></dt>
<dd><p>Transcript information in the two-item list (consisting of RefCDS
and gr_genes) that is output by <code>build_RefCDS</code>. This transcript information will be used for
cancereffectsizeR's annotations. If unspecified, the same reference information as supplied for dNdScv will be used.</p></dd>


<dt id="arg-species-name">species_name<a class="anchor" aria-label="anchor" href="#arg-species-name"></a></dt>
<dd><p>Name of the species, primarily for display (e.g., "human").</p></dd>


<dt id="arg-genome-build-name">genome_build_name<a class="anchor" aria-label="anchor" href="#arg-genome-build-name"></a></dt>
<dd><p>Name of the genome build, primarily for display (e.g., "hg19").</p></dd>


<dt id="arg-bsgenome-name">BSgenome_name<a class="anchor" aria-label="anchor" href="#arg-bsgenome-name"></a></dt>
<dd><p>The name of the BSgenome package to use (e.g., "hg19"); will used by
cancereffectsizeR to load the reference genome via BSgenome::getBSgenome().</p></dd>


<dt id="arg-supported-chr">supported_chr<a class="anchor" aria-label="anchor" href="#arg-supported-chr"></a></dt>
<dd><p>Character vector of supported chromosomes. Note that cancereffectsizeR uses
NCBI-style chromosome names, which means no chr prefixes ("X", not "chrX"). Mitochondrial
contigs shouldn't be included since they would require special handling that hasn't been
implemented.</p></dd>


<dt id="arg-default-exome">default_exome<a class="anchor" aria-label="anchor" href="#arg-default-exome"></a></dt>
<dd><p>A BED file or GRanges object that defines coding regions in the genome as
might be used by an exome capture kit. This file (or GRanges) might be acquired or generated
from exome capture kit documentation, or alternatively, coding regions defined in a GTF file
(or the granges output by build_RefCDS()).</p></dd>


<dt id="arg-exome-interval-padding">exome_interval_padding<a class="anchor" aria-label="anchor" href="#arg-exome-interval-padding"></a></dt>
<dd><p>Number of bases to pad start/end of each covered interval, to allow
for some variants to be called just outside of targeted regions, where there still may be
pretty good sequencing coverage.</p></dd>


<dt id="arg-transcripts">transcripts<a class="anchor" aria-label="anchor" href="#arg-transcripts"></a></dt>
<dd><p>Additional information about coding (and, optionally, noncoding)
transcripts from a Gencode GTF, supplied as a data.table. See the format provided in
ces.refset.hg38. You'll have to match the format (including column names) pretty closely to get
expected behavior. Noncoding transcripts are represented only by records with transcript_type =
"transcript", and protein-coding transcripts are representing with transcript, CDS, and UTR
records. Note that in Gencode format.</p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>How many cores to use (default 1).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To run this function, you'll need to have output from <code><a href="build_RefCDS.html">build_RefCDS()</a></code>.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

