<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.10.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a></li>
    <li><a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a></li>
    <li><a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a></li>
    <li><a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="create-and-manage-projects">Create and manage projects<a class="anchor" aria-label="anchor" href="#create-and-manage-projects"></a></h2>

      <div class="section-desc"><p>Create, load, and save projects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="CESAnalysis.html">CESAnalysis()</a></code>

        </dt>
        <dd>Create a cancereffectsizeR analysis</dd>
      </dl><dl><dt>

          <code><a href="load_cesa.html">load_cesa()</a></code>

        </dt>
        <dd>Load a previously saved CESAnalysis</dd>
      </dl><dl><dt>

          <code><a href="save_cesa.html">save_cesa()</a></code>

        </dt>
        <dd>Save a CESAnalysis in progress</dd>
      </dl><dl><dt>

          <code><a href="set_refset_dir.html">set_refset_dir()</a></code>

        </dt>
        <dd>Set reference data directory</dd>
      </dl></div><div class="section level2">
      <h2 id="obtain-and-prep-maf-data">Obtain and prep MAF data<a class="anchor" aria-label="anchor" href="#obtain-and-prep-maf-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="preload_maf.html">preload_maf()</a></code>

        </dt>
        <dd>Read and verify MAF somatic mutation data</dd>
      </dl><dl><dt>

          <code><a href="check_sample_overlap.html">check_sample_overlap()</a></code>

        </dt>
        <dd>Catch duplicate samples</dd>
      </dl><dl><dt>

          <code><a href="get_TCGA_project_MAF.html">get_TCGA_project_MAF()</a></code>

        </dt>
        <dd>Get MAF data from TCGA cohort</dd>
      </dl><dl><dt>

          <code><a href="vcfs_to_maf_table.html">vcfs_to_maf_table()</a></code>

        </dt>
        <dd>Read a VCF into an MAF-like table</dd>
      </dl><dl><dt>

          <code><a href="make_PathScore_input.html">make_PathScore_input()</a></code>

        </dt>
        <dd>Make a PathScore input file from MAF data</dd>
      </dl><dl><dt>

          <code><a href="lift_bed.html">lift_bed()</a></code>

        </dt>
        <dd>Convert BED intervals between genome builds</dd>
      </dl></div><div class="section level2">
      <h2 id="load-and-manage-variants">Load and manage variants<a class="anchor" aria-label="anchor" href="#load-and-manage-variants"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="load_maf.html">load_maf()</a></code>

        </dt>
        <dd>Load MAF somatic mutation data</dd>
      </dl><dl><dt>

          <code><a href="select_variants.html">select_variants()</a></code>

        </dt>
        <dd>Select and filter variants</dd>
      </dl><dl><dt>

          <code><a href="variant_counts.html">variant_counts()</a></code>

        </dt>
        <dd>Assess variant prevalence and coverage</dd>
      </dl><dl><dt>

          <code><a href="samples_with.html">samples_with()</a></code>

        </dt>
        <dd>Find samples with specified variants</dd>
      </dl><dl><dt>

          <code><a href="add_variants.html">add_variants()</a></code>

        </dt>
        <dd>Add variant annotations</dd>
      </dl><dl><dt>

          <code><a href="add_covered_regions.html">add_covered_regions()</a></code>

        </dt>
        <dd>add_covered_regions</dd>
      </dl><dl><dt>

          <code><a href="baseline_mutation_rates.html">baseline_mutation_rates()</a></code>

        </dt>
        <dd>Baseline mutation rate calculation</dd>
      </dl></div><div class="section level2">
      <h2 id="load-sample-level-data">Load sample-level data<a class="anchor" aria-label="anchor" href="#load-sample-level-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="load_sample_data.html">load_sample_data()</a></code>

        </dt>
        <dd>Add sample data</dd>
      </dl><dl><dt>

          <code><a href="clear_sample_data.html">clear_sample_data()</a></code>

        </dt>
        <dd>Clear sample data</dd>
      </dl></div><div class="section level2">
      <h2 id="compound-variants">Compound variants<a class="anchor" aria-label="anchor" href="#compound-variants"></a></h2>

      <div class="section-desc"><p>Combine variants into arbitrary batches and test for batch-level selection</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="define_compound_variants.html">define_compound_variants()</a></code>

        </dt>
        <dd>Divide batches of variants into a CompoundVariantSet</dd>
      </dl><dl><dt>

          <code><a href="CompoundVariantSet.html">CompoundVariantSet()</a></code>

        </dt>
        <dd>Create CompoundVariantSet from variant IDs</dd>
      </dl></div><div class="section level2">
      <h2 id="trinucleotide-signatures-and-rates">Trinucleotide signatures and rates<a class="anchor" aria-label="anchor" href="#trinucleotide-signatures-and-rates"></a></h2>

      <div class="section-desc"><p>Mutational signature extraction and inference of context-specific mutation rates</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code>

        </dt>
        <dd>Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes</dd>
      </dl><dl><dt>

          <code><a href="suggest_cosmic_signature_exclusions.html">suggest_cosmic_signature_exclusions()</a></code>

        </dt>
        <dd>Tissue-specific mutational signature exclusions</dd>
      </dl><dl><dt>

          <code><a href="trinuc_snv_counts.html">trinuc_snv_counts()</a></code>

        </dt>
        <dd>Tabulate SNVs by trinucleotide context</dd>
      </dl><dl><dt>

          <code><a href="convert_signature_weights_for_mp.html">convert_signature_weights_for_mp()</a></code>

        </dt>
        <dd>Get MutationalPatterns contributions matrix</dd>
      </dl><dl><dt>

          <code><a href="clear_trinuc_rates_and_signatures.html">clear_trinuc_rates_and_signatures()</a></code>

        </dt>
        <dd>Clear mutational signature attributions and related mutation rate information</dd>
      </dl><dl><dt>

          <code><a href="set_signature_weights.html">set_signature_weights()</a></code>

        </dt>
        <dd>Set SNV signature weights</dd>
      </dl><dl><dt>

          <code><a href="set_trinuc_rates.html">set_trinuc_rates()</a></code>

        </dt>
        <dd>Assign pre-calculated relative trinucleotide mutation rates</dd>
      </dl><dl><dt>

          <code><a href="assign_group_average_trinuc_rates.html">assign_group_average_trinuc_rates()</a></code>

        </dt>
        <dd>Skip mutational signature analysis and assign group average relative trinucleotide-context-specific mutation rates to all samples</dd>
      </dl></div><div class="section level2">
      <h2 id="gene-mutation-rates">Gene mutation rates<a class="anchor" aria-label="anchor" href="#gene-mutation-rates"></a></h2>

      <div class="section-desc"><p>Calculate neutral gene mutation rates</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="gene_mutation_rates.html">gene_mutation_rates()</a></code>

        </dt>
        <dd>Use dNdScv with tissue-specific covariates to calculate gene-level mutation rates</dd>
      </dl><dl><dt>

          <code><a href="set_gene_rates.html">set_gene_rates()</a></code>

        </dt>
        <dd>Assign pre-calculated regional mutation rates</dd>
      </dl><dl><dt>

          <code><a href="clear_gene_rates.html">clear_gene_rates()</a></code>

        </dt>
        <dd>Clear regional mutation rates</dd>
      </dl></div><div class="section level2">
      <h2 id="cancer-effect-sizes">Cancer effect sizes<a class="anchor" aria-label="anchor" href="#cancer-effect-sizes"></a></h2>

      <div class="section-desc"><p>Quantify selection for somatic variants</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="ces_variant.html">ces_variant()</a></code>

        </dt>
        <dd>Calculate cancer effects of variants</dd>
      </dl><dl><dt>

          <code><a href="ces_epistasis.html">ces_epistasis()</a></code>

        </dt>
        <dd>Variant-level pairwise epistasis</dd>
      </dl><dl><dt>

          <code><a href="ces_gene_epistasis.html">ces_gene_epistasis()</a></code>

        </dt>
        <dd>Gene-level epistasis</dd>
      </dl><dl><dt>

          <code><a href="mutational_signature_effects.html">mutational_signature_effects()</a></code>

        </dt>
        <dd>Attribute cancer effects to mutational signatures</dd>
      </dl><dl><dt>

          <code><a href="clear_effect_output.html">clear_effect_output()</a></code>

        </dt>
        <dd>Clear variant effect output</dd>
      </dl><dl><dt>

          <code><a href="clear_epistasis_output.html">clear_epistasis_output()</a></code>

        </dt>
        <dd>Clear epistasis output</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>

      <div class="section-desc"><p>Display and compare variant effect sizes</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="plot_effects.html">plot_effects()</a></code>

        </dt>
        <dd>Plot cancer effects</dd>
      </dl><dl><dt>

          <code><a href="plot_signature_effects.html">plot_signature_effects()</a></code>

        </dt>
        <dd>Plot mutational source and effect attributions</dd>
      </dl><dl><dt>

          <code><a href="plot_epistasis.html">plot_epistasis()</a></code>

        </dt>
        <dd>Plot pairwise epistasis</dd>
      </dl><dl><dt>

          <code><a href="epistasis_plot_schematic.html">epistasis_plot_schematic()</a></code>

        </dt>
        <dd>Get epistatic effect schematic</dd>
      </dl></div><div class="section level2">
      <h2 id="selection-models">Selection models<a class="anchor" aria-label="anchor" href="#selection-models"></a></h2>

      <div class="section-desc"><p>Likelihood function generators for various models of selection</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="sswm_lik.html">sswm_lik()</a></code>

        </dt>
        <dd>sswm_lik</dd>
      </dl><dl><dt>

          <code><a href="pairwise_epistasis_lik.html">pairwise_epistasis_lik()</a></code>

        </dt>
        <dd>pairwise_epistasis_lik</dd>
      </dl></div><div class="section level2">
      <h2 id="explore-reference-data">Explore reference data<a class="anchor" aria-label="anchor" href="#explore-reference-data"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="cosmic_signature_info.html">cosmic_signature_info()</a></code>

        </dt>
        <dd>Get COSMIC signature descriptions</dd>
      </dl><dl><dt>

          <code><a href="get_PathScore_coding_regions.html">get_PathScore_coding_regions()</a></code>

        </dt>
        <dd>Get PathScore coding regions</dd>
      </dl><dl><dt>

          <code><a href="list_ces_refsets.html">list_ces_refsets()</a></code>

        </dt>
        <dd>list_ces_refsets</dd>
      </dl><dl><dt>

          <code><a href="list_ces_covariates.html">list_ces_covariates()</a></code>

        </dt>
        <dd>list_ces_covariates</dd>
      </dl><dl><dt>

          <code><a href="list_ces_signature_sets.html">list_ces_signature_sets()</a></code>

        </dt>
        <dd>list_ces_signature_sets</dd>
      </dl><dl><dt>

          <code><a href="get_ces_signature_set.html">get_ces_signature_set()</a></code>

        </dt>
        <dd>get_ces_signature_set</dd>
      </dl></div><div class="section level2">
      <h2 id="create-custom-reference-data">Create custom reference data<a class="anchor" aria-label="anchor" href="#create-custom-reference-data"></a></h2>

      <div class="section-desc"><p>Build your own reference data set for almost any genome or tissue type</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="create_refset.html">create_refset()</a></code>

        </dt>
        <dd>Create a custom refset</dd>
      </dl><dl><dt>

          <code><a href="build_RefCDS.html">build_RefCDS()</a></code>

        </dt>
        <dd>cancereffectsizeR's RefCDS builder</dd>
      </dl><dl><dt>

          <code><a href="validate_signature_set.html">validate_signature_set()</a></code>

        </dt>
        <dd>validate_signature_set</dd>
      </dl></div><div class="section level2">
      <h2 id="accessors">Accessors<a class="anchor" aria-label="anchor" href="#accessors"></a></h2>

      <div class="section-desc"><p>Data accessors that you probably wonâ€™t need (use cesa$maf, cesa$samples, etc. instead)</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="maf_records.html">maf_records()</a></code>

        </dt>
        <dd>View data loaded into CESAnalysis</dd>
      </dl><dl><dt>

          <code><a href="excluded_maf_records.html">excluded_maf_records()</a></code>

        </dt>
        <dd>View excluded MAF data</dd>
      </dl><dl><dt>

          <code><a href="get_sample_info.html">get_sample_info()</a></code>

        </dt>
        <dd>View sample metadata</dd>
      </dl><dl><dt>

          <code><a href="get_trinuc_rates.html">get_trinuc_rates()</a></code>

        </dt>
        <dd>Get estimated relative rates of trinucleotide-specific SNV mutation</dd>
      </dl><dl><dt>

          <code><a href="get_signature_weights.html">get_signature_weights()</a></code>

        </dt>
        <dd>Get table of signature attributions</dd>
      </dl><dl><dt>

          <code><a href="get_gene_rates.html">get_gene_rates()</a></code>

        </dt>
        <dd>Get table of neutral gene mutation rates</dd>
      </dl><dl><dt>

          <code><a href="snv_results.html">snv_results()</a></code>

        </dt>
        <dd>View results from ces_variant</dd>
      </dl><dl><dt>

          <code><a href="epistasis_results.html">epistasis_results()</a></code>

        </dt>
        <dd>View output from epistasis functions</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

