<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • cancereffectsizeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancereffectsizeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1.9006</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cancereffectsizeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-basic-workflow" class="hasAnchor"><a href="#section-basic-workflow" class="anchor"></a>Basic workflow</h2>
          <p class="section-desc"><p>Functions to run sequentially in most analyses</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="CESAnalysis.html">CESAnalysis()</a></code> </p>
        </td>
        <td><p>Create a cancereffectsizeR analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_maf.html">load_maf()</a></code> </p>
        </td>
        <td><p>Load MAF-formatted somatic mutation data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code> </p>
        </td>
        <td><p>Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_mutation_rates.html">gene_mutation_rates()</a></code> </p>
        </td>
        <td><p>Use dNdScv with tissue-specific covariates to calculate gene-level mutation rates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_variants.html">annotate_variants()</a></code> </p>
        </td>
        <td><p>Annotate variants with data from genome and reference transcripts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ces_snv.html">ces_snv()</a></code> </p>
        </td>
        <td><p>Calculate selection intensity for single-nucleotide variants and amino acid changes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-trinuc-mutation-rates-extensions-and-alternatives" class="hasAnchor"><a href="#section-trinuc-mutation-rates-extensions-and-alternatives" class="anchor"></a>trinuc_mutation_rates extensions and alternatives</h2>
          <p class="section-desc"><p>Additional functions for signature extraction and determination of tumor-specific mutational processes</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="suggest_cosmic_v3_signatures_to_remove.html">suggest_cosmic_v3_signatures_to_remove()</a></code> </p>
        </td>
        <td><p>Identify mutational signatures to exclude from your analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_group_average_trinuc_rates.html">assign_group_average_trinuc_rates()</a></code> </p>
        </td>
        <td><p>Skip mutational signature analysis and assign group average relative trinucleotide-context-specific mutation rates to all samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_trinuc_rates.html">set_trinuc_rates()</a></code> </p>
        </td>
        <td><p>Assign pre-calculated relative trinucleotide mutation rates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_deconstructSigs.html">run_deconstructSigs()</a></code> </p>
        </td>
        <td><p>cancereffectsizeR wrapper for deconstructSigs</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-viewing-cesanalysis-data" class="hasAnchor"><a href="#section-viewing-cesanalysis-data" class="anchor"></a>Viewing CESAnalysis data</h2>
          <p class="section-desc"><p>Data accession functions that you probably won’t need (you can use cesa$maf, cesa$samples, etc. instead)</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="maf_records.html">maf_records()</a></code> </p>
        </td>
        <td><p>View data loaded into CESAnalysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="excluded_maf_records.html">excluded_maf_records()</a></code> </p>
        </td>
        <td><p>View excluded MAF data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="samples.html">samples()</a></code> </p>
        </td>
        <td><p>View sample metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_trinuc_rates.html">get_trinuc_rates()</a></code> </p>
        </td>
        <td><p>Get expected relative trinucleotide-specific SNV mutation rates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_signature_weights.html">get_signature_weights()</a></code> </p>
        </td>
        <td><p>Get table of signature weights by tumor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gene_rates.html">get_gene_rates()</a></code> </p>
        </td>
        <td><p>Get table of neutral gene mutation rates by progression state</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="snv_results.html">snv_results()</a></code> </p>
        </td>
        <td><p>View results from ces_snv</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-miscellaneous" class="hasAnchor"><a href="#section-miscellaneous" class="anchor"></a>Miscellaneous</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="set_genome_data_directory.html">set_genome_data_directory()</a></code> </p>
        </td>
        <td><p>Assign a reference data directory to a CESAnalysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_genome_data_directory.html">get_genome_data_directory()</a></code> </p>
        </td>
        <td><p>Get the path to the reference data directory of a CESAnalysis</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-in-beta" class="hasAnchor"><a href="#section-in-beta" class="anchor"></a>In beta</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="ces_gene_epistasis.html">ces_gene_epistasis()</a></code> </p>
        </td>
        <td><p>Calculate gene-level selection intensity under an assumption of pairwise gene epistasis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_epistasis_results.html">gene_epistasis_results()</a></code> </p>
        </td>
        <td><p>View results from gene-level epistasis analysis</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-basic-workflow">Basic workflow</a></li>
      <li><a href="#section-trinuc-mutation-rates-extensions-and-alternatives">trinuc_mutation_rates extensions and alternatives</a></li>
      <li><a href="#section-viewing-cesanalysis-data">Viewing CESAnalysis data</a></li>
      <li><a href="#section-miscellaneous">Miscellaneous</a></li>
      <li><a href="#section-in-beta">In beta</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


