<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference • cancereffectsizeR</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancereffectsizeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1.9012.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cancereffectsizeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-basic-workflow" class="hasAnchor"><a href="#section-basic-workflow" class="anchor"></a>Basic workflow</h2>
          <p class="section-desc"><p>Core functions for cancereffectsizeR analyses</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="CESAnalysis.html">CESAnalysis()</a></code> </p>
        </td>
        <td><p>Create a cancereffectsizeR analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="load_maf.html">load_maf()</a></code> </p>
        </td>
        <td><p>Load MAF-formatted somatic mutation data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trinuc_mutation_rates.html">trinuc_mutation_rates()</a></code> </p>
        </td>
        <td><p>Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene_mutation_rates.html">gene_mutation_rates()</a></code> </p>
        </td>
        <td><p>Use dNdScv with tissue-specific covariates to calculate gene-level mutation rates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ces_snv.html">ces_snv()</a></code> </p>
        </td>
        <td><p>Calculate selection intensity for single-nucleotide variants and amino acid changes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-epistasis" class="hasAnchor"><a href="#section-epistasis" class="anchor"></a>Epistasis</h2>
          <p class="section-desc"><p>Assess pairwise epistasis between variants or genes</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="ces_gene_epistasis.html">ces_gene_epistasis()</a></code> </p>
        </td>
        <td><p>Gene-level epistasis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ces_epistasis.html">ces_epistasis()</a></code> </p>
        </td>
        <td><p>Variant-level pairwise epistasis</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-get-variant-info" class="hasAnchor"><a href="#section-get-variant-info" class="anchor"></a>Get variant info</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="select_variants.html">select_variants()</a></code> </p>
        </td>
        <td><p>Select and filter variants</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="baseline_mutation_rates.html">baseline_mutation_rates()</a></code> </p>
        </td>
        <td><p>Baseline mutation rate calculation</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-management" class="hasAnchor"><a href="#section-data-management" class="anchor"></a>Data management</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="load_cesa.html">load_cesa()</a></code> </p>
        </td>
        <td><p>Load a previously saved CESAnalysis object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_variants.html">add_variants()</a></code> </p>
        </td>
        <td><p>Add variant annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_covered_regions.html">add_covered_regions()</a></code> </p>
        </td>
        <td><p>add_covered_regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="annotate_variants.html">annotate_variants()</a></code> </p>
        </td>
        <td><p>Annotate variants</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-explore-reference-data" class="hasAnchor"><a href="#section-explore-reference-data" class="anchor"></a>Explore reference data</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="list_ces_ref_sets.html">list_ces_ref_sets()</a></code> </p>
        </td>
        <td><p>list_ces_ref_sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="list_ces_covariates.html">list_ces_covariates()</a></code> </p>
        </td>
        <td><p>list_ces_covariates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="list_ces_signature_sets.html">list_ces_signature_sets()</a></code> </p>
        </td>
        <td><p>list_ces_signature_sets</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_ces_signature_set.html">get_ces_signature_set()</a></code> </p>
        </td>
        <td><p>get_ces_signature_set</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-create-custom-reference-data" class="hasAnchor"><a href="#section-create-custom-reference-data" class="anchor"></a>Create custom reference data</h2>
          <p class="section-desc"><p>More coming soon!</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="build_RefCDS.html">build_RefCDS()</a></code> </p>
        </td>
        <td><p>cancereffectsizeR's RefCDS builder</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-mutational-signatures" class="hasAnchor"><a href="#section-mutational-signatures" class="anchor"></a>Mutational signatures</h2>
          <p class="section-desc"><p>Additional methods and options for mutational signatures and trinuc mutation rates</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="suggest_cosmic_signatures_to_remove.html">suggest_cosmic_signatures_to_remove()</a></code> </p>
        </td>
        <td><p>Identify mutational signatures to exclude from analysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_trinuc_rates.html">set_trinuc_rates()</a></code> </p>
        </td>
        <td><p>Assign pre-calculated relative trinucleotide mutation rates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assign_group_average_trinuc_rates.html">assign_group_average_trinuc_rates()</a></code> </p>
        </td>
        <td><p>Skip mutational signature analysis and assign group average relative trinucleotide-context-specific mutation rates to all samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="run_deconstructSigs.html">run_deconstructSigs()</a></code> </p>
        </td>
        <td><p>cancereffectsizeR wrapper for deconstructSigs</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-accessors" class="hasAnchor"><a href="#section-accessors" class="anchor"></a>Accessors</h2>
          <p class="section-desc"><p>Data accession functions that you probably won’t need (use cesa$maf, cesa$samples, etc. instead)</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="maf_records.html">maf_records()</a></code> </p>
        </td>
        <td><p>View data loaded into CESAnalysis</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="excluded_maf_records.html">excluded_maf_records()</a></code> </p>
        </td>
        <td><p>View excluded MAF data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_sample_info.html">get_sample_info()</a></code> </p>
        </td>
        <td><p>View sample metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_mutations.html">get_mutations()</a></code> </p>
        </td>
        <td><p>Get lists of mutations and annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_trinuc_rates.html">get_trinuc_rates()</a></code> </p>
        </td>
        <td><p>Get expected relative trinucleotide-specific SNV mutation rates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_signature_weights.html">get_signature_weights()</a></code> </p>
        </td>
        <td><p>Get table of signature weights by tumor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gene_rates.html">get_gene_rates()</a></code> </p>
        </td>
        <td><p>Get table of neutral gene mutation rates</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="snv_results.html">snv_results()</a></code> </p>
        </td>
        <td><p>View results from ces_snv</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


