<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load MAF-formatted somatic mutation data — load_maf • cancereffectsizeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Load MAF-formatted somatic mutation data — load_maf" />
<meta property="og:description" content="Load MAF data from a text file or data.frame/data.table into a CESAnalysis object. If column names
don't match MAF format specifications (Chromosome, Start_Position, etc., with Tumor_Sample_Barcode
used as the sample ID column), you can supply your own column names. When your CESAnalysis includes
chronological tumor progression states (e.g., pre- vs. post-treatment, stages 1-4, primary vs. metastatic),
specify &quot;progression_col&quot;. By default, data is assumed to be derived from whole-exome sequencing. Whole-genome
data and targeted sequencing data are also supported when the &quot;coverage&quot; option is specified. If the data
you are loading is from a different genome build than your CESAnalysis, you can use the &quot;chain_file&quot; option 
to supply a UCSC-style chain file, and then your MAF coordinates will be automatically converted with liftOver." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancereffectsizeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1.9005</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cancereffectsizeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load MAF-formatted somatic mutation data</h1>
    <small class="dont-index">Source: <a href='https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/master/R/load_maf.R'><code>R/load_maf.R</code></a></small>
    <div class="hidden name"><code>load_maf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Load MAF data from a text file or data.frame/data.table into a CESAnalysis object. If column names
don't match MAF format specifications (Chromosome, Start_Position, etc., with Tumor_Sample_Barcode
used as the sample ID column), you can supply your own column names. When your CESAnalysis includes
chronological tumor progression states (e.g., pre- vs. post-treatment, stages 1-4, primary vs. metastatic),
specify "progression_col". By default, data is assumed to be derived from whole-exome sequencing. Whole-genome
data and targeted sequencing data are also supported when the "coverage" option is specified. If the data
you are loading is from a different genome build than your CESAnalysis, you can use the "chain_file" option 
to supply a UCSC-style chain file, and then your MAF coordinates will be automatically converted with liftOver.</p>
    </div>

    <pre class="usage"><span class='fu'>load_maf</span>(
  <span class='kw'>cesa</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>maf</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sample_col</span> <span class='kw'>=</span> <span class='st'>"Tumor_Sample_Barcode"</span>,
  <span class='kw'>chr_col</span> <span class='kw'>=</span> <span class='st'>"Chromosome"</span>,
  <span class='kw'>start_col</span> <span class='kw'>=</span> <span class='st'>"Start_Position"</span>,
  <span class='kw'>ref_col</span> <span class='kw'>=</span> <span class='st'>"Reference_Allele"</span>,
  <span class='kw'>tumor_allele_col</span> <span class='kw'>=</span> <span class='st'>"guess"</span>,
  <span class='kw'>coverage</span> <span class='kw'>=</span> <span class='st'>"exome"</span>,
  <span class='kw'>covered_regions</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>covered_regions_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>progression_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>chain_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>enforce_generic_exome_coverage</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>cesa</th>
      <td><p>the CESAnalysis object to load the data into</p></td>
    </tr>
    <tr>
      <th>maf</th>
      <td><p>Path of tab-delimited text file in MAF format, or an MAF in data.table or data.frame format</p></td>
    </tr>
    <tr>
      <th>sample_col</th>
      <td><p>column name with sample ID data (Tumor_Sample_Barcode or Unique_Patient_Identifier)</p></td>
    </tr>
    <tr>
      <th>chr_col</th>
      <td><p>column name with chromosome data  (Chromosome)</p></td>
    </tr>
    <tr>
      <th>start_col</th>
      <td><p>column name with start position (Start_Position)</p></td>
    </tr>
    <tr>
      <th>ref_col</th>
      <td><p>column name with reference allele data (Reference_Allele)</p></td>
    </tr>
    <tr>
      <th>tumor_allele_col</th>
      <td><p>column name with alternate allele data (default uses Tumor_Seq_Allele2 and Tumor_Seq_Allele1 if present)</p></td>
    </tr>
    <tr>
      <th>coverage</th>
      <td><p>exome, genome, or targeted (default exome)</p></td>
    </tr>
    <tr>
      <th>covered_regions</th>
      <td><p>optional for exome, required for targeted: a GRanges object or a BED file of covered intervals matching the CESAnalysis genome</p></td>
    </tr>
    <tr>
      <th>covered_regions_name</th>
      <td><p>a name describing the covered regions (e.g., "my_custom_targeted_regions"); required when covered_regions are supplied</p></td>
    </tr>
    <tr>
      <th>progression_col</th>
      <td><p>column in MAF with tumor progression state (see readme)</p></td>
    </tr>
    <tr>
      <th>chain_file</th>
      <td><p>a chain file (text format, name ends in .chain) to convert MAF records to the genome build used in the CESAnalysis</p></td>
    </tr>
    <tr>
      <th>enforce_generic_exome_coverage</th>
      <td><p>when loading generic exome data, exclude records that aren't covered in the (somewhat arbitrary)
generic exome intervals included with CES genome reference data (default FALSE)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>CESAnalysis object with the specified MAF data loaded (in addition to any previously loaded data)</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


