<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Load MAF data from a text file or data table into your CESAnalysis. Column names are
expected to match MAF format specifications (Chromosome, Start_Position, etc.). It's
encouraged to use preload_maf() prep the input (including, optionally, liftOver
conversion of genomic coordinates), but if you have clean MAF data, you can run this
function directly. By default, data is assumed to be derived from whole-exome
sequencing. Whole-genome data and targeted sequencing data are also supported when the
coverage option is specified."><title>Load MAF somatic mutation data — load_maf • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Load MAF somatic mutation data — load_maf"><meta property="og:description" content="Load MAF data from a text file or data table into your CESAnalysis. Column names are
expected to match MAF format specifications (Chromosome, Start_Position, etc.). It's
encouraged to use preload_maf() prep the input (including, optionally, liftOver
conversion of genomic coordinates), but if you have clean MAF data, you can run this
function directly. By default, data is assumed to be derived from whole-exome
sequencing. Whole-genome data and targeted sequencing data are also supported when the
coverage option is specified."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.6.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a>
    <a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Load MAF somatic mutation data</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/load_maf.R" class="external-link"><code>R/load_maf.R</code></a></small>
      <div class="d-none name"><code>load_maf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Load MAF data from a text file or data table into your CESAnalysis. Column names are
expected to match MAF format specifications (Chromosome, Start_Position, etc.). It's
encouraged to use preload_maf() prep the input (including, optionally, liftOver
conversion of genomic coordinates), but if you have clean MAF data, you can run this
function directly. By default, data is assumed to be derived from whole-exome
sequencing. Whole-genome data and targeted sequencing data are also supported when the
<code>coverage</code> option is specified.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">load_maf</span><span class="op">(</span>
  cesa <span class="op">=</span> <span class="cn">NULL</span>,
  maf <span class="op">=</span> <span class="cn">NULL</span>,
  coverage <span class="op">=</span> <span class="st">"exome"</span>,
  covered_regions <span class="op">=</span> <span class="cn">NULL</span>,
  covered_regions_name <span class="op">=</span> <span class="cn">NULL</span>,
  covered_regions_padding <span class="op">=</span> <span class="fl">0</span>,
  group_col <span class="op">=</span> <span class="cn">NULL</span>,
  sample_data_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span>,
  enforce_default_exome_coverage <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>cesa</dt>
<dd><p>CESAnalysis.</p></dd>
<dt>maf</dt>
<dd><p>Path of tab-delimited text file in MAF format, or an MAF in data.table or
data.frame format.</p></dd>
<dt>coverage</dt>
<dd><p>exome, genome, or targeted (default exome).</p></dd>
<dt>covered_regions</dt>
<dd><p>optional for exome, required for targeted: a GRanges object or a
BED file of covered intervals matching the CESAnalysis genome.</p></dd>
<dt>covered_regions_name</dt>
<dd><p>a name describing the covered regions (e.g.,
"my_custom_targeted_regions"); required when covered_regions are supplied.</p></dd>
<dt>covered_regions_padding</dt>
<dd><p>How many bases (default 0) to expand start and end of
each covered_regions interval, to include variants called just outside of targeted
regions. Consider setting from 0-100bp, or up to the sequencing read length. If the
input data has been trimmed to the targeted regions, leave set to 0.</p></dd>
<dt>group_col</dt>
<dd><p>column in MAF with sample group labels (deprecated; no longer needed).</p></dd>
<dt>sample_data_cols</dt>
<dd><p>MAF columns containing sample-level data (e.g., tumor grade)
that you would like to have copied into the CESAnalysis samples table.</p></dd>
<dt>enforce_default_exome_coverage</dt>
<dd><p>When loading default exome data, exclude records
that aren't covered in the default exome capture intervals included with CES genome
reference data (default FALSE).</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>CESAnalysis with the specified MAF data loaded. The MAF data table includes
  CES-generated variant IDs, a list of all genes overlapping the site, and top_gene and
  top_consequence columns that give the most significant annotated coding changes for
  each mutation record. Annotation precedence is determined by MAF prevalence (usually
  equal), essential splice status, premature stop codon, nonsilent status, MAF mutation
  prevalence across the transcript (often favors longer transcripts), and finally
  alphabetical order. The columns are recalculated when more data is loaded, so changes
  in MAF prevalence can change which variants are highlighted. Note that
<code>[CESAnalysis]$variants</code> contains more information about all top_consequence
  variants and all noncoding variants from the MAF.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer></div>

  

  

  </body></html>

