<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Load MAF somatic mutation data — read_in_maf • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Load MAF somatic mutation data — read_in_maf"><meta name="description" content="Load MAF data from a text file or data table into your CESAnalysis. If column names
don't match MAF format specifications (Chromosome, Start_Position, etc., with
Tumor_Sample_Barcode used as the sample ID column), you can supply your own column
names. When your CESAnalysis has defined sample groups (see ?CESAnalysis),
specify &quot;group_col&quot;. By default, data is assumed to be derived from whole-exome
sequencing. Whole-genome data and targeted sequencing data are also supported when the
&quot;coverage&quot; option is specified. If the data you are loading is from a different genome
build than your CESAnalysis, you can use the &quot;chain_file&quot; option to supply a UCSC-style
chain file, and your MAF coordinates will be automatically converted with
rtracklayer's version of liftOver."><meta property="og:description" content="Load MAF data from a text file or data table into your CESAnalysis. If column names
don't match MAF format specifications (Chromosome, Start_Position, etc., with
Tumor_Sample_Barcode used as the sample ID column), you can supply your own column
names. When your CESAnalysis has defined sample groups (see ?CESAnalysis),
specify &quot;group_col&quot;. By default, data is assumed to be derived from whole-exome
sequencing. Whole-genome data and targeted sequencing data are also supported when the
&quot;coverage&quot; option is specified. If the data you are loading is from a different genome
build than your CESAnalysis, you can use the &quot;chain_file&quot; option to supply a UCSC-style
chain file, and your MAF coordinates will be automatically converted with
rtracklayer's version of liftOver."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a></li>
    <li><a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a></li>
    <li><a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a></li>
    <li><a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Load MAF somatic mutation data</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/internal_read_maf.R" class="external-link"><code>R/internal_read_maf.R</code></a></small>
      <div class="d-none name"><code>read_in_maf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Load MAF data from a text file or data table into your CESAnalysis. If column names
don't match MAF format specifications (Chromosome, Start_Position, etc., with
Tumor_Sample_Barcode used as the sample ID column), you can supply your own column
names. When your CESAnalysis has defined sample groups (see <code><a href="CESAnalysis.html">?CESAnalysis</a></code>),
specify "group_col". By default, data is assumed to be derived from whole-exome
sequencing. Whole-genome data and targeted sequencing data are also supported when the
"coverage" option is specified. If the data you are loading is from a different genome
build than your CESAnalysis, you can use the "chain_file" option to supply a UCSC-style
chain file, and your MAF coordinates will be automatically converted with
rtracklayer's version of liftOver.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">read_in_maf</span><span class="op">(</span></span>
<span>  <span class="va">maf</span>,</span>
<span>  <span class="va">refset_env</span>,</span>
<span>  chr_col <span class="op">=</span> <span class="st">"Chromosome"</span>,</span>
<span>  start_col <span class="op">=</span> <span class="st">"Start_Position"</span>,</span>
<span>  ref_col <span class="op">=</span> <span class="st">"Reference_Allele"</span>,</span>
<span>  tumor_allele_col <span class="op">=</span> <span class="st">"guess"</span>,</span>
<span>  sample_col <span class="op">=</span> <span class="st">"patient_id"</span>,</span>
<span>  more_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chain_file <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  separate_old_problems <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>Path of tab-delimited text file in MAF format, or an MAF in data.table or data.frame format</p></dd>


<dt id="arg-refset-env">refset_env<a class="anchor" aria-label="anchor" href="#arg-refset-env"></a></dt>
<dd><p>a refset data environment</p></dd>


<dt id="arg-chr-col">chr_col<a class="anchor" aria-label="anchor" href="#arg-chr-col"></a></dt>
<dd><p>column name with chromosome data  (Chromosome)</p></dd>


<dt id="arg-start-col">start_col<a class="anchor" aria-label="anchor" href="#arg-start-col"></a></dt>
<dd><p>column name with start position (Start_Position)</p></dd>


<dt id="arg-ref-col">ref_col<a class="anchor" aria-label="anchor" href="#arg-ref-col"></a></dt>
<dd><p>column name with reference allele data (Reference_Allele)</p></dd>


<dt id="arg-tumor-allele-col">tumor_allele_col<a class="anchor" aria-label="anchor" href="#arg-tumor-allele-col"></a></dt>
<dd><p>column name with alternate allele data; by default,
values from Tumor_Seq_Allele2 and Tumor_Seq_Allele1 columns are used.</p></dd>


<dt id="arg-sample-col">sample_col<a class="anchor" aria-label="anchor" href="#arg-sample-col"></a></dt>
<dd><p>column name with sample ID data (Tumor_Sample_Barcode or patient_id)</p></dd>


<dt id="arg-chain-file">chain_file<a class="anchor" aria-label="anchor" href="#arg-chain-file"></a></dt>
<dd><p>a LiftOver chain file (text format, name ends in .chain) to convert MAF
records to the genome build used in the CESAnalysis.</p></dd>


<dt id="arg-separate-old-problems">separate_old_problems<a class="anchor" aria-label="anchor" href="#arg-separate-old-problems"></a></dt>
<dd><p>When TRUE (as used by load_maf), respect old problems that
look like they came from cancereffectsizeR (typically from preload_maf). These get
separated as "old_problem", and the records won't be checked. chain_file must be
NULL.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data.table with core MAF columns, any other requested columns, and a "problem" column</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

