<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function helps you find and view variant data from your CESAnalysis's MAF data and
mutation annotation tables. By default, almost all amino-acid-changing mutations and
noncoding SBS are returned. You can apply a series of filters to restrict output to
certain genes or genomic regions, require a minimum variant frequency in MAF data, and/or
specify exact variant_ids to return."><title>Select and filter variants — select_variants • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Select and filter variants — select_variants"><meta property="og:description" content="This function helps you find and view variant data from your CESAnalysis's MAF data and
mutation annotation tables. By default, almost all amino-acid-changing mutations and
noncoding SBS are returned. You can apply a series of filters to restrict output to
certain genes or genomic regions, require a minimum variant frequency in MAF data, and/or
specify exact variant_ids to return."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a>
    <a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Select and filter variants</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/select_variants.R" class="external-link"><code>R/select_variants.R</code></a></small>
      <div class="d-none name"><code>select_variants.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function helps you find and view variant data from your CESAnalysis's MAF data and
mutation annotation tables. By default, almost all amino-acid-changing mutations and
noncoding SBS are returned. You can apply a series of filters to restrict output to
certain genes or genomic regions, require a minimum variant frequency in MAF data, and/or
specify exact <code>variant_ids</code> to return.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">select_variants</span><span class="op">(</span></span>
<span>  <span class="va">cesa</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_freq <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  variant_ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variant_position_table <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  padding <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>cesa</dt>
<dd><p>CESAnalysis with MAF data loaded and annotated (e.g., with <code><a href="load_maf.html">load_maf()</a></code>)</p></dd>


<dt>genes</dt>
<dd><p>Filter variants to specified genes.</p></dd>


<dt>min_freq</dt>
<dd><p>Filter out variants with MAF frequency below threshold (default 0).
Note that variants that are not in the annotation tables will never be returned. Use
<code><a href="add_variants.html">add_variants()</a></code> to include variants absent from MAF data in your CESAnalysis.</p></dd>


<dt>variant_ids</dt>
<dd><p>Filter out any variants besides those specified. (Variants specified here can
still get removed by other filters.)</p></dd>


<dt>type</dt>
<dd><p>Vector of variant types to include, such as "sbs" or "aac".</p></dd>


<dt>gr</dt>
<dd><p>Filter out any variants not within input GRanges +/- <code>padding</code> bases.</p></dd>


<dt>variant_position_table</dt>
<dd><p>Filter out any variants that don't intersect the
positions given in chr/start/end of this table (1-based closed coordinates).
Typically, the table comes from a previous <code>select_variants</code> call and can be expanded
with <code>padding</code>. (Gritty detail: Amino acid substitutions get special handling. Only the
precise positions in start, end, and center_nt_pos are used. Otherwise, coding changes that
span splice sites would capture all the intronic sites.)</p></dd>


<dt>padding</dt>
<dd><p>add +/- this many bases to every range specified in <code>gr</code> or
<code>variant_position_table</code> (stopping at chromosome ends, naturally).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.table with info on selected variants (see details)</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Only variants that are annotated in the CESAnalysis can be returned. To view more annotations,
such as for variants absent from the MAF data, you must first call add_variants() to add them to
the CESAnalysis.</p>
<p>Note that while intergenic SBS have their nearest genes annotated in the SBS
tables, these variants will not be captured by the genes filter of this function.</p>
<p>Definitions of some less self-explanatory columns:</p><ul><li><p>variant_name: In a coding variant, gene and protein change on the (MANE) canonical
  transcripts, such as "BRAF V600E". For coding changes reported on other transcripts, the
  protein ID is included: "POLH W415C (ENSP00000361300.1)". With older reference data sets
  (ces.refset.hg19, versions of ces.refset.hg38 &lt; 1.3, and any custom reference data set that
  doesn't have complete information on canonical transcripts), the variant name is a 
  shortening of the variant_id.</p></li>
<li><p>start/end: lowest/highest genomic positions overlapping variant</p></li>
<li><p>variant_id: unique IDs for variants given the associated genome assembly version and the transcript data</p></li>
<li><p>ref/alt: genomic reference and alternate alleles (for genomic
  variants; NA for AACs)</p></li>
<li><p>gene: the affected gene in AACs; for SBS, the
  overlapping gene (or an arbitrary gene if more than one overlaps), or the nearest gene
  for intergenic SBS</p></li>
<li><p>strand: for AACs, 1 if the reference sequence strand is the coding strand; -1 otherwise</p></li>
<li><p>essential_splice: Variant is 1,2 bp upstream or
  1,2,5 bp downstream of an annotated splice position (edge case: if an SBS has
  multiple gene/transcript annotations, this doesn't say which one it's essential for)</p></li>
<li><p>intergenic: variant does not overlap any coding regions in the reference data</p></li>
<li><p>trinuc_mut: for SBS, the reference trinucleotide context, in deconstructSigs notation</p></li>
<li><p>coding_seq: coding strand nucleotides in order of transcription</p></li>
<li><p>center_nt_pos: regardless of strand, start/end give positions of two out of three AAC nucleotides; this
                      gives the position of the center nucleotide (maybe useful if the AAC spans a splice site)</p></li>
<li><p>multi_anno_site: T/F whether variant has multiple gene/transcript/AAC annotations</p></li>
<li><p>maf_prevalence: number of occurrences of the variant in MAF data</p></li>
<li><p>samples_covering: number of MAF samples with sequencing coverage at the variant site</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

