<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='This function calculates expected relative rates of trinucleotide-context-specific SBS
mutations within tumors by attributing SBS to mutational processes represented in
mutation signature sets (such as "COSMIC v3.2"). Signature extraction can be done with
MutationalPatterns (default) or deconstructSigs. Tumors with targeted sequencing data
are assigned the average trinucleotide mutation rates calculated across all
exome/genome data, which means that you need at least some exome or genome data to run.'><title>Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes — trinuc_mutation_rates • cancereffectsizeR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes — trinuc_mutation_rates"><meta property="og:description" content='This function calculates expected relative rates of trinucleotide-context-specific SBS
mutations within tumors by attributing SBS to mutational processes represented in
mutation signature sets (such as "COSMIC v3.2"). Signature extraction can be done with
MutationalPatterns (default) or deconstructSigs. Tumors with targeted sequencing data
are assigned the average trinucleotide mutation rates calculated across all
exome/genome data, which means that you need at least some exome or genome data to run.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cancereffectsizeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/cancereffectsizeR.html">Tutorial</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/MAF_filtering_tips.html">MAF filtering and sample validation</a>
    <a class="dropdown-item" href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    <a class="dropdown-item" href="../articles/create_custom_covariates.html">Generate tissue covariates data</a>
    <a class="dropdown-item" href="../articles/custom_refset_instructions.html">Running with custom reference data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes</h1>
      <small class="dont-index">Source: <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/main/R/trinuc_mutation_rates.R" class="external-link"><code>R/trinuc_mutation_rates.R</code></a></small>
      <div class="d-none name"><code>trinuc_mutation_rates.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates expected relative rates of trinucleotide-context-specific SBS
mutations within tumors by attributing SBS to mutational processes represented in
mutation signature sets (such as "COSMIC v3.2"). Signature extraction can be done with
MutationalPatterns (default) or deconstructSigs. Tumors with targeted sequencing data
are assigned the average trinucleotide mutation rates calculated across all
exome/genome data, which means that you need at least some exome or genome data to run.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">trinuc_mutation_rates</span><span class="op">(</span></span>
<span>  <span class="va">cesa</span>,</span>
<span>  signature_set <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  signature_exclusions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  signature_extractor <span class="op">=</span> <span class="st">"MutationalPatterns"</span>,</span>
<span>  mp_strict_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  bootstrap_mutations <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  assume_identical_mutational_processes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sample_group <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sig_averaging_threshold <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  signatures_to_remove <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>cesa</dt>
<dd><p>CESAnalysis object</p></dd>


<dt>signature_set</dt>
<dd><p>Name of built-in signature set (see
<code><a href="list_ces_signature_sets.html">list_ces_signature_sets()</a></code>), or a custom signature set (see details)</p></dd>


<dt>signature_exclusions</dt>
<dd><p>Specify any signatures to exclude from analysis; use
<code><a href="suggest_cosmic_signature_exclusions.html">suggest_cosmic_signature_exclusions()</a></code> for advice on COSMIC signatures</p></dd>


<dt>samples</dt>
<dd><p>Which samples to include in the current run. Defaults to all samples.
Can be a vector of patient_ids, or a data.table containing rows from
the CESAnalysis sample table.</p></dd>


<dt>cores</dt>
<dd><p>How many cores to use for processing tumors in parallel.</p></dd>


<dt>signature_extractor</dt>
<dd><p>One of "MutationalPatterns" (default) or "deconstructSigs".</p></dd>


<dt>mp_strict_args</dt>
<dd><p>Named list of arguments to pass to MutationalPatterns'
fit_to_signatures_strict function. Note that mut_matrix and signatures arguments are
generated automatically, and that if you'd rather not use the strict method, you can
emulate fit_to_signatures() by setting max_delta = 0.</p></dd>


<dt>bootstrap_mutations</dt>
<dd><p>T/F (default FALSE). Instead of using actual sbs counts for
the samples, do a single bootstrap sampling of each sample (in other words, run
MutationalPatterns::fit_to_signatures_bootstrapped() with <code>n_boot=1</code>). This can
be useful if you intend to run this function multiple times to get a distribution of
signature attributions and trinuc rates (and downstream cancer effect sizes). This
option may be replaced with more thorough support for bootstrapping in the future.</p></dd>


<dt>assume_identical_mutational_processes</dt>
<dd><p>use well-mutated tumors (those with number
of eligible mutations meeting sig_averaging_threshold) to calculate group average
signature weights, and assign these (and implied trinucleotide mutation rates) to all tumors</p></dd>


<dt>sig_averaging_threshold</dt>
<dd><p>Mutation prevalence threshold (default 50) that
determines which tumors inform the calculation of group-average signature weights.
When assume_identical_mutational_processes == FALSE (the default), these group
averages are blended into the signature weights of sub-threshold tumors.</p></dd>


<dt>signatures_to_remove</dt>
<dd><p>Deprecated; use the renamed argument signature_exclusions.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>CESAnalysis with sample-specific signature weights and inferred
  trinucleotide-context-specific relative mutation rates. The sbs_counts matrix gives
  the counts of sbs in each trinucleotide context for all samples in the CESAnalysis.
  (While recurrent mutations are excluded from signature analysis in
  trinuc_mutation_rates(), they are present in sbs_counts for completeness.) The
  sbs_counts matrix, produced by `trinuc_sbs_counts()`, can be fed directly into
  MutationalPatterns if you wish to run your own extended signature analysis.</p>
  

<p>The raw_attributions table contains signature attributions as produced by
  MutationalPatterns or deconstructSigs. The biological_weights table has several
  differences:</p>
<p></p>
<ul><li><p>Weights for signatures associated with artifactual (as opposed to biological) processes are set to zero.</p></li>
<li><p>The remaining weights are normalized to sum to 1.</p></li>
<li><p>Tumors with few mutations (defined by `sig_averaging_threshold`, default = 50)
  have their weights redefined using a blend of their original weights and weights
  derived from running signature extraction en masse on tumors with above-threshold
  mutation counts. These samples are identifiable by filtering the table on
  `group_avg_blended == TRUE`, and we recommend excluding them from most downstream
  signature analysis. (These weights are useful as a best
  guess of mutational processes, but they shouldn't be reported in any way that
  implies their independence from the group-average weights.)</p>  
<p>Biological weights can be interpreted as follows: Out of all the mutations caused by
  biological processes represented in the signatures, the proportion of mutations
  attributed to given signature is its weight.</p>  
<p>Either signature attributions table can be converted into the matrix format used by MutationalPatterns 
  with `convert_signature_weights_for_mp()`.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To reduce the influence of selection on the estimation of relative trinucleotide mutation
rates, only non-recurrent SBS (those that do not appear in more than one 
sample in the current run) are used.</p>
<p>A custom signature set should be given as a named three-item list, where "signatures"
is a pure data.frame with signature definitions, "name" is a 1-length character naming
the set, and "metadata" is a data.table with a name column that matches rownames
in the signature definitions. The following columns allow special functionality:</p><ul><li><p>Likely_Artifact (logical T/F): Marks signatures that are believed to
derive from sample processing, sequencing, calling error, or other non-biological
sources. cancereffectsizeR adjusts for artifact signatures when inferring relative
trinucleotide mutation rates.</p></li>
<li><p>Exome_Min: Minimum number of mutations a WES sample
must have for the presence of the signature to be plausible. This information is used to
prevent hypermutation signatures from being found in tumors with few mutations. Can be
left NA or 0 for non-hypermutation signatures. If this column is present, Genome_Min
must be present and always greater than or equal to Exome_Min.</p></li>
<li><p>Genome_Min:
Minimum number of mutations a WGS sample must have for the presence of the signature to
be plausible. This information is used to prevent hypermutation signatures from being found
in tumors with few mutations. Can be left NA or 0 for non-hypermutation signatures. If
this column is present, Exome_Min must be present and always less than or equal to
Genome_Min.</p></li>
</ul><p>If you don't have any metadata available for your signature set, an empty data table
can also be supplied. For a template signature set object, run 
<code>sig_set = get_ces_signature_set("ces.refset.hg19", "COSMIC_v3.2")</code>.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

