<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes — trinuc_mutation_rates • cancereffectsizeR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes — trinuc_mutation_rates" />
<meta property="og:description" content="This function calculates expected relative rates of trinucleotide-context-specific SNV
mutations within tumors by attributing SNVs to mutational processes represented
in mutation signature sets (such as COSMIC v3). deconstructSigs is used to extract mutational 
signature weights in each tumor. Tumors with targeted sequencing data are assigned the
average trinucleotide mutation rates calculated across all exome/genome data, 
which means that you need at least some exome or genome data to run." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cancereffectsizeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1.9009</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/cancereffectsizeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/cosmic_cancer_type_note.html">Cancer type considerations for COSMIC signature extraction</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Townsend-Lab-Yale/cancereffectsizeR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate relative rates of trinucleotide-context-specific mutations by extracting underlying mutational processes</h1>
    <small class="dont-index">Source: <a href='https://github.com/Townsend-Lab-Yale/cancereffectsizeR/blob/master/R/trinuc_mutation_rates.R'><code>R/trinuc_mutation_rates.R</code></a></small>
    <div class="hidden name"><code>trinuc_mutation_rates.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function calculates expected relative rates of trinucleotide-context-specific SNV
mutations within tumors by attributing SNVs to mutational processes represented
in mutation signature sets (such as COSMIC v3). deconstructSigs is used to extract mutational 
signature weights in each tumor. Tumors with targeted sequencing data are assigned the
average trinucleotide mutation rates calculated across all exome/genome data, 
which means that you need at least some exome or genome data to run.</p>
    </div>

    <pre class="usage"><span class='fu'>trinuc_mutation_rates</span>(
  <span class='no'>cesa</span>,
  <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>signature_choice</span> <span class='kw'>=</span> <span class='st'>"cosmic_v3"</span>,
  <span class='kw'>assume_identical_mutational_processes</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>sig_averaging_threshold</span> <span class='kw'>=</span> <span class='fl'>50</span>,
  <span class='kw'>v3_artifact_accounting</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>v3_hypermutation_rules</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>use_dS_exome2genome</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>signatures_to_remove</span> <span class='kw'>=</span> <span class='st'>""</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>cesa</th>
      <td><p>CESAnalysis object</p></td>
    </tr>
    <tr>
      <th>signature_choice</th>
      <td><p>"cosmic_v3" (default), "cosmic_v2", or an equalivalently formatted data.frame of custom trinucleotide signatures</p></td>
    </tr>
    <tr>
      <th>assume_identical_mutational_processes</th>
      <td><p>use well-mutated tumors (those with number of eligible mutations meeting sig_averaging_threshold) 
to calculate group average signature weights, and assign these to all tumors</p></td>
    </tr>
    <tr>
      <th>sig_averaging_threshold</th>
      <td><p>Mutational threshold (default 50) that determines which tumors inform the
calculation of group-average signature weightings. When assume_identical_mutational_processes == FALSE (the default), 
these group averages are blended into the signature weightings of tumors with few mutations (those below the threshold).</p></td>
    </tr>
    <tr>
      <th>v3_artifact_accounting</th>
      <td><p>when COSMIC v3 signatures associated with sequencing artifacts are detected, renormalizes to isolate true sources of mutational flux.</p></td>
    </tr>
    <tr>
      <th>v3_hypermutation_rules</th>
      <td><p>T/F on whether to follow mutation count rules outlined in https://doi.org/10.1101/322859 (COSMIC v3 manuscript)</p></td>
    </tr>
    <tr>
      <th>use_dS_exome2genome</th>
      <td><p>internal dev option (don't use)</p></td>
    </tr>
    <tr>
      <th>signatures_to_remove</th>
      <td><p>specify any signatures to exclude from analysis; use <code><a href='suggest_cosmic_v3_signatures_to_remove.html'>suggest_cosmic_v3_signatures_to_remove()</a></code> for advice on COSMIC v3 signatures</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>CESAnalysis with expected relative trinucleotide mutation rates ($trinuc_rates) and a table of tumor-specific 
        signature weights ($mutational_signatures). Note that tumors with group_avg_blended == TRUE have signature
        weights influenced by the average weights of well-mutated tumors; you may want to exclude these from some analyses.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>To reduce the influence of selection on the estimation of relative trinucleotide mutation
rates, only non-recurrent SNVs (those that do not appear in more than one 
sample in the data set) are used.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Vincent L. Cannataro, Jeff Mandell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


